<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Functions are provided that facilitate the import and analysis of
    SNP (single nucleotide polymorphism) and silicodart (presence/absence) data. 
    The main focus is on data generated by DarT (Diversity Arrays Technology). 
    However, once SNP or related fragment presence/absence data from any source 
    is imported into a genlight object many of the functions can be used. 
    Functions are available for input and output of SNP and silicodart data, for
    reporting on and filtering on various criteria (e.g. CallRate,
    Heterozygosity, Reproducibility, maximum allele frequency). Advanced 
    filtering is based on Linkage Disequilibrium and HWE (Hardy-Weinberg 
    equilibrium). Other functions are available for visualization after PCoA 
    (Principle Coordinate Analysis), or to facilitate transfer of data between 
    genlight/genind objects and newhybrids, related, phylip, structure,
    faststructure packages.">
<title>Importing and Analysing SNP and Silicodart Data Generated by
        Genome-Wide Restriction Fragment Analysis • dartR</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Importing and Analysing SNP and Silicodart Data Generated by
        Genome-Wide Restriction Fragment Analysis">
<meta property="og:description" content="Functions are provided that facilitate the import and analysis of
    SNP (single nucleotide polymorphism) and silicodart (presence/absence) data. 
    The main focus is on data generated by DarT (Diversity Arrays Technology). 
    However, once SNP or related fragment presence/absence data from any source 
    is imported into a genlight object many of the functions can be used. 
    Functions are available for input and output of SNP and silicodart data, for
    reporting on and filtering on various criteria (e.g. CallRate,
    Heterozygosity, Reproducibility, maximum allele frequency). Advanced 
    filtering is based on Linkage Disequilibrium and HWE (Hardy-Weinberg 
    equilibrium). Other functions are available for visualization after PCoA 
    (Principle Coordinate Analysis), or to facilitate transfer of data between 
    genlight/genind objects and newhybrids, related, phylip, structure,
    faststructure packages.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/dartRTutorials.html">dartR Tutorials</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="dartr-">
<code>dartR</code> <a href="https://green-striped-gecko.github.io/dartR/"><img src="reference/figures/dartRlogo.png" align="right" height="140"></a>
<a class="anchor" aria-label="anchor" href="#dartr-"></a>
</h1></div>
<div class="section level2">
<h2 id="an-accessible-genetic-analysis-platform-for-conservation-ecology-and-agriculture">An accessible genetic analysis platform for conservation, ecology and agriculture<a class="anchor" aria-label="anchor" href="#an-accessible-genetic-analysis-platform-for-conservation-ecology-and-agriculture"></a>
</h2>
<!-- badges: start -->
<p>Publication: <a href="https://doi.org/10.1111/1755-0998.12745" class="external-link"><img src="https://img.shields.io/badge/doi-10.1111/1755--0998.12745-00cccc.svg"></a></p>
<p>Zenodo: <a href="https://zenodo.org/badge/latestdoi/86640709" class="external-link"><img src="https://zenodo.org/badge/86640709.svg" alt="DOI"></a></p>
<!-- badges: start -->
</div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><code>dartR</code> is a user-friendly R package that delivers a variety of analyses and pipelines on the same platform, as well as excellent user support via high-quality tutorials and documentation.</p>
<p><code>dartR</code> is a collaboration between the University of Canberra, CSIRO and Diversity Arrays Technology, and is supported with funding from the ACT Priority Investment Program, CSIRO and the University of Canberra.</p>
<p align="center"><img src="reference/figures/UC.png" height="100">     <img src="reference/figures/csiro_logo.png" height="100">     <img src="reference/figures/DArT-removebg-preview.png" height="100">     <img src="reference/figures/ACT.png" height="100"></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><code>dartR</code> is on CRAN, so to install it simply type:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"dartR"</span><span class="op">)</span></span></code></pre></div>
<p>Because of CRAN limitations, the default installation does not include all packages needed to run all the functions; hence, specialised analyses need the installation of additional packages. The following code guarantees that all necessary packages for all <code>dartR</code> analyses are set up and ready to go.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/gl.install.vanilla.dartR.html">gl.install.vanilla.dartR</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>You can install the development version of <code>dartR</code> from GitHub with:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/gl.install.vanilla.dartR.html">gl.install.vanilla.dartR</a></span><span class="op">(</span>flavour <span class="op">=</span> <span class="st">"dev"</span><span class="op">)</span></span></code></pre></div>
<p>Please consult <a href="https://github.com/green-striped-gecko/dartR/wiki/Installation-tutorial" class="external-link">this installation tutorial</a> if you run into any problems during setup.</p>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p><code>dartR</code> provides several functions for handling all the steps involved in genetic data analysis, from reading multiple data input formats to manipulating, filtering, exploring and analysing the data.</p>
<p align="center"><img src="reference/figures/Figure_1.png" width="800"></p>
<p>We use the prefix ‘gl’ in function names to acknowledge the use of the genlight object from package <a href="https://doi.org/10.1093/bioinformatics/btn129" class="external-link">adegenet</a> as our input format.</p>
<p>In most cases, the following term in the function name indicates a subset of functions.</p>
<p>Similarly, in most cases the first function parameter (input) is the genlight object.</p>
<p>For instance, you might use the code below to generate a report and then filter your data based on the percentage of missing data:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://green-striped-gecko.github.io/dartR/">dartR</a></span><span class="op">)</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="va">platypus.gl</span></span>
<span><span class="va">rep</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/gl.report.callrate.html">gl.report.callrate</a></span><span class="op">(</span><span class="va">test</span><span class="op">)</span></span>
<span><span class="va">test_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/gl.filter.callrate.html">gl.filter.callrate</a></span><span class="op">(</span><span class="va">test</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<ol style="list-style-type: decimal">
<li><p>Are you a R rookie? If you want to learn R and RStudio without any fuss, have a look at our <a href="http://georges.biomatix.org/storage/app/media/uploaded-files/Tutorial_1_dartR_RStudio_Refresher_22-Dec-21.pdf" class="external-link">R-refresher tutorial</a>.</p></li>
<li><p>Let’s get started by reading your genetic data into <code>dartR</code>; if you have DArT data, follow <a href="http://georges.biomatix.org/storage/app/media/uploaded-files/tutorial3adartrdatastructuresandinput22-dec-21-2.pdf" class="external-link">this tutorial</a>; if not, follow <a href="http://georges.biomatix.org/storage/app/media/uploaded-files/tutorial3bdartrdatastructuresandinputfromsourcesotherthandartlmagv2-2.pdf" class="external-link">this one</a>.</p></li>
<li><p>Checking out our <a href="http://georges.biomatix.org/storage/app/media/uploaded-files/tutorial4dartrdatamanipulation22-dec-21-3.pdf" class="external-link">data manipulation tutorial</a> is the easiest way to get your feet wet with <code>dartR</code>.</p></li>
<li><p><a href="http://georges.biomatix.org/storage/app/media/uploaded-files/tutorial5dartrbasicfiltering22-dec-21-2.pdf" class="external-link">This tutorial</a> will provide some pointers on how to filter your data effectively, an important step that depends on making sound threshold assessments.</p></li>
<li><p>Check out our <a href="https://github.com/green-striped-gecko/dartR/wiki/Simulations-tutorial" class="external-link">simulations tutorial</a> to learn more about our simulation model, a powerful tool for illuminating intricate evolutionary and genetic processes.</p></li>
<li><p>In more advanced topics, check our technical notes on <a href="http://georges.biomatix.org/storage/app/media/uploaded-files/TECHNICAL_NOTE_Genetic_Distance_18-Feb-22.pdf" class="external-link">Genetic Distances and their Visualization in Population Genetics</a> and <a href="http://georges.biomatix.org/storage/app/media/uploaded-files/TechNote_fixed_difference_analysis_25-Feb-22.pdf" class="external-link">Fixed Difference Analysis</a>.</p></li>
</ol>
</div>
<div class="section level2">
<h2 id="getting-help">Getting help<a class="anchor" aria-label="anchor" href="#getting-help"></a>
</h2>
<ol style="list-style-type: decimal">
<li><p>Google groups Q&amp;A forum in support of users can be accessed <a href="https://groups.google.com/g/dartr?pli=1" class="external-link">here</a>.</p></li>
<li><p>The <a href="https://community.rstudio.com/" class="external-link">RStudio community</a> provides a welcoming environment in which to ask any inquiry.</p></li>
<li><p>Answers to frequently asked topics may usually be found on <a href="https://stackoverflow.com/questions/tagged/r" class="external-link">Stack Overflow</a>.</p></li>
</ol>
</div>
<div class="section level2">
<h2 id="contribute">Contribute<a class="anchor" aria-label="anchor" href="#contribute"></a>
</h2>
<p>If you want to help shape the future of <code>dartR</code>, <a href="http://georges.biomatix.org/storage/app/media/uploaded-files/Tutorial_0_dartR_for_the_Developer_2.0_19-Feb-22.pdf" class="external-link">this tutorial</a> is for you.</p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Please acknowledge <code>dartR</code> if you use it in your study. Copy and paste the following code to the R console to retrieve the citation information:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"dartR"</span><span class="op">)</span></span></code></pre></div>
<p>Check out our <a href="https://github.com/green-striped-gecko/dartR/wiki/dartR-team-publications" class="external-link">articles</a> and our <a href="https://github.com/green-striped-gecko/dartR/wiki/dartR-awards" class="external-link">awards</a>.</p>
<p>Have fun working with <code>dartR</code>!</p>
<p>Cheers,</p>
<p>Bernd, Arthur, Luis, Carlo &amp; Olly</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=dartR" class="external-link">View on CRAN</a></li>
<li><a href="https://groups.google.com/g/dartr?pli=1" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing dartR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Bernd Gruber <br><small class="roles"> Author, maintainer </small>  </li>
<li>Arthur Georges <br><small class="roles"> Author </small>  </li>
<li>Jose L. Mijangos <br><small class="roles"> Author </small>  </li>
<li>Carlo Pacioni <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://cran.r-project.org/package=dartR" class="external-link"><img src="https://www.r-pkg.org/badges/version/dartR?color=blue"></a></li>
<li><a href="https://cran.r-project.org/web/checks/check_results_dartR.html" class="external-link"><img src="https://cranchecks.info/badges/summary/dartR" alt="CRAN checks"></a></li>
<li><a href="https://github.com/green-striped-gecko/dartR/actions" class="external-link"><img src="https://github.com/green-striped-gecko/dartR/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://github.com/green-striped-gecko/dartR/actions/workflows/R-CMD-check-beta.yaml" class="external-link"><img src="https://github.com/green-striped-gecko/dartR/actions/workflows/R-CMD-check-beta.yaml/badge.svg?branch=beta" alt="R-CMD-check-beta"></a></li>
<li><a href="https://cran.r-project.org/package=dartR" class="external-link"><img src="http://cranlogs.r-pkg.org/badges/last-week/dartR?color=orange"></a></li>
<li><a href="https://github.com/green-striped-gecko/dartR/actions/workflows/R-CMD-check-dev.yaml" class="external-link"><img src="https://github.com/green-striped-gecko/dartR/actions/workflows/R-CMD-check-dev.yaml/badge.svg?branch=dev" alt="R-CMD-check-dev"></a></li>
<li><a href="https://github.com/green-striped-gecko/dartR/actions/workflows/R-CMD-check-dev_Arthur.yaml" class="external-link"><img src="https://github.com/green-striped-gecko/dartR/actions/workflows/R-CMD-check-dev_Arthur.yaml/badge.svg?branch=dev_arthur" alt="R-CMD-check-dev_Arthur"></a></li>
<li><a href="https://github.com/green-striped-gecko/dartR/actions/workflows/R-CMD-check-dev_Bernd.yaml" class="external-link"><img src="https://github.com/green-striped-gecko/dartR/actions/workflows/R-CMD-check-dev_Bernd.yaml/badge.svg?branch=dev_bernd" alt="R-CMD-check-dev_Bernd"></a></li>
<li><a href="https://github.com/green-striped-gecko/dartR/actions/workflows/R-CMD-check-dev_Luis.yaml" class="external-link"><img src="https://github.com/green-striped-gecko/dartR/actions/workflows/R-CMD-check-dev_Luis.yaml/badge.svg?branch=dev_luis" alt="R-CMD-check-dev_Luis"></a></li>
<li><a href="https://github.com/green-striped-gecko/dartR/actions/workflows/R-CMD-check-dev_Carlo.yaml" class="external-link"><img src="https://github.com/green-striped-gecko/dartR/actions/workflows/R-CMD-check-dev_Carlo.yaml/badge.svg?branch=dev_carlo" alt="R-CMD-check-dev_Carlo"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
