<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Different causes may be responsible for lack of Hardy-Weinberg
proportions. This function helps diagnose potential problems."><title>Provides descriptive stats and plots to diagnose potential problems
  with Hardy-Weinberg proportions — gl.diagnostics.hwe • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Provides descriptive stats and plots to diagnose potential problems
  with Hardy-Weinberg proportions — gl.diagnostics.hwe"><meta property="og:description" content="Different causes may be responsible for lack of Hardy-Weinberg
proportions. This function helps diagnose potential problems."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Provides descriptive stats and plots to diagnose potential problems
  with Hardy-Weinberg proportions</h1>
      
      <div class="d-none name"><code>gl.diagnostics.hwe.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Different causes may be responsible for lack of Hardy-Weinberg
proportions. This function helps diagnose potential problems.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.diagnostics.hwe</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  alpha_val <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  bins <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  stdErr <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  colors_hist <span class="op">=</span> <span class="va">two_colors</span>,</span>
<span>  colors_barplot <span class="op">=</span> <span class="va">two_colors_contrast</span>,</span>
<span>  plot_theme <span class="op">=</span> <span class="fu"><a href="theme_dartR.html">theme_dartR</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  save2tmp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n.cores <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt>alpha_val</dt>
<dd><p>Level of significance for testing [default 0.05].</p></dd>


<dt>bins</dt>
<dd><p>Number of bins to display in histograms [default 20].</p></dd>


<dt>stdErr</dt>
<dd><p>Whether standard errors for Fis and Fst should be computed 
(default: TRUE)</p></dd>


<dt>colors_hist</dt>
<dd><p>List of two color names for the borders and fill of the
histogram [default two_colors].</p></dd>


<dt>colors_barplot</dt>
<dd><p>Vector with two color names for the observed and
expected number of significant HWE tests [default two_colors_contrast].</p></dd>


<dt>plot_theme</dt>
<dd><p>User specified theme [default theme_dartR()].</p></dd>


<dt>save2tmp</dt>
<dd><p>If TRUE, saves any ggplots and listings to the session
temporary directory (tempdir) [default FALSE].</p></dd>


<dt>n.cores</dt>
<dd><p>The number of cores to use. If "auto", it will 
use all but one available cores [default "auto"].</p></dd>


<dt>verbose</dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log ; 3, progress and results summary; 5, full report
[default NULL, unless specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list with the table with the summary of the HWE tests and (if 
stdErr=TRUE) a named vector with the StdErrFis and StdErrFst.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function initially runs <code><a href="gl.report.hwe.html">gl.report.hwe</a></code> and reports
the ternary plots. The remaining outputs follow the recommendations from
 Waples
(2015) paper and De Meeûs 2018. These include:</p><ol><li><p>A histogram
with the distribution of p-values of the HWE tests. The distribution should
be roughly uniform across equal-sized bins.</p></li>
<li><p>A bar plot with observed
and expected (null expectation) number of significant HWE tests for the same
locus in multiple populations (that is, the x-axis shows whether a locus
results significant in 1, 2, ..., n populations. The y axis is the count of
these occurrences. The zero value on x-axis shows the number of
non-significant tests). If HWE tests are significant by chance alone,
observed and expected number of HWE tests should have roughly a similar
distribution.</p></li>
<li><p>A scatter plot with a linear regression between Fst and Fis,
 averaged across subpopulations. De Meeûs 2018 suggests that in the case of
Null alleles, a strong positive relationship is expected (together with the
Fis standard error much larger than the Fst standard error, see below).
<b>Note</b>, this is not the scatter plot that Waples 2015 presents in his
paper. In the lower right corner of the plot, the Pearson correlation
coefficient is reported.</p></li>
<li><p>The Fis and Fst (averaged over loci and
subpopulations) standard errors are also printed on screen and reported in
the returned list (if <code>stdErr=TRUE</code>). These are computed with the 
Jackknife method over loci (See De Meeûs 2007 for details on how this is 
computed) and it may take some time for these computations to complete. 
De Meeûs 2018 suggests that under a global significant heterozygosity 
deficit:</p>
<p>- if the
correlation between Fis and Fst is strongly positive, and StdErrFis &gt;&gt;
StdErrFst, Null alleles are likely to be the cause.</p>
<p>- if the correlation
between Fis and Fst is ~0 or mildly positive, and StdErrFis &gt; StdErrFst,
Wahlund may be the cause.</p>
<p>- if the correlation between Fis and Fst is ~0, and
StdErrFis ~ StdErrFst, selfing or sib mating could to be the cause.</p>
<p>It is
important to realise that these statistics only suggest a pattern (pointers).
Their absence is not conclusive evidence of the absence of the problem, as 
their presence does not confirm the cause of the problem.</p></li>
<li><p>A table where the
number of observed and expected significant HWE tests are reported by each
population, indicating whether these are due to heterozygosity excess or
deficiency. These can be used to have a clue of potential problems (e.g.
deficiency might be due to a Wahlund effect, presence of null alleles or
non-random sampling; excess might be due to sex linkage or different
selection between sexes, demographic changes or small Ne. See Table 1 in
Wapples 2015). The last two columns of the table generated by this function
report chisquare values and their associated p-values. Chisquare is computed
following Fisher's procedure for a global test (Fisher 1970). This basically
tests whether there is at least one test that is truly significant in the
series of tests conducted (De Meeûs et al 2009).</p></li>
</ol></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    
<ul><li><p>de Meeûs, T., McCoy, K.D., Prugnolle, F.,
Chevillon, C., Durand, P., Hurtrez-Boussès, S., Renaud, F., 2007. Population
genetics and molecular epidemiology or how to “débusquer la bête”. Infection,
Genetics and Evolution 7, 308-332.</p></li>
<li><p>De Meeûs, T., Guégan, J.-F., Teriokhin, A.T., 2009. MultiTest V.1.2, 
a program to binomially combine independent tests and performance comparison 
with other related methods on
proportional data. BMC Bioinformatics 10, 443-443.</p></li>
<li><p>De Meeûs, T., 2018. Revisiting FIS, FST, Wahlund Effects, and Null 
Alleles. Journal of Heredity 109, 446-456.</p></li>
<li><p>Fisher, R., 1970.
Statistical methods for research workers Edinburgh: Oliver and Boyd.</p></li>
<li><p>Waples, R. S. (2015). Testing for Hardy–Weinberg proportions: have we lost
the plot?. Journal of heredity, 106(1), 1-19.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="gl.report.hwe.html">gl.report.hwe</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Carlo Pacioni -- Post to
  <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/green-striped-gecko/dartR.data" class="external-link">"dartR.data"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">gl.diagnostics.hwe</span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="gl.filter.allna.html">gl.filter.allna</a></span><span class="op">(</span><span class="va">platypus.gl</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>stdErr<span class="op">=</span><span class="cn">FALSE</span>, n.cores<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

