<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Alleles unique to the Y or W chromosome and monomorphic on the X chromosomes
will appear in the SNP dataset as genotypes that are heterozygotic in all
individuals of the heterogametic sex and homozygous in all individuals of the
homogametic sex. This function keeps or drops loci with alleles that behave
in this way, as putative sex specific SNP markers."><title>Filters loci that are sex linked — gl.filter.sexlinked • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Filters loci that are sex linked — gl.filter.sexlinked"><meta property="og:description" content="Alleles unique to the Y or W chromosome and monomorphic on the X chromosomes
will appear in the SNP dataset as genotypes that are heterozygotic in all
individuals of the heterogametic sex and homozygous in all individuals of the
homogametic sex. This function keeps or drops loci with alleles that behave
in this way, as putative sex specific SNP markers."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.6.7</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Filters loci that are sex linked</h1>
      
      <div class="d-none name"><code>gl.filter.sexlinked.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Alleles unique to the Y or W chromosome and monomorphic on the X chromosomes
will appear in the SNP dataset as genotypes that are heterozygotic in all
individuals of the heterogametic sex and homozygous in all individuals of the
homogametic sex. This function keeps or drops loci with alleles that behave
in this way, as putative sex specific SNP markers.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.filter.sexlinked</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  sex <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filter <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  read.depth <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  t.het <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  t.hom <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  t.pres <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  plot.out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_theme <span class="op">=</span> <span class="fu"><a href="theme_dartR.html">theme_dartR</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot_colors <span class="op">=</span> <span class="va">three_colors</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>Name of the genlight object containing the SNP or presence/absence
(SilicoDArT) data [required].</p></dd>


<dt>sex</dt>
<dd><p>Factor that defines the sex of individuals. See explanation in
details [default NULL].</p></dd>


<dt>filter</dt>
<dd><p>Either 'keep' to keep sex linked markers only or 'drop' to drop
sex linked markers [required].</p></dd>


<dt>read.depth</dt>
<dd><p>Additional filter option to keep only loci above a certain
read.depth. Default to 0, which means read.depth is not taken into account
[default 0].</p></dd>


<dt>t.het</dt>
<dd><p>Tolerance in the heterogametic sex, that is t.het=0.05 means
that 5% of the heterogametic sex can be homozygous and still be regarded as
 consistent with a sex specific marker [default 0.1].</p></dd>


<dt>t.hom</dt>
<dd><p>Tolerance in the homogametic sex, that is t.hom=0.05 means that
5% of the homogametic sex can be heterozygous and still be regarded as
consistent with a sex specific marker [default 0.1].</p></dd>


<dt>t.pres</dt>
<dd><p>Tolerance in presence, that is t.pres=0.05 means that a
silicodart marker can be present in either of the sexes and still be regarded
 as a sex-linked marker [default 0.1].</p></dd>


<dt>plot.out</dt>
<dd><p>Creates a plot that shows the heterozygosity of males and
females at each loci be regarded as consistent with a sex specific marker 
[default TRUE].</p></dd>


<dt>plot_theme</dt>
<dd><p>Theme for the plot. See Details for options
[default theme_dartR()].</p></dd>


<dt>plot_colors</dt>
<dd><p>List of three color names for the not sex-linked loci, for
the sex-linked loci and for the area in which sex-linked loci appear 
[default three_colors].</p></dd>


<dt>verbose</dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2, 
progress log; 3, progress and results summary; 5, full report 
[default NULL, unless specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>The filtered genlight object (filter = 'keep': sex linked loci,
filter='drop', everything except sex linked loci).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Sex of the individuals for which sex is known with certainty can be provided
via a factor (equal to the length of the number of individuals) or to be held
in the variable <code>x@other$ind.metrics$sex</code>.
Coding is: M for male, F for female, U or NA for unknown/missing.
The script abbreviates the entries here to the first character. So, coding of
'Female' and 'Male' works as well. Character are also converted to upper 
cases.</p>
<p>'<strong> Function's output </strong></p>
<p>This function creates also a plot that shows the heterozygosity of males and
females at each loci for SNP data or percentage of present/absent in the case 
of SilicoDArT data.</p>
<p>Examples of other themes that can be used can be consulted in</p><ul><li><p><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">https://ggplot2.tidyverse.org/reference/ggtheme.html</a> and</p></li>
<li><p><a href="https://yutannihilation.github.io/allYourFigureAreBelongToUs/ggthemes/" class="external-link">https://yutannihilation.github.io/allYourFigureAreBelongToUs/ggthemes/</a></p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other filter functions: 
<code><a href="gl.filter.allna.html">gl.filter.allna</a>()</code>,
<code><a href="gl.filter.callrate.html">gl.filter.callrate</a>()</code>,
<code><a href="gl.filter.heterozygosity.html">gl.filter.heterozygosity</a>()</code>,
<code><a href="gl.filter.hwe.html">gl.filter.hwe</a>()</code>,
<code><a href="gl.filter.ld.html">gl.filter.ld</a>()</code>,
<code><a href="gl.filter.locmetric.html">gl.filter.locmetric</a>()</code>,
<code><a href="gl.filter.maf.html">gl.filter.maf</a>()</code>,
<code><a href="gl.filter.monomorphs.html">gl.filter.monomorphs</a>()</code>,
<code><a href="gl.filter.overshoot.html">gl.filter.overshoot</a>()</code>,
<code><a href="gl.filter.parent.offspring.html">gl.filter.parent.offspring</a>()</code>,
<code><a href="gl.filter.pa.html">gl.filter.pa</a>()</code>,
<code><a href="gl.filter.rdepth.html">gl.filter.rdepth</a>()</code>,
<code><a href="gl.filter.reproducibility.html">gl.filter.reproducibility</a>()</code>,
<code><a href="gl.filter.secondaries.html">gl.filter.secondaries</a>()</code>,
<code><a href="gl.filter.taglength.html">gl.filter.taglength</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Arthur Georges, Bernd Gruber &amp; Floriaan Devloo-Delva (Post to
 <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>  <span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">gl.filter.sexlinked</span><span class="op">(</span><span class="va">testset.gl</span>, filter<span class="op">=</span><span class="st">'drop'</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.filter.sexlinked </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: genlight object contains monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Starting gl.filter.sexlinked </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of females: 114 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of males: 120 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sex ratio females:(males+females): 0.49 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  No sex linked markers consistent with female heterogamety (ZZ/ZW)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  No sex linked markers consistent with male heterogamety (XX/XY)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.filter.sexlinked-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  No sex-linked loci identified and filter option was 'drop', therefore the </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;"> genlight object is returned unchanged.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.filter.sexlinked </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">gl.filter.sexlinked</span><span class="op">(</span><span class="va">testset.gs</span>, filter<span class="op">=</span><span class="st">'drop'</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.filter.sexlinked </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with Presence/Absence (SilicoDArT) data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: genlight object contains monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Starting gl.filter.sexlinked </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  No sex linked markers consistent with female heterogamety (ZZ/ZW)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  No sex linked markers consistent with male heterogamety (XX/XY)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.filter.sexlinked-2.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  No sex-linked loci identified and filter option was 'drop', therefore the </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;"> genlight object is returned unchanged.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.filter.sexlinked </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

