<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Estimates the rate of false positives in a fixed difference analysis — gl.fdsim • dartR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Estimates the rate of false positives in a fixed difference analysis — gl.fdsim"><meta property="og:description" content="This function takes two populations and generates allele frequency profiles
for them. It then samples an allele frequency for each, at random, and
estimates a sampling distribution for those two allele frequencies. Drawing
two samples from those sampling distributions, it calculates whether or not
they represent a fixed difference. This is applied to all loci, and the
number of fixed differences so generated are counted, as an expectation. The
script distinguished between true fixed differences (with a tolerance of
delta), and false positives. The simulation is repeated a given number of
times (default=1000) to provide an expectation of the number of false
positives, given the observed allele frequency profiles and the sample sizes.
 The probability of the observed count of fixed differences is greater than
the expected number of false positives is calculated."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dartR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/dartRTutorials.html">dartR Tutorials</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Estimates the rate of false positives in a fixed difference analysis</h1>
    
    <div class="hidden name"><code>gl.fdsim.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function takes two populations and generates allele frequency profiles
for them. It then samples an allele frequency for each, at random, and
estimates a sampling distribution for those two allele frequencies. Drawing
two samples from those sampling distributions, it calculates whether or not
they represent a fixed difference. This is applied to all loci, and the
number of fixed differences so generated are counted, as an expectation. The
script distinguished between true fixed differences (with a tolerance of
delta), and false positives. The simulation is repeated a given number of
times (default=1000) to provide an expectation of the number of false
positives, given the observed allele frequency profiles and the sample sizes.
 The probability of the observed count of fixed differences is greater than
the expected number of false positives is calculated.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.fdsim</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">poppair</span>,</span>
<span>  obs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sympatric <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reps <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  delta <span class="op">=</span> <span class="fl">0.02</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>Name of the genlight containing the SNP genotypes [required].</p></dd>


<dt>poppair</dt>
<dd><p>Labels of two populations for comparison in the form
c(popA,popB) [required].</p></dd>


<dt>obs</dt>
<dd><p>Observed number of fixed differences between the two populations
[default NULL].</p></dd>


<dt>sympatric</dt>
<dd><p>If TRUE, the two populations are sympatric, if FALSE then
allopatric [default FALSE].</p></dd>


<dt>reps</dt>
<dd><p>Number of replications to undertake in the simulation
[default 1000].</p></dd>


<dt>delta</dt>
<dd><p>The threshold value for the minor allele frequency to regard the
difference between two populations to be fixed [default 0.02].</p></dd>


<dt>verbose</dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report [default 2].</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list containing the following square matrices
        [[1]] observed fixed differences;
        [[2]] mean expected number of false positives for each comparison;
        [[3]] standard deviation of the no. of false positives for each
        comparison;
        [[4]] probability the observed fixed differences arose by chance for
        each comparison.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Custodian: Arthur Georges (Post to
 <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a>)</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">fd</span> <span class="op">&lt;-</span> <span class="fu">gl.fdsim</span><span class="op">(</span><span class="va">testset.gl</span>,poppair<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'EmsubRopeMata'</span>,<span class="st">'EmmacBurnBara'</span><span class="op">)</span>,sympatric<span class="op">=</span><span class="cn">TRUE</span>,verbose<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.fdsim </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Populations EmsubRopeMata vs EmmacBurnBara [sympatric]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Sample sizes: 11 5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. of loci: 42 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating false positive rate with 1000 replications. Please be patient</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Threshold minor allele frequency for generating a false positive: 0.02 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Estimated mean count of false positives: 0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Estimated SD of the count of false positives: 0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Prob that observed count of 12 are false positives: 0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.fdsim </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

