<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This script calculates the minor allele frequency for each locus and updates
the locus metadata for FreqHomRef, FreqHomSnp, FreqHets and MAF (if it
exists). It then uses the updated metadata for MAF to filter loci."><title>Filters loci on the basis of minor allele frequency (MAF) in a 
genlight
 adegenet object — gl.filter.maf • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Filters loci on the basis of minor allele frequency (MAF) in a 
genlight
 adegenet object — gl.filter.maf"><meta property="og:description" content="This script calculates the minor allele frequency for each locus and updates
the locus metadata for FreqHomRef, FreqHomSnp, FreqHets and MAF (if it
exists). It then uses the updated metadata for MAF to filter loci."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.6.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Filters loci on the basis of minor allele frequency (MAF) in a 
genlight
 adegenet object</h1>
      
      <div class="d-none name"><code>gl.filter.maf.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This script calculates the minor allele frequency for each locus and updates
the locus metadata for FreqHomRef, FreqHomSnp, FreqHets and MAF (if it
exists). It then uses the updated metadata for MAF to filter loci.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.filter.maf</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  by.pop <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pop.limit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nPop</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  ind.limit <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  recalc <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plot.out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_theme <span class="op">=</span> <span class="fu"><a href="theme_dartR.html">theme_dartR</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot_colors_pop <span class="op">=</span> <span class="va">discrete_palette</span>,</span>
<span>  plot_colors_all <span class="op">=</span> <span class="va">two_colors</span>,</span>
<span>  bins <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  save2tmp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt>threshold</dt>
<dd><p>Threshold MAF -- loci with a MAF less than the threshold
will be removed. If a value &gt; 1 is provided it will be 
interpreted as MAC (i.e. the minimum number of times an allele needs to be 
observed) [default 0.01].</p></dd>


<dt>by.pop</dt>
<dd><p>Whether MAF should be calculated by population [default FALSE].</p></dd>


<dt>pop.limit</dt>
<dd><p>Minimum number of populations in which MAF should be less 
than the threshold for a locus to be filtered out. Only used if by.pop=TRUE. 
The default value is half of the populations [default ceiling(nPop(x)/2)].</p></dd>


<dt>ind.limit</dt>
<dd><p>Minimum number of individuals that a population should 
contain to calculate MAF. Only used if by.pop=TRUE [default 10].</p></dd>


<dt>recalc</dt>
<dd><p>Recalculate the locus metadata statistics if any individuals
are deleted in the filtering [default FALSE].</p></dd>


<dt>plot.out</dt>
<dd><p>Specify if histograms of call rate, before and after, are to
be produced [default TRUE].</p></dd>


<dt>plot_theme</dt>
<dd><p>User specified theme for the plot [default theme_dartR()].</p></dd>


<dt>plot_colors_pop</dt>
<dd><p>A color palette for population plots
[default discrete_palette].</p></dd>


<dt>plot_colors_all</dt>
<dd><p>List of two color names for the borders and fill of
the overall plot [default two_colors].</p></dd>


<dt>bins</dt>
<dd><p>Number of bins to display in histograms [default 25].</p></dd>


<dt>save2tmp</dt>
<dd><p>If TRUE, saves any ggplots and listings to the session
temporary directory (tempdir) [default FALSE].</p></dd>


<dt>verbose</dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
 [default 2, unless specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>The reduced genlight dataset</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Careful consideration needs to be given to the settings to be used for this 
fucntion. When the filter is applied globally (i.e. <code>by.pop=FALSE</code>) but 
the data include multiple population, there is the risk to remove markers 
because the allele frequencies is low (at global level) but the allele 
frequencies
for the same markers may be high within some of the populations (especially 
if 
the per-population sample size is small). Similarly, not always it is a 
sensible choice to run this function using <code>by.pop=TRUE</code> because allele 
that are rare in a population may be very common in other, but the (possible) 
allele frequencies will depend on the sample size within each population. 
Where the purpose of filtering for MAF is to remove possible spurious alleles 
(i.e. sequencing errors), it is perhaps better to filter based on the number 
of times an allele is observed (MAC, Minimum Allele Count), under the 
assumption that if an allele is observed &gt;MAC, it is fairly rare to be an 
error.
<b>From v2.1</b> The threshold can take values &gt; 1. In this case, these are 
interpreted as a threshold for MAC.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other filter functions: 
<code><a href="gl.filter.allna.html">gl.filter.allna</a>()</code>,
<code><a href="gl.filter.callrate.html">gl.filter.callrate</a>()</code>,
<code><a href="gl.filter.heterozygosity.html">gl.filter.heterozygosity</a>()</code>,
<code><a href="gl.filter.hwe.html">gl.filter.hwe</a>()</code>,
<code><a href="gl.filter.ld.html">gl.filter.ld</a>()</code>,
<code><a href="gl.filter.locmetric.html">gl.filter.locmetric</a>()</code>,
<code><a href="gl.filter.monomorphs.html">gl.filter.monomorphs</a>()</code>,
<code><a href="gl.filter.overshoot.html">gl.filter.overshoot</a>()</code>,
<code><a href="gl.filter.parent.offspring.html">gl.filter.parent.offspring</a>()</code>,
<code><a href="gl.filter.pa.html">gl.filter.pa</a>()</code>,
<code><a href="gl.filter.rdepth.html">gl.filter.rdepth</a>()</code>,
<code><a href="gl.filter.reproducibility.html">gl.filter.reproducibility</a>()</code>,
<code><a href="gl.filter.secondaries.html">gl.filter.secondaries</a>()</code>,
<code><a href="gl.filter.sexlinked.html">gl.filter.sexlinked</a>()</code>,
<code><a href="gl.filter.taglength.html">gl.filter.taglength</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Luis Mijangos -- Post to
 <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl.filter.monomorphs.html">gl.filter.monomorphs</a></span><span class="op">(</span><span class="va">testset.gl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.filter.monomorphs </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Identifying monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Removing monomorphic loci and loci with all missing </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">                       data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.filter.monomorphs </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">gl.filter.maf</span><span class="op">(</span><span class="va">result</span>, threshold<span class="op">=</span><span class="fl">0.05</span>, verbose<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.filter.maf </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Removing loci with MAF &lt; 0.05 over all the dataset </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">                and recalculating FreqHoms and FreqHets</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Summary of filtered dataset</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   MAF for loci &gt; 0.05 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Initial number of loci: 111 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Number of loci deleted: 89 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Final number of loci: 22 </span>
<span class="r-plt img"><img src="gl.filter.maf-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.filter.maf </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

