<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Filters loci that show significant departure from Hardy-Weinberg Equilibrium — gl.filter.hwe • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Filters loci that show significant departure from Hardy-Weinberg Equilibrium — gl.filter.hwe"><meta name="description" content="This function filters out loci showing significant departure from H-W
proportions based on observed frequencies of reference homozygotes,
 heterozygotes and alternate homozygotes.
Loci are filtered out if they show HWE departure either in any one population
(n.pop.threshold =1) or in at least X number of populations
(n.pop.threshold &amp;gt; 1)."><meta property="og:description" content="This function filters out loci showing significant departure from H-W
proportions based on observed frequencies of reference homozygotes,
 heterozygotes and alternate homozygotes.
Loci are filtered out if they show HWE departure either in any one population
(n.pop.threshold =1) or in at least X number of populations
(n.pop.threshold &amp;gt; 1)."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/green-striped-gecko/dartR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Filters loci that show significant departure from Hardy-Weinberg Equilibrium</h1>
      <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl.filter.hwe.r" class="external-link"><code>R/gl.filter.hwe.r</code></a></small>
      <div class="d-none name"><code>gl.filter.hwe.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function filters out loci showing significant departure from H-W
proportions based on observed frequencies of reference homozygotes,
 heterozygotes and alternate homozygotes.</p>
<p>Loci are filtered out if they show HWE departure either in any one population
(n.pop.threshold =1) or in at least X number of populations
(n.pop.threshold &gt; 1).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.filter.hwe</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  subset <span class="op">=</span> <span class="st">"each"</span>,</span>
<span>  n.pop.threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  method_sig <span class="op">=</span> <span class="st">"Exact"</span>,</span>
<span>  multi_comp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  multi_comp_method <span class="op">=</span> <span class="st">"BY"</span>,</span>
<span>  alpha_val <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  pvalue_type <span class="op">=</span> <span class="st">"midp"</span>,</span>
<span>  cc_val <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  min_sample_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt id="arg-subset">subset<a class="anchor" aria-label="anchor" href="#arg-subset"></a></dt>
<dd><p>Way to group individuals to perform H-W tests. Either a vector
with population names, 'each', 'all' (see details) [default 'each'].</p></dd>


<dt id="arg-n-pop-threshold">n.pop.threshold<a class="anchor" aria-label="anchor" href="#arg-n-pop-threshold"></a></dt>
<dd><p>The minimum number of populations where the same locus
has to be out of H-W proportions to be removed [default 1].</p></dd>


<dt id="arg-method-sig">method_sig<a class="anchor" aria-label="anchor" href="#arg-method-sig"></a></dt>
<dd><p>Method for determining statistical significance:
'ChiSquare'
or 'Exact' [default 'Exact'].</p></dd>


<dt id="arg-multi-comp">multi_comp<a class="anchor" aria-label="anchor" href="#arg-multi-comp"></a></dt>
<dd><p>Whether to adjust p-values for multiple comparisons
[default FALSE].</p></dd>


<dt id="arg-multi-comp-method">multi_comp_method<a class="anchor" aria-label="anchor" href="#arg-multi-comp-method"></a></dt>
<dd><p>Method to adjust p-values for multiple comparisons:
'holm', 'hochberg', 'hommel', 'bonferroni', 'BH', 'BY', 'fdr'
(see details) [default 'fdr'].</p></dd>


<dt id="arg-alpha-val">alpha_val<a class="anchor" aria-label="anchor" href="#arg-alpha-val"></a></dt>
<dd><p>Level of significance for testing [default 0.05].</p></dd>


<dt id="arg-pvalue-type">pvalue_type<a class="anchor" aria-label="anchor" href="#arg-pvalue-type"></a></dt>
<dd><p>Type of p-value to be used in the Exact method.
Either 'dost','selome','midp' (see details) [default 'midp'].</p></dd>


<dt id="arg-cc-val">cc_val<a class="anchor" aria-label="anchor" href="#arg-cc-val"></a></dt>
<dd><p>The continuity correction applied to the ChiSquare test
[default 0.5].</p></dd>


<dt id="arg-min-sample-size">min_sample_size<a class="anchor" aria-label="anchor" href="#arg-min-sample-size"></a></dt>
<dd><p>Minimum number of individuals per population in which
perform H-W tests [default 5].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2, unless specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A genlight object with the loci departing significantly from H-W
proportions removed.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>There are several factors that can cause deviations from Hardy-Weinberg
 proportions including: mutation, finite population size, selection,
 population structure, age structure, assortative mating, sex linkage,
 nonrandom sampling and genotyping errors. Therefore, testing for
 Hardy-Weinberg proportions should be a process that involves a careful
 evaluation of the results, a good place to start is Waples (2015).</p>
<p>Note that tests for H-W proportions are only valid if there is no population
 substructure (assuming random mating) and have sufficient power only when
 there is sufficient sample size (n individuals &gt; 15).</p>
<p>Populations can be defined in three ways:</p><ul><li><p>Merging all populations in the dataset using subset = 'all'.</p></li>
<li><p>Within each population separately using: subset = 'each'.</p></li>
<li><p>Within selected populations using for example: subset =
c('pop1','pop2').</p></li>
</ul><p>Two different statistical methods to test for deviations from Hardy Weinberg
proportions:</p><ul><li><p>The classical chi-square test (method_sig='ChiSquare') based on the
function <code><a href="https://rdrr.io/pkg/HardyWeinberg/man/HWChisq.html" class="external-link">HWChisq</a></code> of the R package HardyWeinberg.
By default a continuity correction is applied (cc_val=0.5). The
continuity correction can be turned off (by specifying cc_val=0), for example
in cases of extreme allele frequencies in which the continuity correction can
lead to excessive type 1 error rates.</p></li>
<li><p>The exact test (method_sig='Exact') based on the exact calculations
contained in the function <code><a href="https://rdrr.io/pkg/HardyWeinberg/man/HWExactStats.html" class="external-link">HWExactStats</a></code> of the R
package HardyWeinberg, and described in Wigginton  et al. (2005). The exact
test is recommended in most cases (Wigginton  et al., 2005).
Three different methods to estimate p-values (pvalue_type) in the Exact test
can be used:</p><ul><li><p>'dost' p-value is computed as twice the tail area of a one-sided test.</p></li>
<li><p>'selome' p-value is computed as the sum of the probabilities of all
samples less or equally likely as the current sample.</p></li>
<li><p>'midp', p-value is computed as half the probability of the current
sample + the probabilities of all samples that are more extreme.</p></li>
</ul><p>The standard exact p-value is overly conservative, in particular
for small minor allele frequencies. The mid p-value ameliorates this problem
by bringing the rejection rate closer to the nominal level, at the price of
occasionally exceeding the nominal level (Graffelman &amp; Moreno, 2013).</p></li>
</ul><p>Correction for multiple tests can be applied using the following methods
based on the function <code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></code>:</p><ul><li><p>'holm' is also known as the sequential Bonferroni technique
(Rice, 1989).
This method has a greater statistical power than the standard Bonferroni
test,
however this method becomes very stringent when many tests are performed and
many real deviations from the null hypothesis can go undetected
(Waples, 2015).</p></li>
<li><p>'hochberg' based on Hochberg, 1988.</p></li>
<li><p>'hommel' based on Hommel, 1988. This method is more powerful than
Hochberg's, but the difference is usually small.</p></li>
<li><p>'bonferroni' in which p-values are multiplied by the number of tests.
This method is very stringent and therefore has reduced power to detect
multiple departures from the null hypothesis.</p></li>
<li><p>'BH' based on Benjamini &amp; Hochberg, 1995.</p></li>
<li><p>'BY' based on Benjamini &amp; Yekutieli, 2001.</p></li>
</ul><p>The first four methods are designed to give strong control of the family-wise
error rate. The last two methods control the false discovery rate (FDR),
the expected proportion of false discoveries among the rejected hypotheses.
The false discovery rate is a less stringent condition than the family-wise
error rate, so these methods are more powerful than the others, especially
when number of tests is large.
The number of tests on which the adjustment for multiple comparisons is
the number of populations times the number of loci.</p>
<p><b>From v2.1</b> <code>gl.filter.hwe</code> takes the argument
 <code>n.pop.threshold</code>.
if <code>n.pop.threshold &gt; 1</code> loci will be removed only if they are
concurrently
significant (after adjustment if applied) out of hwe in &gt;=
<code>n.pop.threshold &gt; 1</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>

<ul><li><p>Benjamini, Y., and Yekutieli, D. (2001). The control of the false
 discovery rate in multiple testing under dependency. Annals of Statistics,
 29, 1165–1188.</p></li>
<li><p>Graffelman, J. (2015). Exploring Diallelic Genetic Markers: The Hardy
Weinberg Package. Journal of Statistical Software 64:1-23.</p></li>
<li><p>Graffelman, J. &amp; Morales-Camarena, J. (2008). Graphical tests for
Hardy-Weinberg equilibrium based on the ternary plot. Human Heredity
65:77-84.</p></li>
<li><p>Graffelman, J., &amp; Moreno, V. (2013). The mid p-value in exact tests for
Hardy-Weinberg equilibrium. Statistical applications in genetics and
 molecular
biology, 12(4), 433-448.</p></li>
<li><p>Hochberg, Y. (1988). A sharper Bonferroni procedure for multiple tests
 of significance. Biometrika, 75, 800–803.</p></li>
<li><p>Hommel, G. (1988). A stagewise rejective multiple test procedure based
 on a modified Bonferroni test. Biometrika, 75, 383–386.</p></li>
<li><p>Rice, W. R. (1989). Analyzing tables of statistical tests. Evolution,
 43(1), 223-225.</p></li>
<li><p>Waples, R. S. (2015). Testing for Hardy–Weinberg proportions: have we
lost the plot?. Journal of heredity, 106(1), 1-19.</p></li>
<li><p>Wigginton, J.E., Cutler, D.J., &amp; Abecasis, G.R. (2005). A Note on Exact
Tests of Hardy-Weinberg Equilibrium. American Journal of Human Genetics
76:887-893.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="gl.report.hwe.html">gl.report.hwe</a></code></p>
<p>Other filter functions:
<code><a href="gl.filter.allna.html">gl.filter.allna</a>()</code>,
<code><a href="gl.filter.callrate.html">gl.filter.callrate</a>()</code>,
<code><a href="gl.filter.heterozygosity.html">gl.filter.heterozygosity</a>()</code>,
<code><a href="gl.filter.ld.html">gl.filter.ld</a>()</code>,
<code><a href="gl.filter.locmetric.html">gl.filter.locmetric</a>()</code>,
<code><a href="gl.filter.maf.html">gl.filter.maf</a>()</code>,
<code><a href="gl.filter.monomorphs.html">gl.filter.monomorphs</a>()</code>,
<code><a href="gl.filter.overshoot.html">gl.filter.overshoot</a>()</code>,
<code><a href="gl.filter.pa.html">gl.filter.pa</a>()</code>,
<code><a href="gl.filter.parent.offspring.html">gl.filter.parent.offspring</a>()</code>,
<code><a href="gl.filter.rdepth.html">gl.filter.rdepth</a>()</code>,
<code><a href="gl.filter.reproducibility.html">gl.filter.reproducibility</a>()</code>,
<code><a href="gl.filter.secondaries.html">gl.filter.secondaries</a>()</code>,
<code><a href="gl.filter.sexlinked.html">gl.filter.sexlinked</a>()</code>,
<code><a href="gl.filter.taglength.html">gl.filter.taglength</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Luis Mijangos – Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">gl.filter.hwe</span><span class="op">(</span>x <span class="op">=</span> <span class="va">bandicoot.gl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.filter.hwe </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Analysing each population separately</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Loci examined: 1000 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Deleted 128 loci with significant departure from HWE at alpha = 0.05 applied locus by locus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Loci retained: 872 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">    Adjustment of p-values for multiple comparisons vary with </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">                sample size</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.filter.hwe </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

