<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculates an identity by descent matrix — gl.grm • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculates an identity by descent matrix — gl.grm"><meta name="description" content="This function calculates the mean probability of identity by state (IBS)
across loci that would result from all the possible crosses of the
individuals analyzed. IBD is calculated by an additive relationship matrix
approach developed by Endelman and Jannink (2012) as implemented in the
function A.mat (package rrBLUP)."><meta property="og:description" content="This function calculates the mean probability of identity by state (IBS)
across loci that would result from all the possible crosses of the
individuals analyzed. IBD is calculated by an additive relationship matrix
approach developed by Endelman and Jannink (2012) as implemented in the
function A.mat (package rrBLUP)."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/green-striped-gecko/dartR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculates an identity by descent matrix</h1>
      <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl.grm.r" class="external-link"><code>R/gl.grm.r</code></a></small>
      <div class="d-none name"><code>gl.grm.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calculates the mean probability of identity by state (IBS)
across loci that would result from all the possible crosses of the
individuals analyzed. IBD is calculated by an additive relationship matrix
approach developed by Endelman and Jannink (2012) as implemented in the
function <a href="https://rdrr.io/pkg/rrBLUP/man/A.mat.html" class="external-link">A.mat</a> (package rrBLUP).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.grm</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  plotheatmap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  palette_discrete <span class="op">=</span> <span class="va">discrete_palette</span>,</span>
<span>  palette_convergent <span class="op">=</span> <span class="va">convergent_palette</span>,</span>
<span>  legendx <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  legendy <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt id="arg-plotheatmap">plotheatmap<a class="anchor" aria-label="anchor" href="#arg-plotheatmap"></a></dt>
<dd><p>A switch if a heatmap should be shown [default TRUE].</p></dd>


<dt id="arg-palette-discrete">palette_discrete<a class="anchor" aria-label="anchor" href="#arg-palette-discrete"></a></dt>
<dd><p>A discrete palette for the color of populations or a
list with as many colors as there are populations in the dataset
 [default discrete_palette].</p></dd>


<dt id="arg-palette-convergent">palette_convergent<a class="anchor" aria-label="anchor" href="#arg-palette-convergent"></a></dt>
<dd><p>A convergent palette for the IBD values
[default convergent_palette].</p></dd>


<dt id="arg-legendx">legendx<a class="anchor" aria-label="anchor" href="#arg-legendx"></a></dt>
<dd><p>x coordinates for the legend[default 0].</p></dd>


<dt id="arg-legendy">legendy<a class="anchor" aria-label="anchor" href="#arg-legendy"></a></dt>
<dd><p>y coordinates for the legend[default 1].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log ; 3, progress and results summary; 5, full report
[default 2 or as specified using gl.set.verbosity].</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Parameters passed to function A.mat from package rrBLUP.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An identity by descent matrix</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Two or more alleles are identical by descent (IBD) if they are identical
copies of the same ancestral allele in a base population. The additive
relationship matrix is a theoretical framework for estimating a relationship
matrix that is consistent with an approach to estimate the probability that
the alleles at a random locus are identical in state (IBS).</p>
<p>This function also plots a heatmap, and a dendrogram, of IBD values where
each diagonal element has a mean that equals 1+f, where f is the inbreeding
coefficient (i.e. the probability that the two alleles at a randomly chosen
locus are IBD from the base population). As this probability lies between 0
 and 1, the diagonal elements range from 1 to 2. Because the inbreeding
 coefficients are expressed relative to the current population, the mean of
 the off-diagonal elements is -(1+f)/n, where n is the number of loci.
 Individual names are shown in the margins of the heatmap and colors
 represent different populations.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>

<ul><li><p>Endelman, J. B. (2011). Ridge regression and other kernels for genomic
 selection with r package rrblup. The Plant Genome 4, 250.</p></li>
<li><p>Endelman, J. B. , Jannink, J.-L. (2012). Shrinkage estimation of the
realized relationship matrix. G3: Genes, Genomics, Genetics 2, 1405.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="gl.grm.network.html">gl.grm.network</a></code></p>
<p>Other inbreeding functions:
<code><a href="gl.grm.network.html">gl.grm.network</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Arthur Georges – Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">gl.grm</span><span class="op">(</span><span class="va">platypus.gl</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.grm </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.grm-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.grm </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

