<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Visualize patterns of linkage disequilibrium and identification of haplotypes — gl.ld.haplotype • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Visualize patterns of linkage disequilibrium and identification of haplotypes — gl.ld.haplotype"><meta name="description" content="This function plots a Linkage disequilibrium (LD) heatmap, where the colour
shading indicates the strength of LD. Chromosome positions (Mbp) are shown on
 the horizontal axis, and haplotypes appear as triangles and delimited by
 dark yellow vertical lines. Numbers identifying each haplotype are shown in
 the upper part of the plot.
The heatmap also shows heterozygosity for each SNP.
The function identifies haplotypes based on contiguous SNPs that are in
 linkage disequilibrium using as threshold ld_threshold_haplo and
containing more than min_snps SNPs."><meta property="og:description" content="This function plots a Linkage disequilibrium (LD) heatmap, where the colour
shading indicates the strength of LD. Chromosome positions (Mbp) are shown on
 the horizontal axis, and haplotypes appear as triangles and delimited by
 dark yellow vertical lines. Numbers identifying each haplotype are shown in
 the upper part of the plot.
The heatmap also shows heterozygosity for each SNP.
The function identifies haplotypes based on contiguous SNPs that are in
 linkage disequilibrium using as threshold ld_threshold_haplo and
containing more than min_snps SNPs."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/green-striped-gecko/dartR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Visualize patterns of linkage disequilibrium and identification of haplotypes</h1>
      <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl.ld.haplotype.r" class="external-link"><code>R/gl.ld.haplotype.r</code></a></small>
      <div class="d-none name"><code>gl.ld.haplotype.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function plots a Linkage disequilibrium (LD) heatmap, where the colour
shading indicates the strength of LD. Chromosome positions (Mbp) are shown on
 the horizontal axis, and haplotypes appear as triangles and delimited by
 dark yellow vertical lines. Numbers identifying each haplotype are shown in
 the upper part of the plot.</p>
<p>The heatmap also shows heterozygosity for each SNP.</p>
<p>The function identifies haplotypes based on contiguous SNPs that are in
 linkage disequilibrium using as threshold <code>ld_threshold_haplo</code> and
containing more than <code>min_snps</code> SNPs.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.ld.haplotype</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  pop_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chrom_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ld_max_pairwise <span class="op">=</span> <span class="fl">1e+07</span>,</span>
<span>  maf <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  ld_stat <span class="op">=</span> <span class="st">"R.squared"</span>,</span>
<span>  ind.limit <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  min_snps <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  ld_threshold_haplo <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  coordinates <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  color_haplo <span class="op">=</span> <span class="st">"viridis"</span>,</span>
<span>  color_het <span class="op">=</span> <span class="st">"deeppink"</span>,</span>
<span>  plot.out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  save2tmp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt id="arg-pop-name">pop_name<a class="anchor" aria-label="anchor" href="#arg-pop-name"></a></dt>
<dd><p>Name of the population to analyse. If NULL all the
populations are analised [default NULL].</p></dd>


<dt id="arg-chrom-name">chrom_name<a class="anchor" aria-label="anchor" href="#arg-chrom-name"></a></dt>
<dd><p>Nme of the chromosome to analyse. If NULL all the
chromosomes are analised [default NULL].</p></dd>


<dt id="arg-ld-max-pairwise">ld_max_pairwise<a class="anchor" aria-label="anchor" href="#arg-ld-max-pairwise"></a></dt>
<dd><p>Maximum distance in number of base pairs at which LD
should be calculated [default 10000000].</p></dd>


<dt id="arg-maf">maf<a class="anchor" aria-label="anchor" href="#arg-maf"></a></dt>
<dd><p>Minor allele frequency (by population) threshold to filter out
loci. If a value &gt; 1 is provided it will be interpreted as MAC (i.e. the
 minimum number of times an allele needs to be observed) [default 0.05].</p></dd>


<dt id="arg-ld-stat">ld_stat<a class="anchor" aria-label="anchor" href="#arg-ld-stat"></a></dt>
<dd><p>The LD measure to be calculated: "LLR", "OR", "Q", "Covar",
"D.prime", "R.squared", and "R". See <code><a href="https://rdrr.io/pkg/snpStats/man/ld.html" class="external-link">ld</a></code>
 (package snpStats) for details [default "R.squared"].</p></dd>


<dt id="arg-ind-limit">ind.limit<a class="anchor" aria-label="anchor" href="#arg-ind-limit"></a></dt>
<dd><p>Minimum number of individuals that a population should
contain to take it in account to report loci in LD [default 10].</p></dd>


<dt id="arg-min-snps">min_snps<a class="anchor" aria-label="anchor" href="#arg-min-snps"></a></dt>
<dd><p>Minimum number of SNPs that should have a haplotype to call
it [default 10].</p></dd>


<dt id="arg-ld-threshold-haplo">ld_threshold_haplo<a class="anchor" aria-label="anchor" href="#arg-ld-threshold-haplo"></a></dt>
<dd><p>Minimum LD between adjacent SNPs to call a
haplotype [default 0.5].</p></dd>


<dt id="arg-coordinates">coordinates<a class="anchor" aria-label="anchor" href="#arg-coordinates"></a></dt>
<dd><p>A vector of two elements with the start and end
coordinates in base pairs to which restrict the
analysis e.g. c(1,1000000) [default NULL].</p></dd>


<dt id="arg-color-haplo">color_haplo<a class="anchor" aria-label="anchor" href="#arg-color-haplo"></a></dt>
<dd><p>Color palette for haplotype plot. See details
[default "viridis"].</p></dd>


<dt id="arg-color-het">color_het<a class="anchor" aria-label="anchor" href="#arg-color-het"></a></dt>
<dd><p>Color for heterozygosity [default "deeppink"].</p></dd>


<dt id="arg-plot-out">plot.out<a class="anchor" aria-label="anchor" href="#arg-plot-out"></a></dt>
<dd><p>Specify if heatmap plot is to be produced [default TRUE].</p></dd>


<dt id="arg-save-tmp">save2tmp<a class="anchor" aria-label="anchor" href="#arg-save-tmp"></a></dt>
<dd><p>If TRUE, saves any ggplots and listings to the session
temporary directory (tempdir) [default FALSE].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2, unless specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A table with the haplotypes that were identified.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The information for SNP's position should be stored in the genlight accessor
  "@position" and the SNP's chromosome name in the accessor "@chromosome"
  (see examples). The function will then calculate LD within each chromosome.</p>
<p>The output of the function includes a table with the haplotypes
 that were identified and their location.</p>
<p>Colors of the heatmap (<code>color_haplo</code>) are based on the function
   <code><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_fill_viridis</a></code> from  package <code>viridis</code>.
   Other color palettes options are "magma", "inferno", "plasma", "viridis",
    "cividis", "rocket", "mako" and "turbo".</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other ld functions:
<code><a href="gl.ld.distance.html">gl.ld.distance</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Luis Mijangos – Post to
 <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/green-striped-gecko/dartR.data" class="external-link">"dartR.data"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">platypus.gl</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl.filter.callrate.html">gl.filter.callrate</a></span><span class="op">(</span><span class="va">x</span>,threshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.filter.callrate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: Data may include monomorphic loci in call rate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">                    calculations for filtering</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Recalculating Call Rate</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Removing loci based on Call Rate, threshold = 1 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.ld.haplotype-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.filter.callrate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl.keep.pop.html">gl.keep.pop</a></span><span class="op">(</span><span class="va">x</span>, pop.list <span class="op">=</span> <span class="st">"TENTERFIELD"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.keep.pop </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for presence of nominated populations</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Retaining only populations TENTERFIELD </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: Resultant dataset may contain monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Locus metrics not recalculated</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.keep.pop </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">x</span><span class="op">$</span><span class="va">chromosome</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">other</span><span class="op">$</span><span class="va">loc.metrics</span><span class="op">$</span><span class="va">Chrom_Platypus_Chrom_NCBIv1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">x</span><span class="op">$</span><span class="va">position</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">$</span><span class="va">other</span><span class="op">$</span><span class="va">loc.metrics</span><span class="op">$</span><span class="va">ChromPos_Platypus_Chrom_NCBIv1</span></span></span>
<span class="r-in"><span><span class="va">ld_res</span> <span class="op">&lt;-</span> <span class="fu">gl.ld.haplotype</span><span class="op">(</span><span class="va">x</span>,chrom_name <span class="op">=</span> <span class="st">"NC_041728.1_chromosome_1"</span>,</span></span>
<span class="r-in"><span>                          ld_max_pairwise <span class="op">=</span> <span class="fl">10000000</span> <span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.ld.haplotype </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating pairwise LD in population TENTERFIELD </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Analysing chromosome NC_041728.1_chromosome_1 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  The maximum distance at which LD should be calculated </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">                 (ld_max_pairwise) is too short for chromosome NC_041728.1_chromosome_1 . Setting this distance to 33814744 bp</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>`fortify(&lt;SpatialPolygonsDataFrame&gt;)` was deprecated in ggplot2 3.4.4.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Please migrate to sf.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">dartR</span> package.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://groups.google.com/g/dartr?pli=1&gt;</span>.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Regions defined for each Polygons</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  No haplotypes were identified for chromosome NC_041728.1_chromosome_1 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.ld.haplotype-2.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] population       chromosome       haplotype        start           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5] end              start_ld_plot    end_ld_plot      midpoint        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] midpoint_ld_plot labels          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt;0 rows&gt; (or 0-length row.names)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.ld.haplotype </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span>     </span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

