<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Visualize patterns of linkage disequilibrium and identification of 
haplotypes — gl.ld.haplotype • dartR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Visualize patterns of linkage disequilibrium and identification of 
haplotypes — gl.ld.haplotype"><meta property="og:description" content="This function plots a Linkage disequilibrium (LD) heatmap, where the colour 
shading indicates the strength of LD. Chromosome positions (Mbp) are shown on
 the horizontal axis, and haplotypes appear as triangles and delimited by 
 dark yellow vertical lines. Numbers identifying each haplotype are shown in 
 the upper part of the plot. 
The heatmap also shows heterozygosity for each SNP. 
The function identifies haplotypes based on contiguous SNPs that are in 
 linkage disequilibrium using as threshold ld_threshold_haplo and
containing more than min_snps SNPs."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dartR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/dartRTutorials.html">dartR Tutorials</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Visualize patterns of linkage disequilibrium and identification of 
haplotypes</h1>
    
    <div class="hidden name"><code>gl.ld.haplotype.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function plots a Linkage disequilibrium (LD) heatmap, where the colour 
shading indicates the strength of LD. Chromosome positions (Mbp) are shown on
 the horizontal axis, and haplotypes appear as triangles and delimited by 
 dark yellow vertical lines. Numbers identifying each haplotype are shown in 
 the upper part of the plot.</p> 
<p>The heatmap also shows heterozygosity for each SNP.</p> 
<p>The function identifies haplotypes based on contiguous SNPs that are in 
 linkage disequilibrium using as threshold <code>ld_threshold_haplo</code> and
containing more than <code>min_snps</code> SNPs.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.ld.haplotype</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  pop_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chrom_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ld_max_pairwise <span class="op">=</span> <span class="fl">1e+07</span>,</span>
<span>  maf <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  ld_stat <span class="op">=</span> <span class="st">"R.squared"</span>,</span>
<span>  ind.limit <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  min_snps <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  ld_threshold_haplo <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  coordinates <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  color_haplo <span class="op">=</span> <span class="st">"viridis"</span>,</span>
<span>  color_het <span class="op">=</span> <span class="st">"deeppink"</span>,</span>
<span>  plot.out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  save2tmp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt>pop_name</dt>
<dd><p>Name of the population to analyse. If NULL all the 
populations are analised [default NULL].</p></dd>


<dt>chrom_name</dt>
<dd><p>Nme of the chromosome to analyse. If NULL all the 
chromosomes are analised [default NULL].</p></dd>


<dt>ld_max_pairwise</dt>
<dd><p>Maximum distance in number of base pairs at which LD
should be calculated [default 10000000].</p></dd>


<dt>maf</dt>
<dd><p>Minor allele frequency (by population) threshold to filter out 
loci. If a value &gt; 1 is provided it will be interpreted as MAC (i.e. the
 minimum number of times an allele needs to be observed) [default 0.05].</p></dd>


<dt>ld_stat</dt>
<dd><p>The LD measure to be calculated: "LLR", "OR", "Q", "Covar",
"D.prime", "R.squared", and "R". See <code><a href="https://rdrr.io/pkg/snpStats/man/ld.html" class="external-link">ld</a></code>
 (package snpStats) for details [default "R.squared"].</p></dd>


<dt>ind.limit</dt>
<dd><p>Minimum number of individuals that a population should
contain to take it in account to report loci in LD [default 10].</p></dd>


<dt>min_snps</dt>
<dd><p>Minimum number of SNPs that should have a haplotype to call 
it [default 10].</p></dd>


<dt>ld_threshold_haplo</dt>
<dd><p>Minimum LD between adjacent SNPs to call a 
haplotype [default 0.5].</p></dd>


<dt>coordinates</dt>
<dd><p>A vector of two elements with the start and end 
coordinates in base pairs to which restrict the 
analysis e.g. c(1,1000000) [default NULL].</p></dd>


<dt>color_haplo</dt>
<dd><p>Color palette for haplotype plot. See details
[default "viridis"].</p></dd>


<dt>color_het</dt>
<dd><p>Color for heterozygosity [default "deeppink"].</p></dd>


<dt>plot.out</dt>
<dd><p>Specify if heatmap plot is to be produced [default TRUE].</p></dd>


<dt>save2tmp</dt>
<dd><p>If TRUE, saves any ggplots and listings to the session
temporary directory (tempdir) [default FALSE].</p></dd>


<dt>verbose</dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2, unless specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A table with the haplotypes that were identified.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The information for SNP's position should be stored in the genlight accessor
  "@position" and the SNP's chromosome name in the accessor "@chromosome"
  (see examples). The function will then calculate LD within each chromosome.</p>   
<p>The output of the function includes a table with the haplotypes
 that were identified and their location.</p> 
<p>Colors of the heatmap (<code>color_haplo</code>) are based on the function
   <code><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_fill_viridis</a></code> from  package <code>viridis</code>. 
   Other color palettes options are "magma", "inferno", "plasma", "viridis",
    "cividis", "rocket", "mako" and "turbo".</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Other ld functions: 
<code><a href="gl.ld.distance.html">gl.ld.distance</a>()</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Custodian: Luis Mijangos -- Post to
 <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/green-striped-gecko/dartR.data" class="external-link">"dartR.data"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">platypus.gl</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl.filter.callrate.html">gl.filter.callrate</a></span><span class="op">(</span><span class="va">x</span>,threshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.filter.callrate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: Data may include monomorphic loci in call rate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">                    calculations for filtering</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Recalculating Call Rate</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Removing loci based on Call Rate, threshold = 1 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.ld.haplotype-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.filter.callrate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">x</span><span class="op">$</span><span class="va">chromosome</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">other</span><span class="op">$</span><span class="va">loc.metrics</span><span class="op">$</span><span class="va">Chrom_Platypus_Chrom_NCBIv1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">x</span><span class="op">$</span><span class="va">position</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">$</span><span class="va">other</span><span class="op">$</span><span class="va">loc.metrics</span><span class="op">$</span><span class="va">ChromPos_Platypus_Chrom_NCBIv1</span></span></span>
<span class="r-in"><span><span class="va">ld_res</span> <span class="op">&lt;-</span> <span class="fu">gl.ld.haplotype</span><span class="op">(</span><span class="va">x</span>,chrom_name <span class="op">=</span> <span class="st">"NC_041728.1_chromosome_1"</span>,</span></span>
<span class="r-in"><span>                       ld_max_pairwise <span class="op">=</span> <span class="fl">10000000</span> <span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.ld.haplotype </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating pairwise LD in population SEVERN_ABOVE </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Analysing chromosome NC_041728.1_chromosome_1 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  The maximum distance at which LD should be calculated </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">                 (ld_max_pairwise) is too short for chromosome NC_041728.1_chromosome_1 . Setting this distance to 41378719 bp</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Regions defined for each Polygons</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  No haplotypes were identified for chromosome NC_041728.1_chromosome_1 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating pairwise LD in population SEVERN_BELOW </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Analysing chromosome NC_041728.1_chromosome_1 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  The maximum distance at which LD should be calculated </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">                 (ld_max_pairwise) is too short for chromosome NC_041728.1_chromosome_1 . Setting this distance to 39408304 bp</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Regions defined for each Polygons</span>
<span class="r-plt img"><img src="gl.ld.haplotype-2.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  No haplotypes were identified for chromosome NC_041728.1_chromosome_1 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating pairwise LD in population TENTERFIELD </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Analysing chromosome NC_041728.1_chromosome_1 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  The maximum distance at which LD should be calculated </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">                 (ld_max_pairwise) is too short for chromosome NC_041728.1_chromosome_1 . Setting this distance to 33814744 bp</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Regions defined for each Polygons</span>
<span class="r-plt img"><img src="gl.ld.haplotype-3.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  No haplotypes were identified for chromosome NC_041728.1_chromosome_1 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.ld.haplotype-4.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] population       chromosome       haplotype        start           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5] end              start_ld_plot    end_ld_plot      midpoint        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] midpoint_ld_plot labels          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt;0 rows&gt; (or 0-length row.names)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.ld.haplotype </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

