<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plots linkage disequilibrium against distance by population disequilibrium patterns — gl.ld.distance • dartR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plots linkage disequilibrium against distance by population disequilibrium patterns — gl.ld.distance"><meta property="og:description" content="The function creates a plot showing
the pairwise LD measure against distance in number of base pairs pooled over
all the chromosomes and a red line representing the threshold (R.squared =
0.2) that is commonly used to imply that two loci are unlinked (Delourme et
al., 2013; Li et al., 2014)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dartR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.9.9.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/dartRTutorials.html">dartR Tutorials</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/green-striped-gecko/dartR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plots linkage disequilibrium against distance by population disequilibrium patterns</h1>
    <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl.ld.distance.r" class="external-link"><code>R/gl.ld.distance.r</code></a></small>
    <div class="hidden name"><code>gl.ld.distance.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The function creates a plot showing
the pairwise LD measure against distance in number of base pairs pooled over
all the chromosomes and a red line representing the threshold (R.squared =
0.2) that is commonly used to imply that two loci are unlinked (Delourme et
al., 2013; Li et al., 2014).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.ld.distance</span><span class="op">(</span></span>
<span>  <span class="va">ld_report</span>,</span>
<span>  ld_resolution <span class="op">=</span> <span class="fl">1e+05</span>,</span>
<span>  pop_colors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_theme <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot.out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  save2tmp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="st">" "</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-ld-report">ld_report<a class="anchor" aria-label="anchor" href="#arg-ld-report"></a></dt>
<dd><p>Output from function <code><a href="gl.report.ld.map.html">gl.report.ld.map</a></code>
[required].</p></dd>


<dt id="arg-ld-resolution">ld_resolution<a class="anchor" aria-label="anchor" href="#arg-ld-resolution"></a></dt>
<dd><p>Resolution at which LD should be reported in number of
base pairs [default NULL].</p></dd>


<dt id="arg-pop-colors">pop_colors<a class="anchor" aria-label="anchor" href="#arg-pop-colors"></a></dt>
<dd><p>A color palette for box plots by population or a list
 with as many colors as there are populations in the dataset
[default NULL].</p></dd>


<dt id="arg-plot-theme">plot_theme<a class="anchor" aria-label="anchor" href="#arg-plot-theme"></a></dt>
<dd><p>User specified theme [default NULL].</p></dd>


<dt id="arg-plot-out">plot.out<a class="anchor" aria-label="anchor" href="#arg-plot-out"></a></dt>
<dd><p>Specify if plot is to be produced [default TRUE].</p></dd>


<dt id="arg-save-tmp">save2tmp<a class="anchor" aria-label="anchor" href="#arg-save-tmp"></a></dt>
<dd><p>If TRUE, saves any ggplots and listings to the session
temporary directory (tempdir) [default FALSE].</p></dd>


<dt id="arg-plot-title">plot_title<a class="anchor" aria-label="anchor" href="#arg-plot-title"></a></dt>
<dd><p>Title of tyh plot [default " "].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2, unless specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A dataframe with information of LD against distance by population.</p>
    </div>
    <div id="references">
    <h2>References</h2>

<ul><li><p>Delourme, R., Falentin, C., Fomeju, B. F., Boillot, M., Lassalle, G.,
André, I., . . . Marty, A. (2013). High-density SNP-based genetic map
development and linkage disequilibrium assessment in Brassica napusL. BMC
genomics, 14(1), 120.</p></li>
<li><p>Li, X., Han, Y., Wei, Y., Acharya, A., Farmer, A. D., Ho, J., . . .
Brummer, E. C. (2014). Development of an alfalfa SNP array and its use to
evaluate patterns of population structure and linkage disequilibrium. PLoS
One, 9(1), e84329.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Other ld functions:
<code><a href="gl.ld.haplotype.html">gl.ld.haplotype</a>()</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Custodian: Luis Mijangos – Post to
 <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"snpStats"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"fields"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/green-striped-gecko/dartR.data" class="external-link">"dartR.data"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">platypus.gl</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl.filter.callrate.html">gl.filter.callrate</a></span><span class="op">(</span><span class="va">x</span>,threshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl.filter.monomorphs.html">gl.filter.monomorphs</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">x</span><span class="op">$</span><span class="va">position</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">$</span><span class="va">other</span><span class="op">$</span><span class="va">loc.metrics</span><span class="op">$</span><span class="va">ChromPos_Platypus_Chrom_NCBIv1</span></span></span>
<span class="r-in"><span><span class="va">x</span><span class="op">$</span><span class="va">chromosome</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">other</span><span class="op">$</span><span class="va">loc.metrics</span><span class="op">$</span><span class="va">Chrom_Platypus_Chrom_NCBIv1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ld_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl.report.ld.map.html">gl.report.ld.map</a></span><span class="op">(</span><span class="va">x</span>,ld_max_pairwise <span class="op">=</span> <span class="fl">10000000</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ld_res_2</span> <span class="op">&lt;-</span> <span class="fu">gl.ld.distance</span><span class="op">(</span><span class="va">ld_res</span>,ld_resolution<span class="op">=</span> <span class="fl">1000000</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.filter.callrate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: Data may include monomorphic loci in call rate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">                    calculations for filtering</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Recalculating Call Rate</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Removing loci based on Call Rate, threshold = 1 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.ld.distance-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.filter.callrate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.filter.monomorphs </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Identifying monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Removing monomorphic loci and loci with all missing </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">                       data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.filter.monomorphs </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.ld.map </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating pairwise LD in population SEVERN_ABOVE </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating pairwise LD in population SEVERN_BELOW </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating pairwise LD in population TENTERFIELD </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.ld.distance-2.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.ld.map </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.ld.distance </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.ld.distance-3.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           pop distance    ld_stat</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        &lt;fctr&gt;    &lt;num&gt;      &lt;num&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_ABOVE  1000001 0.10239933</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_ABOVE  2000001 0.11837328</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_ABOVE  3000001 0.06542061</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_ABOVE  4000001 0.08078629</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_ABOVE  5000001 0.04427707</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_ABOVE  6000001 0.04725646</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_ABOVE  7000001 0.09087990</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_ABOVE  8000001 0.07732017</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_ABOVE  9000001 0.03934305</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_ABOVE  9992140 0.07103110</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_BELOW  1000001 0.14760863</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_BELOW  2000001 0.14563835</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_BELOW  3000001 0.05268563</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_BELOW  4000001 0.08586598</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_BELOW  5000001 0.10992954</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_BELOW  6000001 0.04812035</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_BELOW  7000001 0.07374916</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_BELOW  8000001 0.10933380</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_BELOW  9000001 0.12975825</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  SEVERN_BELOW  9992140 0.03812468</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TENTERFIELD  1000001 0.03968855</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TENTERFIELD  2000001 0.08432615</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TENTERFIELD  3000001 0.06752635</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TENTERFIELD  4000001 0.03745991</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TENTERFIELD  5000001 0.03536964</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TENTERFIELD  6000001 0.10039732</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TENTERFIELD  7000001 0.02024599</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TENTERFIELD  8000001 0.01268421</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TENTERFIELD  9000001 0.04717859</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TENTERFIELD  9992140 0.08552747</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           pop distance    ld_stat</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.ld.distance </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

