<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Creates an input file for the program NewHybrids and runs it if NewHybrids is installed — gl.nhybrids • dartR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Creates an input file for the program NewHybrids and runs it if NewHybrids is installed — gl.nhybrids"><meta property="og:description" content="This function compares two sets of parental populations to identify loci that
exhibit a fixed difference, returns an genlight object with the reduced
data, and creates an input file for the program NewHybrids using the top 200
(or user-specified lower loc.limit) loci. In the absence of two identified
parental populations, the script will select a random set of up to 200 loci only
(method='random') or up to the first 200 loci ranked on information content
(method='AvgPIC').
A fixed difference occurs when a SNP allele is present in all individuals
of one population and absent in the other. There is provision for setting
a level of tolerance, e.g. threshold = 0.05 which considers alleles present
at greater than 95
a fixed difference. Only up to 200 loci are retained, because of limitations
of NewHybids.
If you specify a directory for the NewHybrids executable file, then the
script will create the input file from the SNP data then run NewHybrids. If
the directory is set to NULL, the execution will stop once the input file
(default='nhyb.txt') has been written to disk. Note: the executable option
will not work on a Mac; Mac users should generate the NewHybrids input file
and run this on their local installation of NewHybrids.
Refer to the New Hybrids manual for further information on the parameters to
set
– http://ib.berkeley.edu/labs/slatkin/eriq/software/new_hybs_doc1_1Beta3.pdf
It is important to stringently filter the data on RepAvg and CallRate if
using the random option. One might elect to repeat the analysis
(method='random') and combine the resultant posterior probabilities should
the maximum of 200 loci be considered insufficient.
The F1 individuals should be homozygous at all loci for which the parental
populations are fixed and different, assuming parental populations have been
specified. Sampling errors can result in this not being the case, especially
where the sample sizes for the parental populations are small. Alternatively,
the threshold for posterior probabilities used to determine assignment
(pprob) or the definition of a fixed difference (threshold) may be too lax.
To assess the error rate in the determination of assignment of F1
individuals, a plot of the frequency of homozygous reference, heterozygotes
and homozygous alternate (SNP) can be produced by setting plot=TRUE (the
default)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dartR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.9.9.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/dartRTutorials.html">dartR Tutorials</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/green-striped-gecko/dartR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Creates an input file for the program NewHybrids and runs it if NewHybrids is installed</h1>
    <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl.nhybrids.r" class="external-link"><code>R/gl.nhybrids.r</code></a></small>
    <div class="hidden name"><code>gl.nhybrids.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function compares two sets of parental populations to identify loci that
exhibit a fixed difference, returns an genlight object with the reduced
data, and creates an input file for the program NewHybrids using the top 200
(or user-specified lower loc.limit) loci. In the absence of two identified
parental populations, the script will select a random set of up to 200 loci only
(method='random') or up to the first 200 loci ranked on information content
(method='AvgPIC').</p>
<p>A fixed difference occurs when a SNP allele is present in all individuals
of one population and absent in the other. There is provision for setting
a level of tolerance, e.g. threshold = 0.05 which considers alleles present
at greater than 95<!-- % in one population and less than 5% in the other to be -->
a fixed difference. Only up to 200 loci are retained, because of limitations
of NewHybids.</p>
<p>If you specify a directory for the NewHybrids executable file, then the
script will create the input file from the SNP data then run NewHybrids. If
the directory is set to NULL, the execution will stop once the input file
(default='nhyb.txt') has been written to disk. Note: the executable option
will not work on a Mac; Mac users should generate the NewHybrids input file
and run this on their local installation of NewHybrids.</p>
<p>Refer to the New Hybrids manual for further information on the parameters to
set
– http://ib.berkeley.edu/labs/slatkin/eriq/software/new_hybs_doc1_1Beta3.pdf</p>
<p>It is important to stringently filter the data on RepAvg and CallRate if
using the random option. One might elect to repeat the analysis
(method='random') and combine the resultant posterior probabilities should
the maximum of 200 loci be considered insufficient.</p>
<p>The F1 individuals should be homozygous at all loci for which the parental
populations are fixed and different, assuming parental populations have been
specified. Sampling errors can result in this not being the case, especially
where the sample sizes for the parental populations are small. Alternatively,
the threshold for posterior probabilities used to determine assignment
(pprob) or the definition of a fixed difference (threshold) may be too lax.
To assess the error rate in the determination of assignment of F1
individuals, a plot of the frequency of homozygous reference, heterozygotes
and homozygous alternate (SNP) can be produced by setting plot=TRUE (the
default).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.nhybrids</span><span class="op">(</span></span>
<span>  <span class="va">gl</span>,</span>
<span>  outpath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  p0 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"random"</span>,</span>
<span>  loc.limit <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_theme <span class="op">=</span> <span class="fu"><a href="theme_dartR.html">theme_dartR</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot_colors <span class="op">=</span> <span class="va">two_colors</span>,</span>
<span>  pprob <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  nhyb.directory <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  BurnIn <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  sweeps <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  GtypFile <span class="op">=</span> <span class="st">"TwoGensGtypFreq.txt"</span>,</span>
<span>  AFPriorFile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  PiPrior <span class="op">=</span> <span class="st">"Jeffreys"</span>,</span>
<span>  ThetaPrior <span class="op">=</span> <span class="st">"Jeffreys"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-gl">gl<a class="anchor" aria-label="anchor" href="#arg-gl"></a></dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt id="arg-outpath">outpath<a class="anchor" aria-label="anchor" href="#arg-outpath"></a></dt>
<dd><p>Path where to save the output file [default tempdir()].</p></dd>


<dt id="arg-p-">p0<a class="anchor" aria-label="anchor" href="#arg-p-"></a></dt>
<dd><p>List of populations to be regarded as parental population 0
[default NULL].</p></dd>


<dt id="arg-p-">p1<a class="anchor" aria-label="anchor" href="#arg-p-"></a></dt>
<dd><p>List of populations to be regarded as parental population 1
[default NULL].</p></dd>


<dt id="arg-threshold">threshold<a class="anchor" aria-label="anchor" href="#arg-threshold"></a></dt>
<dd><p>Sets the level at which a gene frequency difference is
considered to be fixed [default 0].</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>Specifies the method (random or AvgPIC) to select 200 loci for
NewHybrids [default random].</p></dd>


<dt id="arg-loc-limit">loc.limit<a class="anchor" aria-label="anchor" href="#arg-loc-limit"></a></dt>
<dd><p>Specifies the number of loci to use in the analysis
[default 200]</p></dd>


<dt id="arg-plot">plot<a class="anchor" aria-label="anchor" href="#arg-plot"></a></dt>
<dd><p>If TRUE, a plot of the frequency of homozygous reference,
heterozygotes and homozygous alternate (SNP) is produced for the F1
individuals
[default TRUE, applies only if both parental populations are specified].</p></dd>


<dt id="arg-plot-theme">plot_theme<a class="anchor" aria-label="anchor" href="#arg-plot-theme"></a></dt>
<dd><p>User specified theme [default theme_dartR()].</p></dd>


<dt id="arg-plot-colors">plot_colors<a class="anchor" aria-label="anchor" href="#arg-plot-colors"></a></dt>
<dd><p>Vector with two color names for the borders and fill
[default two_colors].</p></dd>


<dt id="arg-pprob">pprob<a class="anchor" aria-label="anchor" href="#arg-pprob"></a></dt>
<dd><p>Threshold level for assignment to likelihood bins
[default 0.95, used only if plot=TRUE].</p></dd>


<dt id="arg-nhyb-directory">nhyb.directory<a class="anchor" aria-label="anchor" href="#arg-nhyb-directory"></a></dt>
<dd><p>Directory that holds the NewHybrids executable file
e.g. C:/NewHybsPC [default NULL].</p></dd>


<dt id="arg-burnin">BurnIn<a class="anchor" aria-label="anchor" href="#arg-burnin"></a></dt>
<dd><p>Number of sweeps to use in the burn in [default 10000].</p></dd>


<dt id="arg-sweeps">sweeps<a class="anchor" aria-label="anchor" href="#arg-sweeps"></a></dt>
<dd><p>Number  of  sweeps  to  use  in  computing  the  actual Monte
Carlo averages [default 10000].</p></dd>


<dt id="arg-gtypfile">GtypFile<a class="anchor" aria-label="anchor" href="#arg-gtypfile"></a></dt>
<dd><p>Name of a file containing the genotype frequency classes
[default TwoGensGtypFreq.txt].</p></dd>


<dt id="arg-afpriorfile">AFPriorFile<a class="anchor" aria-label="anchor" href="#arg-afpriorfile"></a></dt>
<dd><p>Name of the file containing prior allele frequency
information [default NULL].</p></dd>


<dt id="arg-piprior">PiPrior<a class="anchor" aria-label="anchor" href="#arg-piprior"></a></dt>
<dd><p>Jeffreys-like priors or Uniform priors for the parameter pi
[default Jeffreys].</p></dd>


<dt id="arg-thetaprior">ThetaPrior<a class="anchor" aria-label="anchor" href="#arg-thetaprior"></a></dt>
<dd><p>Jeffreys-like priors or Uniform priors for the parameter
theta [default Jeffreys].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2 or as specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>The reduced genlight object, if parentals are provided; output of
 NewHybrids is saved to the working directory.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Anderson, E.C. and Thompson, E.A.(2002). A model-based method for identifying
species hybrids using multilocus genetic data. Genetics. 160:1217-1229.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Custodian: Arthur Georges – Post to
 <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">gl.nhybrids</span><span class="op">(</span><span class="va">testset.gl</span>, </span></span>
<span class="r-in"><span>p0<span class="op">=</span><span class="cn">NULL</span>, p1<span class="op">=</span><span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>nhyb.directory<span class="op">=</span><span class="st">'D:/workspace/R/NewHybsPC'</span>, <span class="co"># Specify as necessary</span></span></span>
<span class="r-in"><span>outpath<span class="op">=</span><span class="st">"D:/workspace"</span>,  <span class="co"># Specify as necessary, usually getwd() [= workspace]</span></span></span>
<span class="r-in"><span>BurnIn<span class="op">=</span><span class="fl">100</span>,</span></span>
<span class="r-in"><span>sweeps<span class="op">=</span><span class="fl">100</span>,</span></span>
<span class="r-in"><span>verbose<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

