<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Reports private alleles (and fixed alleles) per pair of populations — gl.report.pa • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reports private alleles (and fixed alleles) per pair of populations — gl.report.pa"><meta name="description" content="This function reports private alleles in one population compared with a
second population, for all populations taken pairwise. It also reports a
count of fixed allelic differences and the mean absolute allele frequency
differences (AFD) between pairs of populations."><meta property="og:description" content="This function reports private alleles in one population compared with a
second population, for all populations taken pairwise. It also reports a
count of fixed allelic differences and the mean absolute allele frequency
differences (AFD) between pairs of populations."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/green-striped-gecko/dartR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Reports private alleles (and fixed alleles) per pair of populations</h1>
      <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl.report.pa.r" class="external-link"><code>R/gl.report.pa.r</code></a></small>
      <div class="d-none name"><code>gl.report.pa.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function reports private alleles in one population compared with a
second population, for all populations taken pairwise. It also reports a
count of fixed allelic differences and the mean absolute allele frequency
differences (AFD) between pairs of populations.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.report.pa</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  x2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"pairwise"</span>,</span>
<span>  loc_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plot.out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  font_plot <span class="op">=</span> <span class="fl">14</span>,</span>
<span>  map.interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  palette_discrete <span class="op">=</span> <span class="va">discrete_palette</span>,</span>
<span>  save2tmp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt id="arg-x-">x2<a class="anchor" aria-label="anchor" href="#arg-x-"></a></dt>
<dd><p>If two separate genlight objects are to be compared this can be
provided here, but they must have the same number of SNPs [default NULL].</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>Method to calculate private alleles: 'pairwise' comparison or
compare each population against the rest 'one2rest' [default 'pairwise'].</p></dd>


<dt id="arg-loc-names">loc_names<a class="anchor" aria-label="anchor" href="#arg-loc-names"></a></dt>
<dd><p>Whether names of loci with private alleles and fixed
differences should reported. If TRUE, loci names are reported using a list
 [default FALSE].</p></dd>


<dt id="arg-plot-out">plot.out<a class="anchor" aria-label="anchor" href="#arg-plot-out"></a></dt>
<dd><p>Specify if Sankey plot is to be produced [default TRUE].</p></dd>


<dt id="arg-font-plot">font_plot<a class="anchor" aria-label="anchor" href="#arg-font-plot"></a></dt>
<dd><p>Numeric font size in pixels for the node text labels
[default 14].</p></dd>


<dt id="arg-map-interactive">map.interactive<a class="anchor" aria-label="anchor" href="#arg-map-interactive"></a></dt>
<dd><p>Specify whether an interactive map showing private
alleles between populations is to be produced [default FALSE].</p></dd>


<dt id="arg-palette-discrete">palette_discrete<a class="anchor" aria-label="anchor" href="#arg-palette-discrete"></a></dt>
<dd><p>A discrete palette for the color of populations or a
list with as many colors as there are populations in the dataset
 [default discrete_palette].</p></dd>


<dt id="arg-save-tmp">save2tmp<a class="anchor" aria-label="anchor" href="#arg-save-tmp"></a></dt>
<dd><p>If TRUE, saves any ggplots and listings to the session
temporary directory (tempdir) [default FALSE].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent, fatal errors only; 1, flag function
begin and end; 2, progress log; 3, progress and results summary; 5, full
report [default 2 or as specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data.frame. Each row shows, for each pair of populations the number
 of individuals in each population, the number of loci with fixed differences
 (same for both populations) in pop1 (compared to pop2) and vice versa. Same
 for private alleles and finally the absolute mean allele frequency
 difference between loci (AFD). If loc_names = TRUE, loci names with private
  alleles and fixed differences are reported in a list in addition to the
  dataframe.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Note that the number of paired alleles between two populations is not a
symmetric dissimilarity measure.</p>
<p>If no x2 is provided, the function uses the pop(gl) hierarchy to determine
pairs of populations, otherwise it runs a single comparison between x and
x2.</p>
<p><strong>Hint:</strong> in case you want to run comparisons between individuals
(assuming individual names are unique), you can simply redefine your
population names with your individual names, as below:</p>
<p><code>pop(gl) &lt;- indNames(gl)</code></p>
<p><strong> Definition of fixed and private alleles </strong></p>
<p>The table below shows the possible cases of allele frequencies between
two populations (0 = homozygote for Allele 1, x = both Alleles are present,
1 = homozygote for Allele 2).</p>
<ul><li><p>p: cases where there is a private allele in pop1 compared to pop2 (but
not vice versa)</p></li>
<li><p>f: cases where there is a fixed allele in pop1 (and pop2, as those cases
are symmetric)</p></li>
</ul><table class="table table"><tr><td></td><td></td><td></td><td><em>pop1</em></td><td></td></tr><tr><td></td><td></td><td><strong>0</strong></td><td><strong>x</strong></td><td><strong>1</strong></td></tr><tr><td></td><td><strong>0</strong></td><td>-</td><td>p</td><td>p,f</td></tr><tr><td><em>pop2</em></td><td><strong>x</strong></td><td>-</td><td>-</td><td>-</td></tr><tr><td></td><td><strong>1</strong></td><td>p,f</td><td>p</td><td>-</td></tr></table><p>The absolute allele frequency difference (AFD) in this function is a simple
differentiation metric displaying intuitive properties which provides a
valuable alternative to FST. For details about its properties and how it is
calculated see Berner (2019).</p>
<p>The function also reports an estimation of the lower bound of the number of
 undetected private alleles using the Good-Turing frequency formula,
 originally developed for cryptography, which estimates in an ecological
 context the true frequencies of rare species in a single assemblage based on
  an incomplete sample of individuals. The approach is described in Chao et
  al. (2017). For this function, the equation 2c is used. This estimate is
  reported in the output table as Chao1 and Chao2.</p>
<p>In this function a Sankey Diagram is used to visualize patterns of private
alleles between populations. This diagram allows to display flows (private
alleles) between nodes (populations). Their links are represented with arcs
that have a width proportional to the importance of the flow (number of
private alleles).</p>
<p>if save2temp=TRUE, resultant plot(s) and the tabulation(s) are saved to the
session's temporary directory.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>

<ul><li><p>Berner, D. (2019). Allele frequency difference AFD – an intuitive
alternative to FST for quantifying genetic population differentiation. Genes,
 10(4), 308.</p></li>
<li><p>Chao, Anne, et al. "Deciphering the enigma of undetected species,
 phylogenetic, and functional diversity based on Good-Turing theory."
 Ecology 98.11 (2017): 2914-2929.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="gl.list.reports.html">gl.list.reports</a></code>,
 <code><a href="gl.print.reports.html">gl.print.reports</a></code></p>
<p>Other report functions:
<code><a href="gl.report.bases.html">gl.report.bases</a>()</code>,
<code><a href="gl.report.callrate.html">gl.report.callrate</a>()</code>,
<code><a href="gl.report.diversity.html">gl.report.diversity</a>()</code>,
<code><a href="gl.report.hamming.html">gl.report.hamming</a>()</code>,
<code><a href="gl.report.heterozygosity.html">gl.report.heterozygosity</a>()</code>,
<code><a href="gl.report.hwe.html">gl.report.hwe</a>()</code>,
<code><a href="gl.report.ld.map.html">gl.report.ld.map</a>()</code>,
<code><a href="gl.report.locmetric.html">gl.report.locmetric</a>()</code>,
<code><a href="gl.report.maf.html">gl.report.maf</a>()</code>,
<code><a href="gl.report.monomorphs.html">gl.report.monomorphs</a>()</code>,
<code><a href="gl.report.overshoot.html">gl.report.overshoot</a>()</code>,
<code><a href="gl.report.parent.offspring.html">gl.report.parent.offspring</a>()</code>,
<code><a href="gl.report.rdepth.html">gl.report.rdepth</a>()</code>,
<code><a href="gl.report.reproducibility.html">gl.report.reproducibility</a>()</code>,
<code><a href="gl.report.secondaries.html">gl.report.secondaries</a>()</code>,
<code><a href="gl.report.sexlinked.html">gl.report.sexlinked</a>()</code>,
<code><a href="gl.report.taglength.html">gl.report.taglength</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Bernd Gruber – Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">gl.report.pa</span><span class="op">(</span><span class="va">platypus.gl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.pa </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">Package tidyverse  needed for this function to work. Please install it.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

