<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Calculates the probabilities of agreement with H-W proportions based on 
observed
frequencies of reference homozygotes, heterozygotes and alternate 
homozygotes."><title>Reports departure from Hardy-Weinberg proportions — gl.report.hwe • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reports departure from Hardy-Weinberg proportions — gl.report.hwe"><meta property="og:description" content="Calculates the probabilities of agreement with H-W proportions based on 
observed
frequencies of reference homozygotes, heterozygotes and alternate 
homozygotes."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Reports departure from Hardy-Weinberg proportions</h1>
      
      <div class="d-none name"><code>gl.report.hwe.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Calculates the probabilities of agreement with H-W proportions based on 
observed
frequencies of reference homozygotes, heterozygotes and alternate 
homozygotes.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.report.hwe</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  subset <span class="op">=</span> <span class="st">"each"</span>,</span>
<span>  method_sig <span class="op">=</span> <span class="st">"Exact"</span>,</span>
<span>  multi_comp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  multi_comp_method <span class="op">=</span> <span class="st">"BY"</span>,</span>
<span>  alpha_val <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  pvalue_type <span class="op">=</span> <span class="st">"midp"</span>,</span>
<span>  cc_val <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  sig_only <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min_sample_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  plot.out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_colors <span class="op">=</span> <span class="va">two_colors_contrast</span>,</span>
<span>  max_plots <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  save2tmp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt>subset</dt>
<dd><p>Way to group individuals to perform H-W tests. Either a vector
with population names, 'each', 'all' (see details) [default 'each'].</p></dd>


<dt>method_sig</dt>
<dd><p>Method for determining statistical significance: 
'ChiSquare'
or 'Exact' [default 'Exact'].</p></dd>


<dt>multi_comp</dt>
<dd><p>Whether to adjust p-values for multiple comparisons
[default FALSE].</p></dd>


<dt>multi_comp_method</dt>
<dd><p>Method to adjust p-values for multiple comparisons:
'holm', 'hochberg', 'hommel', 'bonferroni', 'BH', 'BY', 'fdr'
(see details) [default 'fdr'].</p></dd>


<dt>alpha_val</dt>
<dd><p>Level of significance for testing [default 0.05].</p></dd>


<dt>pvalue_type</dt>
<dd><p>Type of p-value to be used in the Exact method.
Either 'dost','selome','midp' (see details) [default 'midp'].</p></dd>


<dt>cc_val</dt>
<dd><p>The continuity correction applied to the ChiSquare test
[default 0.5].</p></dd>


<dt>sig_only</dt>
<dd><p>Whether the returned table should include loci with a 
significant departure from Hardy-Weinberg proportions [default TRUE].</p></dd>


<dt>min_sample_size</dt>
<dd><p>Minimum number of individuals per population in which
perform H-W tests [default 5].</p></dd>


<dt>plot.out</dt>
<dd><p>If TRUE, will produce Ternary Plot(s) [default TRUE].</p></dd>


<dt>plot_colors</dt>
<dd><p>Vector with two color names for the significant and
not-significant loci [default two_colors_contrast].</p></dd>


<dt>max_plots</dt>
<dd><p>Maximum number of plots to print per page [default 4].</p></dd>


<dt>save2tmp</dt>
<dd><p>If TRUE, saves any ggplots and listings to the session
temporary directory (tempdir) [default FALSE].</p></dd>


<dt>verbose</dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log ; 3, progress and results summary; 5, full report
[default NULL, unless specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A dataframe containing loci, counts of reference SNP homozygotes,
heterozygotes and alternate SNP homozygotes; probability of departure from
H-W proportions, per locus significance with and without correction for
multiple comparisons and the number of population where the same locus is 
significantly out of HWE.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>There are several factors that can cause deviations from Hardy-Weinberg
 proportions including: mutation, finite population size, selection,
 population structure, age structure, assortative mating, sex linkage,
 nonrandom sampling and genotyping errors. Therefore, testing for
 Hardy-Weinberg proportions should be a process that involves a careful
 evaluation of the results, a good place to start is Waples (2015).</p>
<p>Note that tests for H-W proportions are only valid if there is no population
 substructure (assuming random mating) and have sufficient power only when
 there is sufficient sample size (n individuals &gt; 15).</p>
<p>Populations can be defined in three ways:</p><ul><li><p>Merging all populations in the dataset using subset = 'all'.</p></li>
<li><p>Within each population separately using: subset = 'each'.</p></li>
<li><p>Within selected populations using for example: subset =
 c('pop1','pop2').</p></li>
</ul><p>Two different statistical methods to test for deviations from Hardy Weinberg
proportions:</p><ul><li><p>The classical chi-square test (method_sig='ChiSquare') based on the
function <code><a href="https://rdrr.io/pkg/HardyWeinberg/man/HWChisq.html" class="external-link">HWChisq</a></code> of the R package HardyWeinberg.
By default a continuity correction is applied (cc_val=0.5). The
continuity correction can be turned off (by specifying cc_val=0), for example
in cases of extreme allele frequencies in which the continuity correction can
lead to excessive type 1 error rates.</p></li>
<li><p>The exact test (method_sig='Exact') based on the exact calculations
contained in the function <code><a href="https://rdrr.io/pkg/HardyWeinberg/man/HWExactStats.html" class="external-link">HWExactStats</a></code> of the R
package HardyWeinberg, and described in Wigginton  et al. (2005). The exact
test is recommended in most cases (Wigginton  et al., 2005).
Three different methods to estimate p-values (pvalue_type) in the Exact test
can be used:</p><ul><li><p>'dost' p-value is computed as twice the tail area of a one-sided test.</p></li>
<li><p>'selome' p-value is computed as the sum of the probabilities of all
samples less or equally likely as the current sample.</p></li>
<li><p>'midp', p-value is computed as half the probability of the current
sample + the probabilities of all samples that are more extreme.</p></li>
</ul><p>The standard exact p-value is overly conservative, in particular
for small minor allele frequencies. The mid p-value ameliorates this problem
by bringing the rejection rate closer to the nominal level, at the price of
occasionally exceeding the nominal level (Graffelman &amp; Moreno, 2013).</p></li>
</ul><p>Correction for multiple tests can be applied using the following methods
based on the function <code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></code>:</p><ul><li><p>'holm' is also known as the sequential Bonferroni technique (Rice,
1989). This method has a greater statistical power than the standard
Bonferroni test, however this method becomes very stringent when many tests
are performed and many real deviations from the null hypothesis can go
 undetected (Waples, 2015).</p></li>
<li><p>'hochberg' based on Hochberg, 1988.</p></li>
<li><p>'hommel' based on Hommel, 1988. This method is more powerful than
 Hochberg's, but the difference is usually small.</p></li>
<li><p>'bonferroni' in which p-values are multiplied by the number of tests.
This method is very stringent and therefore has reduced power to detect
multiple departures from the null hypothesis.</p></li>
<li><p>'BH' based on Benjamini &amp; Hochberg, 1995.</p></li>
<li><p>'BY' based on Benjamini &amp; Yekutieli, 2001.</p></li>
</ul><p>The first four methods are designed to give strong control of the family-wise
error rate. The last two methods control the false discovery rate (FDR),
the expected proportion of false discoveries among the rejected hypotheses.
The false discovery rate is a less stringent condition than the family-wise
error rate, so these methods are more powerful than the others, especially
when number of tests is large.
The number of tests on which the adjustment for multiple comparisons is
the number of populations times the number of loci.</p>
<p><strong>Ternary plots</strong></p>
<p>Ternary plots can be used to visualise patterns of H-W proportions (plot.out
= TRUE). P-values and the statistical (non)significance of a large number of
bi-allelic markers can be inferred from their position in a ternary plot.
See Graffelman &amp; Morales-Camarena (2008) for further details. Ternary plots
are based on the function  <code><a href="https://rdrr.io/pkg/HardyWeinberg/man/HWTernaryPlot.html" class="external-link">HWTernaryPlot</a></code> from
the package HardyWeinberg. Each vertex of the Ternary plot represents one of 
the three possible genotypes for SNP data: homozygous for the reference 
allele (AA), heterozygous (AB) and homozygous for the alternative allele
 (BB). Loci deviating significantly from Hardy-Weinberg proportions after 
 correction for multiple tests are shown in pink. The blue parabola 
 represents Hardy-Weinberg equilibrium, and the area between green lines 
 represents the acceptance region.</p>
<p>For these plots to work it is necessary to install the package ggtern.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    
<ul><li><p>Benjamini, Y., and Yekutieli, D. (2001). The control of the false
 discovery rate in multiple testing under dependency. Annals of Statistics,
 29, 1165–1188.</p></li>
<li><p>Graffelman, J. (2015). Exploring Diallelic Genetic Markers: The Hardy
Weinberg Package. Journal of Statistical Software 64:1-23.</p></li>
<li><p>Graffelman, J. &amp; Morales-Camarena, J. (2008). Graphical tests for
Hardy-Weinberg equilibrium based on the ternary plot. Human Heredity
65:77-84.</p></li>
<li><p>Graffelman, J., &amp; Moreno, V. (2013). The mid p-value in exact tests for
Hardy-Weinberg equilibrium. Statistical applications in genetics and
molecular biology, 12(4), 433-448.</p></li>
<li><p>Hochberg, Y. (1988). A sharper Bonferroni procedure for multiple tests
 of significance. Biometrika, 75, 800–803.</p></li>
<li><p>Hommel, G. (1988). A stagewise rejective multiple test procedure based
 on a modified Bonferroni test. Biometrika, 75, 383–386.</p></li>
<li><p>Rice, W. R. (1989). Analyzing tables of statistical tests. Evolution,
 43(1), 223-225.</p></li>
<li><p>Waples, R. S. (2015). Testing for Hardy–Weinberg proportions: have we
lost the plot?. Journal of heredity, 106(1), 1-19.</p></li>
<li><p>Wigginton, J.E., Cutler, D.J., &amp; Abecasis, G.R. (2005). A Note on Exact
Tests of Hardy-Weinberg Equilibrium. American Journal of Human Genetics
76:887-893.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="gl.filter.hwe.html">gl.filter.hwe</a></code></p>
<p>Other report functions: 
<code><a href="gl.report.bases.html">gl.report.bases</a>()</code>,
<code><a href="gl.report.callrate.html">gl.report.callrate</a>()</code>,
<code><a href="gl.report.diversity.html">gl.report.diversity</a>()</code>,
<code><a href="gl.report.hamming.html">gl.report.hamming</a>()</code>,
<code><a href="gl.report.heterozygosity.html">gl.report.heterozygosity</a>()</code>,
<code><a href="gl.report.ld.map.html">gl.report.ld.map</a>()</code>,
<code><a href="gl.report.locmetric.html">gl.report.locmetric</a>()</code>,
<code><a href="gl.report.maf.html">gl.report.maf</a>()</code>,
<code><a href="gl.report.monomorphs.html">gl.report.monomorphs</a>()</code>,
<code><a href="gl.report.overshoot.html">gl.report.overshoot</a>()</code>,
<code><a href="gl.report.parent.offspring.html">gl.report.parent.offspring</a>()</code>,
<code><a href="gl.report.pa.html">gl.report.pa</a>()</code>,
<code><a href="gl.report.rdepth.html">gl.report.rdepth</a>()</code>,
<code><a href="gl.report.reproducibility.html">gl.report.reproducibility</a>()</code>,
<code><a href="gl.report.secondaries.html">gl.report.secondaries</a>()</code>,
<code><a href="gl.report.sexlinked.html">gl.report.sexlinked</a>()</code>,
<code><a href="gl.report.taglength.html">gl.report.taglength</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Luis Mijangos -- Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="fu">gl.report.hwe</span><span class="op">(</span>x <span class="op">=</span> <span class="va">bandicoot.gl</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

