<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Reads FASTA files and converts them to genlight object — gl.read.fasta • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reads FASTA files and converts them to genlight object — gl.read.fasta"><meta name="description" content="The following IUPAC Ambiguity Codes are taken as heterozygotes:
M is heterozygote for	AC and CA
R is heterozygote	for AG and GA
W is heterozygote	for AT and TA
S is heterozygote	for CG and GC
Y is heterozygote	for CT and TC
K is heterozygote	for GT and TG


The following IUPAC Ambiguity Codes are taken as missing data:
V
H
D
B
N


The function can deal with missing data in individuals, e.g. when FASTA
 files have different number of individuals due to missing data.
The allele with the highest frequency is taken as the reference allele.
SNPs with more than two alleles are skipped."><meta property="og:description" content="The following IUPAC Ambiguity Codes are taken as heterozygotes:
M is heterozygote for	AC and CA
R is heterozygote	for AG and GA
W is heterozygote	for AT and TA
S is heterozygote	for CG and GC
Y is heterozygote	for CT and TC
K is heterozygote	for GT and TG


The following IUPAC Ambiguity Codes are taken as missing data:
V
H
D
B
N


The function can deal with missing data in individuals, e.g. when FASTA
 files have different number of individuals due to missing data.
The allele with the highest frequency is taken as the reference allele.
SNPs with more than two alleles are skipped."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/green-striped-gecko/dartR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Reads FASTA files and converts them to genlight object</h1>
      <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl.read.fasta.r" class="external-link"><code>R/gl.read.fasta.r</code></a></small>
      <div class="d-none name"><code>gl.read.fasta.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The following IUPAC Ambiguity Codes are taken as heterozygotes:</p><ul><li><p>M is heterozygote for	AC and CA</p></li>
<li><p>R is heterozygote	for AG and GA</p></li>
<li><p>W is heterozygote	for AT and TA</p></li>
<li><p>S is heterozygote	for CG and GC</p></li>
<li><p>Y is heterozygote	for CT and TC</p></li>
<li><p>K is heterozygote	for GT and TG</p></li>
</ul><p>The following IUPAC Ambiguity Codes are taken as missing data:</p><ul><li><p>V</p></li>
<li><p>H</p></li>
<li><p>D</p></li>
<li><p>B</p></li>
<li><p>N</p></li>
</ul><p>The function can deal with missing data in individuals, e.g. when FASTA
 files have different number of individuals due to missing data.</p>
<p>The allele with the highest frequency is taken as the reference allele.</p>
<p>SNPs with more than two alleles are skipped.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.read.fasta</span><span class="op">(</span><span class="va">fasta_files</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span>, n_cores <span class="op">=</span> <span class="cn">NULL</span>, verbose <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-fasta-files">fasta_files<a class="anchor" aria-label="anchor" href="#arg-fasta-files"></a></dt>
<dd><p>Fasta files to read [required].</p></dd>


<dt id="arg-parallel">parallel<a class="anchor" aria-label="anchor" href="#arg-parallel"></a></dt>
<dd><p>A logical indicating whether multiple cores -if available-
should be used for the computations (TRUE), or not (FALSE); requires the
 package parallel to be installed [default FALSE].</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>If parallel is TRUE, the number of cores to be used in the
computations; if NULL, then the maximum number of cores available on the
 computer is used [default NULL].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2, unless specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A genlight object.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Ambiguity characters are often used to code heterozygotes. However, using
 heterozygotes as ambiguity characters may bias many estimates. See more
  information in the link below:
<a href="https://evodify.com/heterozygotes-ambiguity-characters/" class="external-link">https://evodify.com/heterozygotes-ambiguity-characters/</a></p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Luis Mijangos – Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span> <span class="co"># Folder where the fasta files are located. </span></span></span>
<span class="r-in"><span> <span class="va">folder_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, package <span class="op">=</span><span class="st">'dartR'</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="co"># listing the FASTA files, including their path. Files have an extension</span></span></span>
<span class="r-in"><span> <span class="co"># that contains "fas".</span></span></span>
<span class="r-in"><span> <span class="va">file_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">folder_samples</span>, pattern <span class="op">=</span> <span class="st">"*.fas"</span>, </span></span>
<span class="r-in"><span>                          full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="co"># reading fasta files</span></span></span>
<span class="r-in"><span>  <span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">gl.read.fasta</span><span class="op">(</span><span class="va">file_names</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.read.fasta </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Reading locus_1.fas </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Merging files...</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.compliance.check </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking coding of SNPs</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    SNP data scored NA, 0, 1 or 2 confirmed</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking locus metrics and flags</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Recalculating locus metrics</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    No monomorphic loci detected</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for loci with all missing data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    No loci with all missing data detected</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking whether individual names are unique.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for individual metrics</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: Creating a slot for individual metrics</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for population assignments</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Population assignments not detected, individuals assigned</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">                    to a single population labelled 'pop1'</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Spelling of coordinates checked and changed if necessary to </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">            lat/lon</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.compliance.check </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  Genlight object does not have individual metrics. You need to add them 'manually' to the @other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.read.fasta </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

