<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Reads FASTA files and converts them to genlight object — gl.read.fasta • dartR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Reads FASTA files and converts them to genlight object — gl.read.fasta"><meta property="og:description" content="The following IUPAC Ambiguity Codes are taken as heterozygotes:
M is heterozygote for	AC and CA
R is heterozygote	for AG and GA
W is heterozygote	for AT and TA
S is heterozygote	for CG and GC
Y is heterozygote	for CT and TC
K is heterozygote	for GT and TG

 
The following IUPAC Ambiguity Codes are taken as missing data:
V
H
D
B
N

 
The function can deal with missing data in individuals, e.g. when FASTA 
 files have different number of individuals due to missing data. 
The allele with the highest frequency is taken as the reference allele. 
SNPs with more than two alleles are skipped."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dartR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/dartRTutorials.html">dartR Tutorials</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Reads FASTA files and converts them to genlight object</h1>
    
    <div class="hidden name"><code>gl.read.fasta.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The following IUPAC Ambiguity Codes are taken as heterozygotes:</p><ul><li><p>M is heterozygote for	AC and CA</p></li>
<li><p>R is heterozygote	for AG and GA</p></li>
<li><p>W is heterozygote	for AT and TA</p></li>
<li><p>S is heterozygote	for CG and GC</p></li>
<li><p>Y is heterozygote	for CT and TC</p></li>
<li><p>K is heterozygote	for GT and TG</p></li>
</ul><p>The following IUPAC Ambiguity Codes are taken as missing data:</p><ul><li><p>V</p></li>
<li><p>H</p></li>
<li><p>D</p></li>
<li><p>B</p></li>
<li><p>N</p></li>
</ul><p>The function can deal with missing data in individuals, e.g. when FASTA 
 files have different number of individuals due to missing data.</p> 
<p>The allele with the highest frequency is taken as the reference allele.</p> 
<p>SNPs with more than two alleles are skipped.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.read.fasta</span><span class="op">(</span><span class="va">fasta_files</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span>, n_cores <span class="op">=</span> <span class="cn">NULL</span>, verbose <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>fasta_files</dt>
<dd><p>Fasta files to read [required].</p></dd>


<dt>parallel</dt>
<dd><p>A logical indicating whether multiple cores -if available-
should be used for the computations (TRUE), or not (FALSE); requires the
 package parallel to be installed [default FALSE].</p></dd>


<dt>n_cores</dt>
<dd><p>If parallel is TRUE, the number of cores to be used in the
computations; if NULL, then the maximum number of cores available on the
 computer is used [default NULL].</p></dd>


<dt>verbose</dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2, unless specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A genlight object.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Ambiguity characters are often used to code heterozygotes. However, using
 heterozygotes as ambiguity characters may bias many estimates. See more
  information in the link below:
<a href="https://evodify.com/heterozygotes-ambiguity-characters/" class="external-link">https://evodify.com/heterozygotes-ambiguity-characters/</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Custodian: Luis Mijangos -- Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span> <span class="co"># Folder where the fasta files are located. </span></span></span>
<span class="r-in"><span> <span class="va">folder_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, package <span class="op">=</span><span class="st">'dartR'</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="co"># listing the FASTA files, including their path. Files have an extension</span></span></span>
<span class="r-in"><span> <span class="co"># that contains "fas".</span></span></span>
<span class="r-in"><span> <span class="va">file_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">folder_samples</span>, pattern <span class="op">=</span> <span class="st">"*.fas"</span>, </span></span>
<span class="r-in"><span>                          full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="co"># reading fasta files</span></span></span>
<span class="r-in"><span>  <span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">gl.read.fasta</span><span class="op">(</span><span class="va">file_names</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.read.fasta </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Reading locus_1.fas </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Merging files...</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.compliance.check </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking coding of SNPs</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    SNP data scored NA, 0, 1 or 2 confirmed</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking locus metrics and flags</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Recalculating locus metrics</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    No monomorphic loci detected</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for loci with all missing data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    No loci with all missing data detected</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking whether individual names are unique.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for individual metrics</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: Creating a slot for individual metrics</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for population assignments</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Population assignments not detected, individuals assigned</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">                    to a single population labelled 'pop1'</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Spelling of coordinates checked and changed if necessary to </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">            lat/lon</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.compliance.check </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  Genlight object does not have individual metrics. You need to add them 'manually' to the @other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.read.fasta </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

