<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Identifies loci that are sex linked — gl.report.sexlinked • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Identifies loci that are sex linked — gl.report.sexlinked"><meta name="description" content="Alleles unique to the Y or W chromosome and monomorphic on the X chromosomes
will appear in the SNP dataset as genotypes that are heterozygotic in all
individuals of the heterogametic sex and homozygous in all individuals of the
homogametic sex. This function identifies loci with alleles that behave in
this way, as putative sex specific SNP markers."><meta property="og:description" content="Alleles unique to the Y or W chromosome and monomorphic on the X chromosomes
will appear in the SNP dataset as genotypes that are heterozygotic in all
individuals of the heterogametic sex and homozygous in all individuals of the
homogametic sex. This function identifies loci with alleles that behave in
this way, as putative sex specific SNP markers."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/green-striped-gecko/dartR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Identifies loci that are sex linked</h1>
      <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl.report.sexlinked.r" class="external-link"><code>R/gl.report.sexlinked.r</code></a></small>
      <div class="d-none name"><code>gl.report.sexlinked.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Alleles unique to the Y or W chromosome and monomorphic on the X chromosomes
will appear in the SNP dataset as genotypes that are heterozygotic in all
individuals of the heterogametic sex and homozygous in all individuals of the
homogametic sex. This function identifies loci with alleles that behave in
this way, as putative sex specific SNP markers.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.report.sexlinked</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  sex <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  t.het <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  t.hom <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  t.pres <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  plot.out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_theme <span class="op">=</span> <span class="fu"><a href="theme_dartR.html">theme_dartR</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot_colors <span class="op">=</span> <span class="va">three_colors</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Name of the genlight object containing the SNP or presence/absence
(SilicoDArT) data [required].</p></dd>


<dt id="arg-sex">sex<a class="anchor" aria-label="anchor" href="#arg-sex"></a></dt>
<dd><p>Factor that defines the sex of individuals. See explanation in
details [default NULL].</p></dd>


<dt id="arg-t-het">t.het<a class="anchor" aria-label="anchor" href="#arg-t-het"></a></dt>
<dd><p>Tolerance in the heterogametic sex, that is t.het=0.05 means
that 5% of the heterogametic sex can be homozygous and still be regarded as
consistent with a sex specific marker [default 0.1].</p></dd>


<dt id="arg-t-hom">t.hom<a class="anchor" aria-label="anchor" href="#arg-t-hom"></a></dt>
<dd><p>Tolerance in the homogametic sex, that is t.hom=0.05 means that
5% of the homogametic sex can be heterozygous and still be regarded as
consistent with a sex specific marker [default 0.1].</p></dd>


<dt id="arg-t-pres">t.pres<a class="anchor" aria-label="anchor" href="#arg-t-pres"></a></dt>
<dd><p>Tolerance in presence, that is t.pres=0.05 means that a
silicodart marker can be present in either of the sexes and still be regarded
as a sex-linked marker [default 0.1].</p></dd>


<dt id="arg-plot-out">plot.out<a class="anchor" aria-label="anchor" href="#arg-plot-out"></a></dt>
<dd><p>Creates a plot that shows the heterozygosity of males and
females at each loci and shaded area in which loci can be regarded as
consistent with a sex specific marker [default TRUE].</p></dd>


<dt id="arg-plot-theme">plot_theme<a class="anchor" aria-label="anchor" href="#arg-plot-theme"></a></dt>
<dd><p>Theme for the plot. See Details for options
[default theme_dartR()].</p></dd>


<dt id="arg-plot-colors">plot_colors<a class="anchor" aria-label="anchor" href="#arg-plot-colors"></a></dt>
<dd><p>List of three color names for the not sex-linked loci, for
the sex-linked loci and for the area in which sex-linked loci appear
[default three_colors].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default NULL, unless specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Two lists of sex-linked loci, one for XX/XY and one for ZZ/ZW systems
and a plot.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Sex of the individuals for which sex is known with certainty can be provided
via a factor (equal to the length of the number of individuals) or to be held
in the variable <code>x@other$ind.metrics$sex</code>.
Coding is: M for male, F for female, U or NA for unknown/missing.
The script abbreviates the entries here to the first character. So, coding of
'Female' and 'Male' works as well. Character are also converted to upper
cases.</p>
<p>'<strong> Function's output </strong></p>
<p>This function creates a plot that shows the heterozygosity of males and
females at each loci or SNP data or percentage of present/absent in the case
of SilicoDArT data.</p>
<p>Examples of other themes that can be used can be consulted in</p><ul><li><p><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">https://ggplot2.tidyverse.org/reference/ggtheme.html</a> and</p></li>
<li><p><a href="https://yutannihilation.github.io/allYourFigureAreBelongToUs/ggthemes/" class="external-link">https://yutannihilation.github.io/allYourFigureAreBelongToUs/ggthemes/</a></p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other report functions:
<code><a href="gl.report.bases.html">gl.report.bases</a>()</code>,
<code><a href="gl.report.callrate.html">gl.report.callrate</a>()</code>,
<code><a href="gl.report.diversity.html">gl.report.diversity</a>()</code>,
<code><a href="gl.report.hamming.html">gl.report.hamming</a>()</code>,
<code><a href="gl.report.heterozygosity.html">gl.report.heterozygosity</a>()</code>,
<code><a href="gl.report.hwe.html">gl.report.hwe</a>()</code>,
<code><a href="gl.report.ld.map.html">gl.report.ld.map</a>()</code>,
<code><a href="gl.report.locmetric.html">gl.report.locmetric</a>()</code>,
<code><a href="gl.report.maf.html">gl.report.maf</a>()</code>,
<code><a href="gl.report.monomorphs.html">gl.report.monomorphs</a>()</code>,
<code><a href="gl.report.overshoot.html">gl.report.overshoot</a>()</code>,
<code><a href="gl.report.pa.html">gl.report.pa</a>()</code>,
<code><a href="gl.report.parent.offspring.html">gl.report.parent.offspring</a>()</code>,
<code><a href="gl.report.rdepth.html">gl.report.rdepth</a>()</code>,
<code><a href="gl.report.reproducibility.html">gl.report.reproducibility</a>()</code>,
<code><a href="gl.report.secondaries.html">gl.report.secondaries</a>()</code>,
<code><a href="gl.report.taglength.html">gl.report.taglength</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Arthur Georges, Bernd Gruber &amp; Floriaan Devloo-Delva
(Post to <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span> <span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">gl.report.sexlinked</span><span class="op">(</span><span class="va">testset.gl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.sexlinked </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: genlight object contains monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Starting gl.report.sexlinked </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of females: 114 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of males: 120 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sex ratio females:(males+females): 0.49 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  No sex linked markers consistent with female </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">                    heterogamety (ZZ/ZW)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  No sex linked markers consistent with male heterogamety </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">                    (XX/XY)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.sexlinked-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.sexlinked </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">gl.report.sexlinked</span><span class="op">(</span><span class="va">testset.gs</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.sexlinked </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with Presence/Absence (SilicoDArT) data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: genlight object contains monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Starting gl.report.sexlinked </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  No sex linked markers consistent with female </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">                    heterogamety (ZZ/ZW)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  No sex linked markers consistent with male </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">                        heterogamety (XX/XY)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.sexlinked-2.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.sexlinked </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl.filter.callrate.html">gl.filter.callrate</a></span><span class="op">(</span><span class="va">platypus.gl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.filter.callrate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: Data may include monomorphic loci in call rate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">                    calculations for filtering</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Recalculating Call Rate</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Removing loci based on Call Rate, threshold = 0.95 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.sexlinked-3.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.filter.callrate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl.filter.monomorphs.html">gl.filter.monomorphs</a></span><span class="op">(</span><span class="va">test</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.filter.monomorphs </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Identifying monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Removing monomorphic loci and loci with all missing </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">                       data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.filter.monomorphs </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">gl.report.sexlinked</span><span class="op">(</span><span class="va">test</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.sexlinked </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Starting gl.report.sexlinked </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of females: 40 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of males: 40 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sex ratio females:(males+females): 0.5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  No sex linked markers consistent with female </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">                    heterogamety (ZZ/ZW)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">  No sex linked markers consistent with male heterogamety </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">                    (XX/XY)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.sexlinked-4.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.sexlinked </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

