<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Reports various statistics of genetic differentiation between populations with confident intervals — gl.report.fstat • dartR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Reports various statistics of genetic differentiation between populations with confident intervals — gl.report.fstat"><meta property="og:description" content='This function calculates four genetic differentiation between populations
statistics (see the "Details" section for further information).

Fst - Measure of the degree of genetic differentiation of
subpopulations (Nei, 1987).
Fstp - Unbiased (i.e. corrected for sampling error, see
explanation below) Fst (Nei, 1987).
Dest - Jost’s D (Jost, 2008).
Gst_H - Gst standardized by the maximum level that it can obtain for
the observed amount of genetic variation (Hedrick 2005).


Sampling errors arise because allele frequencies in our samples differ from
those in the subpopulations from which they were taken (Holsinger, 2012).
Confident Intervals are obtained using bootstrapping.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dartR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.9.9.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/dartRTutorials.html">dartR Tutorials</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/green-striped-gecko/dartR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Reports various statistics of genetic differentiation between populations with confident intervals</h1>
    <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl.report.fstat.r" class="external-link"><code>R/gl.report.fstat.r</code></a></small>
    <div class="hidden name"><code>gl.report.fstat.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function calculates four genetic differentiation between populations
statistics (see the "Details" section for further information).</p>
<ul><li><p><strong>Fst</strong> - Measure of the degree of genetic differentiation of
subpopulations (Nei, 1987).</p></li>
<li><p><strong>Fstp</strong> - Unbiased (i.e. corrected for sampling error, see
explanation below) Fst (Nei, 1987).</p></li>
<li><p><strong>Dest</strong> - Jost’s D (Jost, 2008).</p></li>
<li><p><strong>Gst_H</strong> - Gst standardized by the maximum level that it can obtain for
the observed amount of genetic variation (Hedrick 2005).</p></li>
</ul><p>Sampling errors arise because allele frequencies in our samples differ from
those in the subpopulations from which they were taken (Holsinger, 2012).</p>
<p>Confident Intervals are obtained using bootstrapping.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.report.fstat</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  nboots <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  conf <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  CI.type <span class="op">=</span> <span class="st">"bca"</span>,</span>
<span>  ncpus <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  plot.stat <span class="op">=</span> <span class="st">"Fstp"</span>,</span>
<span>  plot.display <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  palette.divergent <span class="op">=</span> <span class="fu"><a href="gl.colors.html">gl.colors</a></span><span class="op">(</span><span class="st">"div"</span><span class="op">)</span>,</span>
<span>  font.size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  plot.dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot.file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt id="arg-nboots">nboots<a class="anchor" aria-label="anchor" href="#arg-nboots"></a></dt>
<dd><p>Number of bootstrap replicates to obtain confident intervals
[default 0].</p></dd>


<dt id="arg-conf">conf<a class="anchor" aria-label="anchor" href="#arg-conf"></a></dt>
<dd><p>The confidence level of the required interval  [default 0.95].</p></dd>


<dt id="arg-ci-type">CI.type<a class="anchor" aria-label="anchor" href="#arg-ci-type"></a></dt>
<dd><p>Method to estimate confident intervals. One of
"norm", "basic", "perc" or "bca" [default "bca"].</p></dd>


<dt id="arg-ncpus">ncpus<a class="anchor" aria-label="anchor" href="#arg-ncpus"></a></dt>
<dd><p>Number of processes to be used in parallel operation. If ncpus
&gt; 1 parallel operation is activated,see "Details" section [default 1].</p></dd>


<dt id="arg-plot-stat">plot.stat<a class="anchor" aria-label="anchor" href="#arg-plot-stat"></a></dt>
<dd><p>Statistic to plot. One of "Fst","Fstp","Dest" or "Gst_H"
[default "Fstp"].</p></dd>


<dt id="arg-plot-display">plot.display<a class="anchor" aria-label="anchor" href="#arg-plot-display"></a></dt>
<dd><p>If TRUE, a heatmap of the pairwise static chosen is
displayed in the plot window [default TRUE].</p></dd>


<dt id="arg-palette-divergent">palette.divergent<a class="anchor" aria-label="anchor" href="#arg-palette-divergent"></a></dt>
<dd><p>A color palette function for the heatmap plot
[default gl.colors("div")].</p></dd>


<dt id="arg-font-size">font.size<a class="anchor" aria-label="anchor" href="#arg-font-size"></a></dt>
<dd><p>Size of font for the labels of horizontal and vertical axes
of the heatmap [default 0.5].</p></dd>


<dt id="arg-plot-dir">plot.dir<a class="anchor" aria-label="anchor" href="#arg-plot-dir"></a></dt>
<dd><p>Directory in which to save files [default working directory].</p></dd>


<dt id="arg-plot-file">plot.file<a class="anchor" aria-label="anchor" href="#arg-plot-file"></a></dt>
<dd><p>Name for the RDS binary file to save (base name only,
exclude extension) [default NULL].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
 [default NULL, unless specified using gl.set.verbosity]</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Parameters passed to function <a href="https://talgalili.github.io/gplots/reference/heatmap.2.html" class="external-link">heatmap.2</a> (package
gplots).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Two lists, the first list contains matrices with genetic statistics
taken pairwise by population, the second list contains tables with the
genetic statistics for each pair of populations. If nboots &gt; 0, tables with
the four statistics calculated with Low Confidence Intervals (LCI) and High
Confidence Intervals (HCI).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Even though Fst and its relatives can predict evolutionary processes
 (Holsinger &amp; Weir, 2009), they are not true measures of genetic
  differentiation in the sense that they are dependent on the diversity
   within populations (Meirmans &amp; Hedrick, 2011), the number of populations
   analysed (Alcala &amp; Rosenberg, 2017) and are not monotonic
   (Sherwin et al., 2017). Recent approaches have been developed to
   accommodate these mathematical restrictions (G'ST; "Gst_H"; Hedrick, 2005,
and Jost's D; "Dest"; Jost, 2008). More recently, novel approaches based on
information theory (Mutual Information; Sherwin et al., 2017) and allele
frequencies (Allele Frequency Difference; Berner, 2019) have distinct
properties that make them valuable resources to interpret genetic
 differentiation between populations.</p>
<p>Note that each measure of genetic differentiation has advantages and
    drawbacks, and the decision of using a particular measure is usually
    based on the research question.</p>
<p><strong>Statistics calculated</strong></p>
<p>The equations used to calculate the statistics are shown below.</p>
<p></p><ul><li><p><em>Ho</em> - Unbiased estimate of observed heterozygosity across
    subpopulations (Nei, 1987, pp. 164, eq. 7.38) is calculated as:</p>
<p><img src="figures/Ho-equation.jpg" alt=""></p>
<p>where <em>Pkii</em> represents the proportion of homozygote <em>ii</em> for
    allele <em>i</em> in individual <em>k</em> and <em>s</em> represents the number
     of subpopulations.</p></li>
<li><p><em>Hs</em> - Unbiased estimate of the expected heterozygosity under
    Hardy-Weinberg equilibrium across subpopulations (Nei, 1987, pp. 164,
     eq. 7.39) is calculated as:</p>
<p><img src="figures/Hs-equation.jpg" alt=""></p>
<p>where <em>ñ</em> is the harmonic mean of <em>nk</em> (the number of
    individuals in each subpopulation), <em>pki</em> is the proportion
    (sometimes misleadingly called frequency) of allele <em>i</em> in
    subpopulation <em>k</em>.</p></li>
<li><p><em>Ht</em> - Heterozygosity for the total population (Nei, 1987, pp. 164,
     eq. 7.40) is calculated as:</p>
<p><img src="figures/Ht-equation.jpg" alt=""></p></li>
<li><p><em>Dst</em> - The average allele frequency differentiation between
    populations (Nei, 1987, pp. 163) is calculated as:</p>
<p><img src="figures/Dst-equation.jpg" alt=""></p></li>
<li><p><em>Htp</em> - Unbiased estimate of Heterozygosity for the total population
    (Nei, 1987, pp. 165) is calculated as:</p>
<p><img src="figures/Htp-equation.jpg" alt=""></p></li>
<li><p><em>Dstp</em> - Unbiased estimate of the average allele frequency
    differentiation between populations (Nei, 1987, pp. 165)</p></li>
<li><p><em>Fst</em> - Measure of the extent of genetic differentiation
    of subpopulations (Nei, 1987, pp. 162, eq. 7.34) is calculated as:</p>
<p><img src="figures/Fst-equation.jpg" alt=""></p></li>
<li><p><em>Fstp</em> - Unbiased measure of the extent of genetic differentiation
    of subpopulations (Nei, 1987, pp. 163, eq. 7.36) is calculated as:</p>
<p><img src="figures/Fstp-equation.jpg" alt=""></p></li>
<li><p><em>Dest</em> - Jost’s D (Jost, 2008, eq. 12)</p></li>
<li><p><em>Gst-max</em> - The maximum level that Gst can obtain for the observed
    amount of genetic variation (Hedrick 2005, eq. 4a) is calculated as:</p>
<p><img src="figures/GstMax-equation.jpg" alt=""></p></li>
<li><p><em>Gst-H</em> - Gst standardized by the maximum level that it can obtain
    for the observed amount of genetic variation (Hedrick 2005, eq. 4b) is
    calculated as:</p>
<p><img src="figures/Gst-H.jpg" alt=""></p></li>
</ul><p><strong>Confident Intervals</strong></p>
<p>The uncertainty of a parameter, in this case the mean of the statistic, can
be summarised by a confidence interval (CI) which includes the true parameter
value with a specified probability (i.e. confidence level; the parameter
"conf" in this function).</p>
<p>In this function, CI are obtained using Bootstrap which is an inference
method that samples with replacement the data (i.e. loci) and calculates the
 statistics every time.</p>
<p>This function uses the function <a href="https://rdrr.io/pkg/boot/man/boot.html" class="external-link">boot</a> (package boot) to perform
 the bootstrap replicates and the function <a href="https://rdrr.io/pkg/boot/man/boot.ci.html" class="external-link">boot.ci</a>
 (package boot) to perform the calculations for the CI.</p>
<p>Four different types of nonparametric CI can be calculated
  (parameter "CI.type" in this function):</p><ul><li><p>First order normal approximation interval ("norm").</p></li>
<li><p>Basic bootstrap interval ("basic").</p></li>
<li><p>Bootstrap percentile interval ("perc").</p></li>
<li><p>Adjusted bootstrap percentile interval ("bca").</p></li>
</ul><p>The studentized bootstrap interval ("stud") was not included in the CI types
 because it is computationally intensive, it may produce estimates outside
 the range of plausible values and it has been found to be erratic in
 practice, see for example the "Studentized (t) Intervals" section in:</p>
<p>www.r-bloggers.com/2019/09/understanding-bootstrap-confidence-interval-output-from-the-r-boot-package</p>
<p>Nice tutorials about the different types of CI can be found at:</p>
<p>https://www.datacamp.com/tutorial/bootstrap-r</p>

<p>Efron and Tibshirani (1993, p. 162) and Davison and Hinkley
     (1997, p. 194) suggest that the number of bootstrap replicates should
     be between 1000 and 2000.</p>
<p><strong>It is important</strong> to note that unreliable confident intervals will be
  obtained if too few number of bootstrap replicates are used.
  Therefore, the function <a href="https://rdrr.io/pkg/boot/man/boot.ci.html" class="external-link">boot.ci</a> will throw warnings and errors
   if bootstrap replicates are too few. Consider increasing then number of
   bootstrap replicates to at least 200.</p>
<p>The "bca" interval is often cited as the best for theoretical reasons,
   however it may produce unstable results if the bootstrap distribution
    is skewed or has extreme values. For example, you might get the warning
    "extreme order statistics used as endpoints" or the error "estimated
    adjustment 'a' is NA". In this case, you may want to use more bootstrap
    replicates or a different method or check your data for outliers.</p>
<p>The error "estimated adjustment 'w' is infinite" means that the estimated
   adjustment ‘w’ for the "bca" interval is infinite, which can happen when
   the empirical influence values are zero or very close to zero. This can
   be caused by various reasons, such as:</p>
<p>The number of bootstrap replicates is too small, the statistic of interest
    is constant or nearly constant across the bootstrap samples, the data
    contains outliers or extreme values.</p>
<p>You can try some possible solutions, such as:</p>
<p>Increasing the number of bootstrap replicates, using a different type of
bootstrap confidence interval or removing or transforming the outliers or
 extreme values.</p>
<p><strong>Plotting</strong></p>
<p>The plot can be customised by including any parameter(s) from the function
 <a href="https://talgalili.github.io/gplots/reference/heatmap.2.html" class="external-link">heatmap.2</a> (package gplots).</p>
<p>For the color palette you could try for example:</p>
<p>&gt; <code><a href="https://sjmgarnier.github.io/viridis/" class="external-link">library(viridis)</a></code></p>
<p>&gt; <code>res &lt;- gl.report.fstat(platypus.gl, palette.divergent = viridis)</code></p>
<p>If a plot.file is given, the plot arising from this function is saved as an
 "RDS" binary file using the function <a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a> (package base);
  can be reloaded with function <a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a> (package base). A file
  name must be specified for the plot to be saved.</p>
<p>If a plot directory (plot.dir) is specified, the gplot binary is saved to
 that directory; otherwise to the tempdir().</p>
<p>Your plot might not shown in full because your 'Plots' pane is too small
 (in RStudio).
 Increase the size of the 'Plots' pane before running the function.
 Alternatively, use the parameter 'plot.file' to save the plot to a file.</p>
<p><strong>Parallelisation</strong></p>
<p>If the parameter ncpus &gt; 1, parallelisation is enabled. In Windows, parallel
  computing employs a "socket" approach that starts new copies of R on each
   core. POSIX systems, on the other hand (Mac, Linux, Unix, and BSD),
   utilise a "forking" approach that replicates the whole current version of
    R and transfers it to a new core.</p>
<p>Opening and terminating R sessions in each core involves a significant
    amount of processing time, therefore parallelisation in Windows machines
   is only quicker than not usung parallelisation when nboots &gt; 1000-2000.</p>
    </div>
    <div id="references">
    <h2>References</h2>

<ul><li><p>Alcala, N., &amp; Rosenberg, N. A. (2017). Mathematical constraints on FST:
Biallelic markers in arbitrarily many populations. Genetics (206), 1581-1600.</p></li>
<li><p>Berner, D. (2019). Allele frequency difference AFD–an intuitive alternative
to FST for quantifying genetic population differentiation. Genes, 10(4), 308.</p></li>
<li><p>Davison AC, Hinkley DV (1997). Bootstrap Methods and their Application.
 Cambridge University Press: Cambridge.</p></li>
<li><p>Efron, B. (1979). Bootstrap methods: Another look at the jackknife. Annals of
Statistics 7, 1–26.</p></li>
<li><p>Efron B, Tibshirani RJ (1993). An Introduction to the Bootstrap. Chapman and
 Hall: London.</p></li>
<li><p>Hedrick, P. W. (2005). A standardized genetic differentiation measure.
Evolution, 59(8), 1633-1638.</p></li>
<li><p>Holsinger, K. E. (2012). Lecture notes in population genetics.</p></li>
<li><p>Holsinger, K. E., &amp; Weir, B. S. (2009). Genetics in geographically structured
 populations: defining, estimating and interpreting FST. Nature Reviews
 Genetics, 10(9), 639- 650.</p></li>
<li><p>Jost, L. (2008). GST and its relatives do not measure differentiation.
 Molecular Ecology, 17(18), 4015-4026.</p></li>
<li><p>Meirmans, P. G., &amp; Hedrick, P. W. (2011). Assessing population structure:
 FST and related measures. Molecular Ecology Resources, 11(1), 5-18.</p></li>
<li><p>Nei, M. (1987). Molecular evolutionary genetics: Columbia University Press.</p></li>
<li><p>Sherwin, W. B., Chao, A., Jost, L., &amp; Smouse, P. E. (2017). Information
 theory broadens the spectrum of molecular ecology and evolution. Trends in
  Ecology &amp; Evolution, 32(12), 948-963.</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>Custodian: Luis Mijangos – Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">gl.report.fstat</span><span class="op">(</span><span class="va">platypus.gl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.fstat </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Stat_matrices</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Stat_matrices$Fst</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              SEVERN_ABOVE SEVERN_BELOW TENTERFIELD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE           NA       0.0309      0.0392</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW       0.0309           NA      0.0436</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD        0.0392       0.0436          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Stat_matrices$Fstp</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              SEVERN_ABOVE SEVERN_BELOW TENTERFIELD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE           NA       0.0599      0.0754</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW       0.0599           NA      0.0836</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD        0.0754       0.0836          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Stat_matrices$Dest</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              SEVERN_ABOVE SEVERN_BELOW TENTERFIELD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE           NA       0.0101      0.0134</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW       0.0101           NA      0.0150</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD        0.0134       0.0150          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Stat_matrices$Gst_H</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              SEVERN_ABOVE SEVERN_BELOW TENTERFIELD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE           NA       0.0388      0.0498</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW       0.0388           NA      0.0554</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD        0.0498       0.0554          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Stat_tables</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       Stat_tables.SEVERN_ABOVE_vs_SEVERN_BELOW</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Fst                                     0.0309</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Fstp                                    0.0599</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dest                                    0.0101</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gst_H                                   0.0388</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       Stat_tables.SEVERN_ABOVE_vs_TENTERFIELD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Fst                                    0.0392</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Fstp                                   0.0754</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dest                                   0.0134</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gst_H                                  0.0498</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       Stat_tables.SEVERN_BELOW_vs_TENTERFIELD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Fst                                    0.0436</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Fstp                                   0.0836</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dest                                   0.0150</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gst_H                                  0.0554</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.plot.heatmap </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found more than one class "dist" in cache; using the first, from namespace 'spam'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Also defined by ‘BiocGenerics’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing a data matrix</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.fstat-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.plot.heatmap </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.fstat </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

