<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Imports presence/absence data from SilicoDArT to genlight {agegenet}
format (ploidy=1) — gl.read.silicodart • dartR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Imports presence/absence data from SilicoDArT to genlight {agegenet}
format (ploidy=1) — gl.read.silicodart"><meta property="og:description" content="DaRT provide the data as a matrix of entities (individual animals) across the
 top and attributes (P/A of sequenced fragment) down the side in a format
 that is unique to DArT. This program reads the data in to adegenet format
 for consistency with other programming activity. The script may require
 modification as DArT modify their data formats from time to time."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dartR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/dartRTutorials.html">dartR Tutorials</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Imports presence/absence data from SilicoDArT to genlight {agegenet}
format (ploidy=1)</h1>
    
    <div class="hidden name"><code>gl.read.silicodart.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>DaRT provide the data as a matrix of entities (individual animals) across the
 top and attributes (P/A of sequenced fragment) down the side in a format
 that is unique to DArT. This program reads the data in to adegenet format
 for consistency with other programming activity. The script may require
 modification as DArT modify their data formats from time to time.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.read.silicodart</span><span class="op">(</span></span>
<span>  <span class="va">filename</span>,</span>
<span>  ind.metafile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nas <span class="op">=</span> <span class="st">"-"</span>,</span>
<span>  topskip <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lastmetric <span class="op">=</span> <span class="st">"Reproducibility"</span>,</span>
<span>  probar <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>filename</dt>
<dd><p>Name of csv file containing the SilicoDArT data [required].</p></dd>


<dt>ind.metafile</dt>
<dd><p>Name of csv file containing metadata assigned to each
entity (individual) [default NULL].</p></dd>


<dt>nas</dt>
<dd><p>Missing data character [default '-'].</p></dd>


<dt>topskip</dt>
<dd><p>Number of rows to skip before the header row (containing the
specimen identities) [optional].</p></dd>


<dt>lastmetric</dt>
<dd><p>Specifies the last non genetic column (Default is
'Reproducibility'). Be sure to check if that is true, otherwise the number of
individuals will not match. You can also specify the last column by a number
 [default "Reproducibility"].</p></dd>


<dt>probar</dt>
<dd><p>Show progress bar [default TRUE].</p></dd>


<dt>verbose</dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2, or as set by gl.set.verbose()].</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An object of class <code>genlight</code> with ploidy set to 1, containing
the presence/absence data, and locus and individual metadata.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>gl.read.silicodart() opens the data file (csv comma delimited) and skips the
first n=topskip lines. The script assumes that the next line contains the
entity labels (specimen ids) followed immediately by the SNP data for the
first locus.</p>
<p>It reads the presence/absence data into a matrix of 1s and 0s, and inputs the
 locus metadata and specimen metadata. The locus metadata comprises a series
 of columns of values for each locus including the essential columns of
 CloneID and the desirable variables Reproducibility and PIC. Refer to
  documentation provide by DArT for an explanation of these columns.</p>
<p>The specimen metadata provides the opportunity to reassign specimens to
 populations, and to add other data relevant to the specimen. The key
 variables are id (specimen identity which must be the same and in the same
 order as the SilicoDArT file, each unique), pop (population assignment), lat
 (latitude, optional) and lon (longitude, optional). id, pop, lat, lon are
 the column headers in the csv file. Other optional columns can be added.</p>
<p>The data matrix, locus names (forced to be unique), locus metadata, specimen
 names, specimen metadata are combined into a genind object. Refer to the
 documentation for {adegenet} for further details.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="gl.read.dart.html">gl.read.dart</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Custodian: Bernd Gruber -- Post to
 <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">silicodartfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>,<span class="st">'testset_SilicoDArT.csv'</span>, package<span class="op">=</span><span class="st">'dartR'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>,ind.metafile <span class="op">=</span><span class="st">'testset_metadata_silicodart.csv'</span>, package<span class="op">=</span><span class="st">'dartR'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">testset.gs</span> <span class="op">&lt;-</span> <span class="fu">gl.read.silicodart</span><span class="op">(</span>filename <span class="op">=</span> <span class="va">silicodartfile</span>, ind.metafile <span class="op">=</span> <span class="va">metadata</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.read.silicodart </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Reading data from file: /home/runner/work/_temp/Library/dartR/extdata/testset_SilicoDArT.csv </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    This may take some time, please wait!</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Topskip not provided. Guessing topskip...</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Set topskip to  5 . Proceeding ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Added the following locus metrics:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CloneID AlleleSequence TrimmedSequence Chrom_Turtle_v4 ChromPos_Turtle_v4 AlnCnt_Turtle_v4 AlnEvalue_Turtle_v4 CallRate OneRatio PIC AvgReadDepth StDevReadDepth Qpmr Reproducibility .</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Recognised: 218 individuals and 255 Sequence Tags using /home/runner/work/_temp/Library/dartR/extdata/testset_SilicoDArT.csv </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Starting conversion to a genlight object ....</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    Please note conversion of bigger data sets will take some time!</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    Once finished, we recommend you save the object using gl.save(object, file="object.rdata")</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Adding individual metadata: /home/runner/work/_temp/Library/dartR/extdata/testset_metadata_silicodart.csv .</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Ids for individual metadata (at least a subset of) are matching!</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Found  218 matching ids out of 218 ids provided in the ind.metadata file. Subsetting loci now!.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">     Added pop factor.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    Added latlon data.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    Added  id  to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;">    Added  pop  to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;">    Added  lat  to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;">    Added  lon  to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;">    Added  sex  to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;">    Added  maturity  to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Genlight object created to hold Tag P/A data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.read.silicodart </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

