<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculates pairwise linkage disequilibrium by population — gl.report.ld.map • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculates pairwise linkage disequilibrium by population — gl.report.ld.map"><meta name="description" content="This function calculates pairwise linkage disequilibrium (LD) by population
using the function ld (package snpStats).
If SNPs are not mapped to a reference genome, the parameter
 ld_max_pairwise
 should be set as NULL (the default). In this case, the
function will assign the same chromosome (&quot;1&quot;) to all the SNPs in the dataset
 and assign a sequence from 1 to n loci as the position of each SNP. The
 function will then calculate LD for all possible SNP pair combinations.
If SNPs are mapped to a reference genome, the parameter
ld_max_pairwise
 should be filled out (i.e. not NULL). In this case, the
 information for SNP's position should be stored in the genlight accessor
  &quot;@position&quot; and the SNP's chromosome name in the accessor &quot;@chromosome&quot;
   (see examples). The function will then calculate LD within each chromosome
    and for all possible SNP pair combinations within a distance of
     ld_max_pairwise."><meta property="og:description" content="This function calculates pairwise linkage disequilibrium (LD) by population
using the function ld (package snpStats).
If SNPs are not mapped to a reference genome, the parameter
 ld_max_pairwise
 should be set as NULL (the default). In this case, the
function will assign the same chromosome (&quot;1&quot;) to all the SNPs in the dataset
 and assign a sequence from 1 to n loci as the position of each SNP. The
 function will then calculate LD for all possible SNP pair combinations.
If SNPs are mapped to a reference genome, the parameter
ld_max_pairwise
 should be filled out (i.e. not NULL). In this case, the
 information for SNP's position should be stored in the genlight accessor
  &quot;@position&quot; and the SNP's chromosome name in the accessor &quot;@chromosome&quot;
   (see examples). The function will then calculate LD within each chromosome
    and for all possible SNP pair combinations within a distance of
     ld_max_pairwise."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/green-striped-gecko/dartR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculates pairwise linkage disequilibrium by population</h1>
      <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl.report.ld.map.r" class="external-link"><code>R/gl.report.ld.map.r</code></a></small>
      <div class="d-none name"><code>gl.report.ld.map.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calculates pairwise linkage disequilibrium (LD) by population
using the function <code><a href="https://rdrr.io/pkg/snpStats/man/ld.html" class="external-link">ld</a></code> (package snpStats).</p>
<p>If SNPs are not mapped to a reference genome, the parameter
 <code>ld_max_pairwise</code>
 should be set as NULL (the default). In this case, the
function will assign the same chromosome ("1") to all the SNPs in the dataset
 and assign a sequence from 1 to n loci as the position of each SNP. The
 function will then calculate LD for all possible SNP pair combinations.</p>
<p>If SNPs are mapped to a reference genome, the parameter
<code>ld_max_pairwise</code>
 should be filled out (i.e. not NULL). In this case, the
 information for SNP's position should be stored in the genlight accessor
  "@position" and the SNP's chromosome name in the accessor "@chromosome"
   (see examples). The function will then calculate LD within each chromosome
    and for all possible SNP pair combinations within a distance of
     <code>ld_max_pairwise</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.report.ld.map</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  ld_max_pairwise <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maf <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  ld_stat <span class="op">=</span> <span class="st">"R.squared"</span>,</span>
<span>  ind.limit <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  stat_keep <span class="op">=</span> <span class="st">"AvgPIC"</span>,</span>
<span>  ld_threshold_pops <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  plot.out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_theme <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  histogram_colors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  boxplot_colors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bins <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  save2tmp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt id="arg-ld-max-pairwise">ld_max_pairwise<a class="anchor" aria-label="anchor" href="#arg-ld-max-pairwise"></a></dt>
<dd><p>Maximum distance in number of base pairs at which LD
should be calculated [default NULL].</p></dd>


<dt id="arg-maf">maf<a class="anchor" aria-label="anchor" href="#arg-maf"></a></dt>
<dd><p>Minor allele frequency (by population) threshold to filter out
loci. If a value &gt; 1 is provided it will be interpreted as MAC (i.e. the
 minimum number of times an allele needs to be observed) [default 0.05].</p></dd>


<dt id="arg-ld-stat">ld_stat<a class="anchor" aria-label="anchor" href="#arg-ld-stat"></a></dt>
<dd><p>The LD measure to be calculated: "LLR", "OR", "Q", "Covar",
"D.prime", "R.squared", and "R". See <code><a href="https://rdrr.io/pkg/snpStats/man/ld.html" class="external-link">ld</a></code>
 (package snpStats) for details [default "R.squared"].</p></dd>


<dt id="arg-ind-limit">ind.limit<a class="anchor" aria-label="anchor" href="#arg-ind-limit"></a></dt>
<dd><p>Minimum number of individuals that a population should
contain to take it in account to report loci in LD [default 10].</p></dd>


<dt id="arg-stat-keep">stat_keep<a class="anchor" aria-label="anchor" href="#arg-stat-keep"></a></dt>
<dd><p>Name of the column from the slot <code>loc.metrics</code> to be
used to choose SNP to be kept [default "AvgPIC"].</p></dd>


<dt id="arg-ld-threshold-pops">ld_threshold_pops<a class="anchor" aria-label="anchor" href="#arg-ld-threshold-pops"></a></dt>
<dd><p>LD threshold to report in the plot of "Number of
populations in which the same SNP pair are in LD" [default 0.2].</p></dd>


<dt id="arg-plot-out">plot.out<a class="anchor" aria-label="anchor" href="#arg-plot-out"></a></dt>
<dd><p>Specify if plot is to be produced [default TRUE].</p></dd>


<dt id="arg-plot-theme">plot_theme<a class="anchor" aria-label="anchor" href="#arg-plot-theme"></a></dt>
<dd><p>User specified theme [default NULL].</p></dd>


<dt id="arg-histogram-colors">histogram_colors<a class="anchor" aria-label="anchor" href="#arg-histogram-colors"></a></dt>
<dd><p>Vector with two color names for the borders and fill
[default NULL].</p></dd>


<dt id="arg-boxplot-colors">boxplot_colors<a class="anchor" aria-label="anchor" href="#arg-boxplot-colors"></a></dt>
<dd><p>A color palette for box plots by population or a list
 with as many colors as there are populations in the dataset
[default NULL].</p></dd>


<dt id="arg-bins">bins<a class="anchor" aria-label="anchor" href="#arg-bins"></a></dt>
<dd><p>Number of bins to display in histograms [default 50].</p></dd>


<dt id="arg-save-tmp">save2tmp<a class="anchor" aria-label="anchor" href="#arg-save-tmp"></a></dt>
<dd><p>If TRUE, saves any ggplots and listings to the session
temporary directory (tempdir) [default FALSE].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2, unless specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A dataframe with information for each SNP pair in LD.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function reports LD between SNP pairs by population.
The function <code><a href="gl.filter.ld.html">gl.filter.ld</a></code> filters out the SNPs in LD using as
input the results of <code>gl.report.ld.map</code>. The actual number of
SNPs to be filtered out depends on the parameters set in the function
<code><a href="gl.filter.ld.html">gl.filter.ld</a></code>.</p>
<p>Boxplots of LD by population and
a histogram showing LD frequency are presented.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="gl.filter.ld.html">gl.filter.ld</a></code></p>
<p>Other report functions:
<code><a href="gl.report.bases.html">gl.report.bases</a>()</code>,
<code><a href="gl.report.callrate.html">gl.report.callrate</a>()</code>,
<code><a href="gl.report.diversity.html">gl.report.diversity</a>()</code>,
<code><a href="gl.report.hamming.html">gl.report.hamming</a>()</code>,
<code><a href="gl.report.heterozygosity.html">gl.report.heterozygosity</a>()</code>,
<code><a href="gl.report.hwe.html">gl.report.hwe</a>()</code>,
<code><a href="gl.report.locmetric.html">gl.report.locmetric</a>()</code>,
<code><a href="gl.report.maf.html">gl.report.maf</a>()</code>,
<code><a href="gl.report.monomorphs.html">gl.report.monomorphs</a>()</code>,
<code><a href="gl.report.overshoot.html">gl.report.overshoot</a>()</code>,
<code><a href="gl.report.pa.html">gl.report.pa</a>()</code>,
<code><a href="gl.report.parent.offspring.html">gl.report.parent.offspring</a>()</code>,
<code><a href="gl.report.rdepth.html">gl.report.rdepth</a>()</code>,
<code><a href="gl.report.reproducibility.html">gl.report.reproducibility</a>()</code>,
<code><a href="gl.report.secondaries.html">gl.report.secondaries</a>()</code>,
<code><a href="gl.report.sexlinked.html">gl.report.sexlinked</a>()</code>,
<code><a href="gl.report.taglength.html">gl.report.taglength</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Luis Mijangos – Post to
 <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/green-striped-gecko/dartR.data" class="external-link">"dartR.data"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">platypus.gl</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl.filter.callrate.html">gl.filter.callrate</a></span><span class="op">(</span><span class="va">x</span>,threshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.filter.callrate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: Data may include monomorphic loci in call rate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">                    calculations for filtering</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Recalculating Call Rate</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Removing loci based on Call Rate, threshold = 1 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.ld.map-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.filter.callrate </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl.filter.monomorphs.html">gl.filter.monomorphs</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.filter.monomorphs </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Identifying monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Removing monomorphic loci and loci with all missing </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">                       data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.filter.monomorphs </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">x</span><span class="op">$</span><span class="va">position</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">$</span><span class="va">other</span><span class="op">$</span><span class="va">loc.metrics</span><span class="op">$</span><span class="va">ChromPos_Platypus_Chrom_NCBIv1</span></span></span>
<span class="r-in"><span><span class="va">x</span><span class="op">$</span><span class="va">chromosome</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">other</span><span class="op">$</span><span class="va">loc.metrics</span><span class="op">$</span><span class="va">Chrom_Platypus_Chrom_NCBIv1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ld_res</span> <span class="op">&lt;-</span> <span class="fu">gl.report.ld.map</span><span class="op">(</span><span class="va">x</span>,ld_max_pairwise <span class="op">=</span> <span class="fl">10000000</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.ld.map </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating pairwise LD in population SEVERN_ABOVE </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating pairwise LD in population SEVERN_BELOW </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating pairwise LD in population TENTERFIELD </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.ld.map-2.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.ld.map </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

