<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Reads SNP data from a csv file into a genlight object — gl.read.csv • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reads SNP data from a csv file into a genlight object — gl.read.csv"><meta name="description" content="This script takes SNP genotypes from a csv file, combines them with
individual and locus metrics and creates a genlight object."><meta property="og:description" content="This script takes SNP genotypes from a csv file, combines them with
individual and locus metrics and creates a genlight object."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/green-striped-gecko/dartR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Reads SNP data from a csv file into a genlight object</h1>
      <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl.read.csv.r" class="external-link"><code>R/gl.read.csv.r</code></a></small>
      <div class="d-none name"><code>gl.read.csv.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This script takes SNP genotypes from a csv file, combines them with
individual and locus metrics and creates a genlight object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.read.csv</span><span class="op">(</span></span>
<span>  <span class="va">filename</span>,</span>
<span>  transpose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ind.metafile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  loc.metafile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>Name of the csv file containing the SNP genotypes [required].</p></dd>


<dt id="arg-transpose">transpose<a class="anchor" aria-label="anchor" href="#arg-transpose"></a></dt>
<dd><p>If TRUE, rows are loci and columns are individuals
[default FALSE].</p></dd>


<dt id="arg-ind-metafile">ind.metafile<a class="anchor" aria-label="anchor" href="#arg-ind-metafile"></a></dt>
<dd><p>Name of the csv file containing the metrics for
individuals [optional].</p></dd>


<dt id="arg-loc-metafile">loc.metafile<a class="anchor" aria-label="anchor" href="#arg-loc-metafile"></a></dt>
<dd><p>Name of the csv file containing the metrics for
loci [optional].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2 or as specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A genlight object with the SNP data and associated metadata included.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The SNP data need to be in one of two forms. SNPs can be coded 0 for
homozygous reference, 2 for homozygous alternate, 1 for heterozygous, and NA
for missing values; or the SNP data can be coded A/A, A/C, C/T, G/A etc,
and -/- for missing data. In this format, the reference allele is the most
frequent allele, as used by DArT. Other formats will throw an error.</p>
<p>The SNP data need to be individuals as rows, labeled, and loci as columns,
also labeled. If the orientation is individuals as columns and loci by rows,
 then set transpose=TRUE.</p>
<p>The individual metrics need to be in a csv file, with headings, with a
 mandatory id column corresponding exactly to the individual identity labels
 provided with the SNP data and in the same order.</p>
<p>The locus metadata needs to be in a csv file with headings, with a mandatory
column headed AlleleID corresponding exactly to the locus identity labels
provided with the SNP data and in the same order.</p>
<p>Note that the locus metadata will be complemented by calculable statistics
corresponding to those that would be provided by Diversity Arrays Technology
(e.g. CallRate).</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Luis Mijangos – Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">csv_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>,<span class="st">'platy_test.csv'</span>, package<span class="op">=</span><span class="st">'dartR'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ind_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>,<span class="st">'platy_ind.csv'</span>, package<span class="op">=</span><span class="st">'dartR'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gl</span>  <span class="op">&lt;-</span> <span class="fu">gl.read.csv</span><span class="op">(</span>filename <span class="op">=</span> <span class="va">csv_file</span>, ind.metafile <span class="op">=</span> <span class="va">ind_metadata</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.read.csv </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">Warning: Locus metafile not provided, locus metrics will be</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">        calculated where this is possible</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Input data should be a csv file with individuals as rows, loci as columns</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    6 loci, confirming first 5: loci1 loci2 loci3 loci4 loci5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    13 individuals, confirming first 5: T158 T306 T305 T148 T149 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">    If these are reversed, re-run the script with transpose=TRUE</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Character data detected, assume genotypes are of the</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">            form C/C, A/T, C/G, -/- etc</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Data confirmed as biallelic</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  SNP coding converted to 0, 1, 2 and NA</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.compliance.check </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  The slot loc.all, which stores allele name for each locus, is empty. </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">Creating a dummy variable (A/C) to insert in this slot. </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking coding of SNPs</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    SNP data scored NA, 0, 1 or 2 confirmed</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking locus metrics and flags</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Recalculating locus metrics</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">    Dataset contains monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for loci with all missing data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    No loci with all missing data detected</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking whether individual names are unique.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for individual metrics</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: Creating a slot for individual metrics</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for population assignments</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    Population assignments confirmed</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Spelling of coordinates checked and changed if necessary to </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">            lat/lon</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.compliance.check </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;"> Added or updated  array(NA, nLoc(x)) to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added or updated  AvgPIC to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added or updated  OneRatioRef to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added or updated  OneRatioSnp to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added or updated  PICRef to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added or updated  PICSnp to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added or updated  CallRate to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added or updated  FreqHomRef to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added or updated  FreqHomSnp to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added or updated  FreqHets to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added or updated  monomorphs to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added or updated  maf to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added or updated  OneRatio to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added or updated  PIC to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;"> Added  id  to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added  pop  to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added  lat  to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added  long  to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added  group  to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;"> Added  age  to the other$ind.metrics slot.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.compliance.check </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking coding of SNPs</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    SNP data scored NA, 0, 1 or 2 confirmed</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking locus metrics and flags</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Recalculating locus metrics</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">    Dataset contains monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for loci with all missing data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    No loci with all missing data detected</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking whether individual names are unique.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for individual metrics</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    Individual metrics confirmed</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Checking for population assignments</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    Population assignments confirmed</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Spelling of coordinates checked and changed if necessary to </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">            lat/lon</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.compliance.check </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.read.csv </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

