<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Reports observed, expected and unbiased heterozygosities and FIS (inbreeding coefficient) by population or by individual from SNP data — gl.report.heterozygosity • dartR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Reports observed, expected and unbiased heterozygosities and FIS (inbreeding coefficient) by population or by individual from SNP data — gl.report.heterozygosity"><meta property="og:description" content="Calculates the observed, expected and unbiased expected (i.e.
corrected for sample size) heterozygosities and FIS (inbreeding coefficient)
for each population or the observed heterozygosity for each individual in a
genlight object."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dartR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.9.9.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/dartRTutorials.html">dartR Tutorials</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/green-striped-gecko/dartR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Reports observed, expected and unbiased heterozygosities and FIS (inbreeding coefficient) by population or by individual from SNP data</h1>
    <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl.report.heterozygosity.r" class="external-link"><code>R/gl.report.heterozygosity.r</code></a></small>
    <div class="hidden name"><code>gl.report.heterozygosity.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Calculates the observed, expected and unbiased expected (i.e.
corrected for sample size) heterozygosities and FIS (inbreeding coefficient)
for each population or the observed heterozygosity for each individual in a
genlight object.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.report.heterozygosity</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"pop"</span>,</span>
<span>  n.invariant <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  nboots <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  conf <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  CI.type <span class="op">=</span> <span class="st">"bca"</span>,</span>
<span>  ncpus <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  plot.display <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot.theme <span class="op">=</span> <span class="fu"><a href="theme_dartR.html">theme_dartR</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  plot.colors.pop <span class="op">=</span> <span class="fu"><a href="gl.colors.html">gl.colors</a></span><span class="op">(</span><span class="st">"dis"</span><span class="op">)</span>,</span>
<span>  plot.colors.ind <span class="op">=</span> <span class="fu"><a href="gl.colors.html">gl.colors</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  error.bar <span class="op">=</span> <span class="st">"SD"</span>,</span>
<span>  save2tmp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Name of the genlight object containing the SNP [required].</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>Calculate heterozygosity by population (method='pop') or by
individual (method='ind') [default 'pop'].</p></dd>


<dt id="arg-n-invariant">n.invariant<a class="anchor" aria-label="anchor" href="#arg-n-invariant"></a></dt>
<dd><p>An estimate of the number of invariant sequence tags used
to adjust the heterozygosity rate [default 0].</p></dd>


<dt id="arg-nboots">nboots<a class="anchor" aria-label="anchor" href="#arg-nboots"></a></dt>
<dd><p>Number of bootstrap replicates to obtain confident intervals
[default 0].</p></dd>


<dt id="arg-conf">conf<a class="anchor" aria-label="anchor" href="#arg-conf"></a></dt>
<dd><p>The confidence level of the required interval  [default 0.95].</p></dd>


<dt id="arg-ci-type">CI.type<a class="anchor" aria-label="anchor" href="#arg-ci-type"></a></dt>
<dd><p>Method to estimate confident intervals. One of
"norm", "basic", "perc" or "bca" [default "bca"].</p></dd>


<dt id="arg-ncpus">ncpus<a class="anchor" aria-label="anchor" href="#arg-ncpus"></a></dt>
<dd><p>Number of processes to be used in parallel operation. If ncpus
&gt; 1 parallel operation is activated,see "Details" section [default 1].</p></dd>


<dt id="arg-plot-display">plot.display<a class="anchor" aria-label="anchor" href="#arg-plot-display"></a></dt>
<dd><p>Specify if plot is to be produced [default TRUE].</p></dd>


<dt id="arg-plot-theme">plot.theme<a class="anchor" aria-label="anchor" href="#arg-plot-theme"></a></dt>
<dd><p>Theme for the plot. See Details for options
[default theme_dartR()].</p></dd>


<dt id="arg-plot-colors-pop">plot.colors.pop<a class="anchor" aria-label="anchor" href="#arg-plot-colors-pop"></a></dt>
<dd><p>A color palette for population plots or a list with
as many colors as there are populations in the dataset
[default gl.colors("dis")].</p></dd>


<dt id="arg-plot-colors-ind">plot.colors.ind<a class="anchor" aria-label="anchor" href="#arg-plot-colors-ind"></a></dt>
<dd><p>List of two color names for the borders and fill of
the plot by individual [default gl.colors(2)].</p></dd>


<dt id="arg-error-bar">error.bar<a class="anchor" aria-label="anchor" href="#arg-error-bar"></a></dt>
<dd><p>statistic to be plotted as error bar either "SD" (standard
deviation) or "SE" (standard error) or "CI" (confident intervals)
 [default "SD"].</p></dd>


<dt id="arg-save-tmp">save2tmp<a class="anchor" aria-label="anchor" href="#arg-save-tmp"></a></dt>
<dd><p>If TRUE, saves any ggplots and listings to the session
temporary directory (tempdir) [default FALSE].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default NULL, unless specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A dataframe containing population labels, heterozygosities, FIS,
their standard deviations and sample sizes</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Observed heterozygosity for a population takes the proportion of
heterozygous loci for each individual then averages over the individuals in
that population. The calculations take into account missing values.
Expected heterozygosity for a population takes the expected proportion of
heterozygotes, that is, expected under Hardy-Weinberg equilibrium, for each
locus, then averages this across the loci for an average estimate for the
population.</p>
<p>Expected heterozygosity is calculated using the correction for sample size
following equation 2 from Nei 1978.</p>
<p>Observed heterozygosity for individuals is calculated as the proportion of
loci that are heterozygous for that individual.</p>
<p>Finally, the loci that are invariant across all individuals in the dataset
(that is, across populations), is typically unknown. This can render
estimates of heterozygosity analysis specific, and so it is not valid to
compare such estimates across species or even across different analyses. This
is a similar problem faced by microsatellites. If you have an estimate of the
number of invariant sequence tags (loci) in your data, such as provided by
<code><a href="gl.report.secondaries.html">gl.report.secondaries</a></code>, you can specify it with the n.invariant
parameter to standardize your estimates of heterozygosity.</p>
<p><strong>NOTE</strong>: It is important to realise that estimation of adjusted
heterozygosity requires that secondaries not to be removed.</p>
<p>Heterozygosities and FIS (inbreeding coefficient) are calculated by locus
within each population using the following equations:</p><ul><li><p>Observed heterozygosity (Ho) = number of homozygotes / n_Ind,
where n_Ind is the number of individuals without missing data.</p></li>
<li><p>Observed heterozygosity adjusted (Ho.adj) &lt;- Ho * n_Loc /
 (n_Loc + n.invariant),
where n_Loc is the number of loci that do not have all missing data  and
n.invariant is an estimate of the number of invariant loci to adjust
heterozygosity.</p></li>
<li><p>Expected heterozygosity (He) = 1 - (p^2 + q^2),
where p is the frequency of the reference allele and q is the frequency of
the alternative allele.</p></li>
<li><p>Expected heterozygosity adjusted (He.adj) = He * n_Loc /
(n_Loc + n.invariant)</p></li>
<li><p>Unbiased expected heterozygosity (uHe) = He * (2 * n_Ind /
(2 * n_Ind - 1))</p></li>
<li><p>Inbreeding coefficient (FIS) = 1 - (mean(Ho) / mean(uHe))</p></li>
</ul><p><strong> Function's output </strong>
Output for method='pop' is an ordered barchart of observed heterozygosity,
unbiased expected heterozygosity and FIS (Inbreeding coefficient) across populations
together with a table of mean observed and expected heterozygosities and FIS
by population and their respective standard deviations (SD).
In the output, it is also reported by population: the number of loci used to
 estimate heterozygosity(n.Loc), the number of polymorphic loci (polyLoc),
 the number of monomorphic loci (monoLoc) and loci with all missing data
  (all_NALoc).
Output for method='ind' is a histogram and a boxplot of heterozygosity across
individuals.
 Plots and table are saved to the session temporary directory (tempdir)
 Examples of other themes that can be used can be consulted in</p><ul><li><p><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">https://ggplot2.tidyverse.org/reference/ggtheme.html</a> and</p></li>
<li><p><a href="https://yutannihilation.github.io/allYourFigureAreBelongToUs/ggthemes/" class="external-link">https://yutannihilation.github.io/allYourFigureAreBelongToUs/ggthemes/</a></p></li>
</ul><p><strong>Error bars</strong></p>
<p>The best method for presenting or assessing genetic statistics depends on
 the type of data you have and the specific questions you're trying to
 answer. Here's a brief overview of when you might use each method:</p>
<p><strong>1. Confidence Intervals ("CI"):</strong></p>
<p>- Usage: Often used to convey the precision of an estimate.</p>
<p>- Advantage: Confidence intervals give a range in which the true parameter
 (like a population mean) is likely to fall, given the data and a specified
 probability (like 95<!-- %). --></p>
<p>- In Context: For genetic statistics, if you're estimating a parameter,
  a 95<!-- % CI gives you a range in which you're 95% confident the true parameter -->
   lies.</p>
<p><strong>2. Standard Deviation ("SD"):</strong></p>
<p>- Usage: Describes the amount of variation from the average in a set of data.</p>
<p>- Advantage: Allows for an understanding of the spread of individual data
  points around the mean.</p>
<p>- In Context: If you're looking at the distribution of a quantitative trait
 (like height) in a population with a particular genotype, the SD can
 describe how much individual heights vary around the average height.</p>
<p><strong>3. Standard Error ("SE"):</strong></p>
<p>- Usage: Describes the precision of the sample mean as an estimate of the
 population mean.</p>
<p>- Advantage: Smaller than the SD in large samples; it takes into account
 both the SD and the sample size.</p>
<p>- In Context: If you want to know how accurately your sample mean represents
  the population mean, you'd look at the SE.</p>
<p><strong>Recommendation:</strong></p>
<p>- If you're trying to convey the precision of an estimate, confidence
  intervals are very useful.</p>
<p>- For understanding variability within a sample, standard deviation is key.</p>
<p>- To see how well a sample mean might estimate a population mean, consider
  the standard error.</p>
<p>In practice, geneticists often use a combination of these methods to
  analyze and present their data, depending on their research questions and
  the nature of the data.</p>
<p><strong>Confident Intervals</strong></p>
<p>The uncertainty of a parameter, in this case the mean of the statistic, can
be summarised by a confidence interval (CI) which includes the true parameter
value with a specified probability (i.e. confidence level; the parameter
"conf" in this function).</p>
<p>In this function, CI are obtained using Bootstrap which is an inference
method that samples with replacement the data (i.e. loci) and calculates the
 statistics every time.</p>
<p>This function uses the function <a href="https://rdrr.io/pkg/boot/man/boot.html" class="external-link">boot</a> (package boot) to perform
 the bootstrap replicates and the function <a href="https://rdrr.io/pkg/boot/man/boot.ci.html" class="external-link">boot.ci</a>
 (package boot) to perform the calculations for the CI.</p>
<p>Four different types of nonparametric CI can be calculated
  (parameter "CI.type" in this function):</p><ul><li><p>First order normal approximation interval ("norm").</p></li>
<li><p>Basic bootstrap interval ("basic").</p></li>
<li><p>Bootstrap percentile interval ("perc").</p></li>
<li><p>Adjusted bootstrap percentile interval ("bca").</p></li>
</ul><p>The studentized bootstrap interval ("stud") was not included in the CI types
 because it is computationally intensive, it may produce estimates outside
 the range of plausible values and it has been found to be erratic in
 practice, see for example the "Studentized (t) Intervals" section in:</p>
<p>www.r-bloggers.com/2019/09/understanding-bootstrap-confidence-interval-output-from-the-r-boot-package
     Efron and Tibshirani (1993, p. 162) and Davison and Hinkley
     (1997, p. 194) suggest that the number of bootstrap replicates should
     be between 1000 and 2000.</p>
<p><strong>It is important</strong> to note that unreliable confident intervals will be
  obtained if too few number of bootstrap replicates are used.
  Therefore, the function <a href="https://rdrr.io/pkg/boot/man/boot.ci.html" class="external-link">boot.ci</a> will throw warnings and errors
   if bootstrap replicates are too few. Consider increasing then number of
   bootstrap replicates to at least 200.</p>
<p>The "bca" interval is often cited as the best for theoretical reasons,
   however it may produce unstable results if the bootstrap distribution
    is skewed or has extreme values. For example, you might get the warning
    "extreme order statistics used as endpoints" or the error "estimated
    adjustment 'a' is NA". In this case, you may want to use more bootstrap
    replicates or a different method or check your data for outliers.</p>
<p>The error "estimated adjustment 'w' is infinite" means that the estimated
   adjustment ‘w’ for the "bca" interval is infinite, which can happen when
   the empirical influence values are zero or very close to zero. This can
   be caused by various reasons, such as:</p>
<p>The number of bootstrap replicates is too small, the statistic of interest
    is constant or nearly constant across the bootstrap samples, the data
    contains outliers or extreme values.</p>
<p>You can try some possible solutions, such as:</p>
<p>Increasing the number of bootstrap replicates, using a different type of
bootstrap confidence interval or removing or transforming the outliers or
 extreme values.</p>
<p><strong>Parallelisation</strong></p>
<p>If the parameter ncpus &gt; 1, parallelisation is enabled. In Windows, parallel
  computing employs a "socket" approach that starts new copies of R on each
   core. POSIX systems, on the other hand (Mac, Linux, Unix, and BSD),
   utilise a "forking" approach that replicates the whole current version of
    R and transfers it to a new core.</p>
<p>Opening and terminating R sessions in each core involves a significant
    amount of processing time, therefore parallelisation in Windows machines
   is only quicker than not usung parallelisation when nboots &gt; 1000-2000.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Nei, M. (1978). Estimation of average heterozygosity and genetic distance
from a small number of individuals. Genetics, 89(3), 583-590.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="gl.filter.heterozygosity.html">gl.filter.heterozygosity</a></code></p>
<p>Other unmatched report:
<code><a href="gl.allele.freq.html">gl.allele.freq</a>()</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Custodian: Luis Mijangos (Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a>)</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span> <span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/green-striped-gecko/dartR.data" class="external-link">"dartR.data"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">gl.report.heterozygosity</span><span class="op">(</span><span class="va">platypus.gl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.heterozygosity </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating Observed Heterozygosities, averaged across</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">                    loci, for each population</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating Expected Heterozygosities</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.heterozygosity-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE SEVERN_ABOVE 21.69526   991         1     496     495         9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW SEVERN_BELOW 15.90918   991         1     461     530         9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD   TENTERFIELD 38.06036   994         1     581     413         6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Ho     HoSD     HoSE HoLCI HoHCI   Ho.adj Ho.adjSD Ho.adjSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE 0.119874 0.175321 0.005544    NA    NA 0.119874 0.175321 0.005569</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW 0.114486 0.174360 0.005514    NA    NA 0.114486 0.174360 0.005539</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD  0.119842 0.160903 0.005088    NA    NA 0.119842 0.160903 0.005104</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              Ho.adjLCI Ho.adjHCI       He     HeSD     HeSE HeLCI HeHCI</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE        NA        NA 0.132964 0.174096 0.005505    NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW        NA        NA 0.129830 0.174659 0.005523    NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD         NA        NA 0.142741 0.173964 0.005501    NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   uHe    uHeSD    uHeSE uHeLCI uHeHCI   He.adj He.adjSD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE 0.136101 0.178203 0.005635     NA     NA 0.132964 0.174096</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW 0.134042 0.180327 0.005702     NA     NA 0.129830 0.174659</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD  0.144642 0.176279 0.005574     NA     NA 0.142741 0.173964</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              He.adjSE He.adjLCI He.adjHCI      FIS    FISSD    FISSE FISLCI</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE 0.005530        NA        NA 0.105976 0.312179 0.009872     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW 0.005548        NA        NA 0.126833 0.339163 0.010725     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD  0.005518        NA        NA 0.148111 0.304848 0.009640     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              FISHCI</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD      NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.heterozygosity </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">gl.report.heterozygosity</span><span class="op">(</span><span class="va">platypus.gl</span>,method<span class="op">=</span><span class="st">'ind'</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.heterozygosity </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating observed heterozygosity for individuals</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Note: No adjustment for invariant loci (n.invariant set to 0)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.heterozygosity-2.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  ind.name         Ho f.hom.ref  f.hom.alt n.Loc</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T27 0.10108696 0.7673913 0.13152174   920</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T35 0.14102564 0.7457265 0.11324786   936</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      SDS4 0.12000000 0.7751351 0.10486486   925</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SDS12 0.13246471 0.7578719 0.10966341   921</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS20 0.14482030 0.7452431 0.10993658   946</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS28 0.12404788 0.7660501 0.10990207   919</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS36 0.12459016 0.7628415 0.11256831   915</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        T3 0.12124464 0.7628755 0.11587983   932</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T11 0.12872340 0.7542553 0.11702128   940</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T19 0.12688172 0.7505376 0.12258065   930</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T28 0.13841202 0.7607296 0.10085837   932</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T36 0.12229437 0.7640693 0.11363636   924</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      SDS5 0.12742981 0.7624190 0.11015119   926</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SDS13 0.12955032 0.7601713 0.11027837   934</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS21 0.12540193 0.7652733 0.10932476   933</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS29 0.12605932 0.7648305 0.10911017   944</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS37 0.16230366 0.7392670 0.09842932   955</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        T4 0.12593383 0.7566702 0.11739594   937</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T12 0.12211221 0.7535754 0.12431243   909</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T20 0.13001083 0.7453954 0.12459372   923</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T29 0.13057671 0.7442873 0.12513602   919</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      SDS6 0.10064935 0.7738095 0.12554113   924</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SDS14 0.11563169 0.7591006 0.12526767   934</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS22 0.11620469 0.7771855 0.10660981   938</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS30 0.11111111 0.7777778 0.11111111   936</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS38 0.11220044 0.7657952 0.12200436   918</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        T5 0.12327311 0.7608927 0.11583422   941</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T13 0.13260870 0.7554348 0.11195652   920</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T21 0.13311331 0.7667767 0.10011001   909</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T30 0.11925602 0.7582057 0.12253829   914</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T38 0.13782051 0.7606838 0.10149573   936</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      SDS7 0.10951327 0.7876106 0.10287611   904</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SDS15 0.12850972 0.7678186 0.10367171   926</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS23 0.13811563 0.7526767 0.10920771   934</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS31 0.10825295 0.7642015 0.12754555   933</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS39 0.11995754 0.7749469 0.10509554   942</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        T6 0.13024758 0.7578041 0.11194833   929</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T14 0.13066955 0.7526998 0.11663067   926</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T22 0.13015184 0.7559653 0.11388286   922</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T31 0.11745690 0.7640086 0.11853448   928</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T39 0.08255814 0.7872093 0.13023256   860</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      SDS8 0.12676056 0.7573131 0.11592633   923</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SDS16 0.12229437 0.7673160 0.11038961   924</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS24 0.13432836 0.7473348 0.11833689   938</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS32 0.12030075 0.7669173 0.11278195   931</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS40 0.12913554 0.7502668 0.12059765   937</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        T7 0.12995595 0.7599119 0.11013216   908</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T15 0.12280702 0.7796053 0.09758772   912</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T23 0.11733046 0.7621098 0.12055974   929</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T32 0.13311688 0.7532468 0.11363636   924</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T40 0.14004376 0.7439825 0.11597374   914</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      SDS9 0.12606838 0.7628205 0.11111111   936</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SDS17 0.11612903 0.7655914 0.11827957   930</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS25 0.12192513 0.7668449 0.11122995   935</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS33 0.13856069 0.7561762 0.10526316   931</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS41 0.11970339 0.7669492 0.11334746   944</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        T8 0.12284483 0.7683190 0.10883621   928</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T16 0.13577586 0.7532328 0.11099138   928</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T24 0.11256831 0.7726776 0.11475410   915</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      SDS2 0.13163482 0.7505308 0.11783439   942</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T33 0.12500000 0.7586207 0.11637931   928</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T41 0.12568306 0.7693989 0.10491803   915</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SDS10 0.12861736 0.7491961 0.12218650   933</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SDS18 0.08441558 0.7792208 0.13636364   924</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS26 0.13291139 0.7594937 0.10759494   948</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS34 0.14285714 0.7532468 0.10389610   924</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        T1 0.12850972 0.7613391 0.11015119   926</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        T9 0.11281490 0.7601314 0.12705367   913</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T17 0.11675676 0.7783784 0.10486486   925</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T25 0.13275299 0.7529924 0.11425462   919</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      SDS3 0.13135593 0.7627119 0.10593220   944</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T34 0.12784399 0.7638137 0.10834236   923</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T42 0.14761376 0.7624861 0.08990011   901</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SDS11 0.12227074 0.7696507 0.10807860   916</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS19 0.10363248 0.7777778 0.11858974   936</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS27 0.13723404 0.7563830 0.10638298   940</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     SUS35 0.08884074 0.7757313 0.13542795   923</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        T2 0.12752391 0.7640808 0.10839532   941</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T10 0.12809473 0.7674919 0.10441335   929</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T18 0.13513514 0.7632432 0.10162162   925</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       T26 0.13588110 0.7600849 0.10403397   942</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.heterozygosity </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">n.inv</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl.report.secondaries.html">gl.report.secondaries</a></span><span class="op">(</span><span class="va">platypus.gl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.secondaries </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Counting ....</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Estimating parameters (lambda) of the Poisson expectation</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1.009082</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.6412174</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.4776479</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.3832083</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.3212201</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.2772315</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.2443198</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.2187315</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1982468</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1814658</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1674603</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1555902</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1453987</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1365515</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1287978</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1219458</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1158462</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1103813</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1054565</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1009953</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.09693515</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.09322415</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.08981909</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.08668362</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.08378692</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.08110273</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.07860852</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.07628483</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.07411477</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.07208364</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.07017853</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.06838812</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.06670239</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.06511244</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.06361039</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.06218918</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.06084248</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.05956463</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.05835052</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.05719552</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.05609546</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.05504654</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.05404531</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.05308862</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.05217359</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.05129759</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0504582</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04965321</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04888057</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04813839</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04742493</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04673858</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04607786</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04544136</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04482782</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04423602</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04366486</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0431133</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04258036</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04206513</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04156677</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04108448</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04061751</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.04016516</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03972677</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03930173</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03888944</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03848936</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03810096</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03772377</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0373573</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03700114</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03665486</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03631806</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03599039</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03567148</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.035361</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03505863</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03476407</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03447704</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03419725</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03392445</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0336584</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03339884</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03314557</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03289836</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03265701</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03242132</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0321911</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03196618</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03174638</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03153154</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03132149</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0311161</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03091522</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0307187</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03052641</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03033823</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.03015403</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02997369</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02979711</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02962417</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02945477</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02928881</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02912618</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0289668</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02881058</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02865743</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02850727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02836001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02821558</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0280739</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02793491</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02779852</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02766468</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02753332</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02740437</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02727778</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02715348</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02703142</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02691155</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0267938</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02667814</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0265645</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02645284</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02634312</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02623528</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02612929</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0260251</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02592267</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02582196</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02572293</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02562554</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02552976</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02543556</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02534288</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02525171</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02516201</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02507375</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0249869</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02490142</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0248173</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02473449</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02465298</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02457273</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02449372</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02441593</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02433932</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02426388</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02418959</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02411641</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02404433</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02397333</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02390339</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02383447</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02376657</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02369967</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02363374</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02356877</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02350474</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02344163</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02337942</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02331811</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02325766</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02319807</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02313931</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02308139</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02302427</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02296794</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0229124</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02285762</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0228036</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02275031</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02269776</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02264591</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02259477</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02254432</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02249455</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02244544</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02239699</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02234918</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02230201</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02225545</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02220951</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02216418</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02211943</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02207527</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02203168</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02198865</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02194618</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02190425</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02186286</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.021822</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02178165</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02174182</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0217025</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02166366</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02162532</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02158746</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02155007</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02151314</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02147667</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02144066</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02140509</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02136995</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02133525</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02130097</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02126711</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02123367</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02120063</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02116798</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02113574</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02110388</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0210724</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0210413</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02101057</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02098021</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02095021</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02092056</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02089127</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02086232</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02083371</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02080543</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02077749</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02074987</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02072258</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0206956</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02066894</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02064258</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02061653</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02059078</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02056533</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02054017</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02051529</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0204907</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02046639</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02044236</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0204186</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0203951</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02037188</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02034891</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02032621</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02030376</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02028156</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02025961</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0202379</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02021644</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02019521</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02017422</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02015346</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02013294</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02011264</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02009256</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0200727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02005307</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02003364</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.02001443</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01999543</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01997664</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01995805</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01993966</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01992148</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01990349</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01988569</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01986809</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01985067</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01983345</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01981641</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01979955</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01978287</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01976637</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01975005</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0197339</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01971792</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01970211</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01968647</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.019671</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01965569</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01964054</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01962555</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01961072</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01959604</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01958152</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01956716</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01955294</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01953887</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01952495</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01951117</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01949754</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01948405</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01947069</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01945748</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01944441</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01943147</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01941866</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01940599</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01939344</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01938103</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01936874</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01935658</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01934455</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01933264</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01932085</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01930918</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01929763</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0192862</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01927488</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01926368</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.0192526</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01924163</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01923076</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01922001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01920937</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01919884</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01918841</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01917808</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01916787</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01915775</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01914774</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Converged on Lambda of 0.0191378238139665 </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.heterozygosity-3.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Total number of SNP loci scored: 1000 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Number of sequence tags in total: 991 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Estimated number of invariant sequence tags: 51288 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Number of sequence tags with secondaries: 9 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Number of secondary SNP loci that would be removed on </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             filtering: 9 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Number of SNP loci that would be retained on filtering: 991 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Number of invariant sites in sequenced tags: 65820 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Mean length of sequence tags: 67.42684 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Total Number of invariant sites (including invariant sequence </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             tags): 3524008 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.secondaries </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu">gl.report.heterozygosity</span><span class="op">(</span><span class="va">platypus.gl</span>, n.invariant <span class="op">=</span> <span class="va">n.inv</span><span class="op">[</span><span class="fl">7</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.heterozygosity </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating Observed Heterozygosities, averaged across</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">                    loci, for each population</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating Expected Heterozygosities</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.heterozygosity-4.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       pop    n.Ind n.Loc    n.Loc.adj polyLoc monoLoc all_NALoc</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE SEVERN_ABOVE 21.69526   991 0.0002811348     496     495         9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW SEVERN_BELOW 15.90918   991 0.0002811348     461     530         9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD   TENTERFIELD 38.06036   994 0.0002819857     581     413         6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Ho     HoSD     HoSE HoLCI HoHCI  Ho.adj Ho.adjSD Ho.adjSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE 0.119874 0.175321 0.005544    NA    NA 3.4e-05 0.003560 0.000113</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW 0.114486 0.174360 0.005514    NA    NA 3.2e-05 0.003496 0.000111</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD  0.119842 0.160903 0.005088    NA    NA 3.4e-05 0.003368 0.000107</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              Ho.adjLCI Ho.adjHCI       He     HeSD     HeSE HeLCI HeHCI</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE        NA        NA 0.132964 0.174096 0.005505    NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW        NA        NA 0.129830 0.174659 0.005523    NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD         NA        NA 0.142741 0.173964 0.005501    NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   uHe    uHeSD    uHeSE uHeLCI uHeHCI  He.adj He.adjSD He.adjSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE 0.136101 0.178203 0.005635     NA     NA 3.7e-05 0.003672 0.000117</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW 0.134042 0.180327 0.005702     NA     NA 3.6e-05 0.003648 0.000116</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD  0.144642 0.176279 0.005574     NA     NA 4.0e-05 0.003777 0.000120</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              He.adjLCI He.adjHCI      FIS    FISSD    FISSE FISLCI FISHCI</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE        NA        NA 0.105976 0.312179 0.009872     NA     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW        NA        NA 0.126833 0.339163 0.010725     NA     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD         NA        NA 0.148111 0.304848 0.009640     NA     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.heterozygosity </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">gl.report.heterozygosity</span><span class="op">(</span><span class="va">platypus.gl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.heterozygosity </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating Observed Heterozygosities, averaged across</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">                    loci, for each population</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Calculating Expected Heterozygosities</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.heterozygosity-5.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE SEVERN_ABOVE 21.69526   991         1     496     495         9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW SEVERN_BELOW 15.90918   991         1     461     530         9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD   TENTERFIELD 38.06036   994         1     581     413         6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    Ho     HoSD     HoSE HoLCI HoHCI   Ho.adj Ho.adjSD Ho.adjSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE 0.119874 0.175321 0.005544    NA    NA 0.119874 0.175321 0.005569</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW 0.114486 0.174360 0.005514    NA    NA 0.114486 0.174360 0.005539</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD  0.119842 0.160903 0.005088    NA    NA 0.119842 0.160903 0.005104</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              Ho.adjLCI Ho.adjHCI       He     HeSD     HeSE HeLCI HeHCI</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE        NA        NA 0.132964 0.174096 0.005505    NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW        NA        NA 0.129830 0.174659 0.005523    NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD         NA        NA 0.142741 0.173964 0.005501    NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   uHe    uHeSD    uHeSE uHeLCI uHeHCI   He.adj He.adjSD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE 0.136101 0.178203 0.005635     NA     NA 0.132964 0.174096</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW 0.134042 0.180327 0.005702     NA     NA 0.129830 0.174659</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD  0.144642 0.176279 0.005574     NA     NA 0.142741 0.173964</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              He.adjSE He.adjLCI He.adjHCI      FIS    FISSD    FISSE FISLCI</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE 0.005530        NA        NA 0.105976 0.312179 0.009872     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW 0.005548        NA        NA 0.126833 0.339163 0.010725     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD  0.005518        NA        NA 0.148111 0.304848 0.009640     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              FISHCI</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_ABOVE     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SEVERN_BELOW     NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TENTERFIELD      NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.heterozygosity </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

