<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Runs Wright-Fisher simulations — gl.sim.WF.run • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Runs Wright-Fisher simulations — gl.sim.WF.run"><meta name="description" content="This function simulates populations made up of diploid organisms that
reproduce in non-overlapping generations. Each individual has a pair of
homologous chromosomes that contains interspersed selected and neutral loci.
For the initial generation, the genotype for each individual’s chromosomes is
 randomly drawn from distributions at linkage equilibrium and in
 Hardy-Weinberg equilibrium.
See documentation and tutorial for a complete description of the simulations.
These documents can be accessed at http://georges.biomatix.org/dartR
Take into account that the simulations will take a little bit longer the
 first time you use the function gl.sim.WF.run() because C++ functions must
  be compiled."><meta property="og:description" content="This function simulates populations made up of diploid organisms that
reproduce in non-overlapping generations. Each individual has a pair of
homologous chromosomes that contains interspersed selected and neutral loci.
For the initial generation, the genotype for each individual’s chromosomes is
 randomly drawn from distributions at linkage equilibrium and in
 Hardy-Weinberg equilibrium.
See documentation and tutorial for a complete description of the simulations.
These documents can be accessed at http://georges.biomatix.org/dartR
Take into account that the simulations will take a little bit longer the
 first time you use the function gl.sim.WF.run() because C++ functions must
  be compiled."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/green-striped-gecko/dartR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Runs Wright-Fisher simulations</h1>
      <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl.sim.WF.run.r" class="external-link"><code>R/gl.sim.WF.run.r</code></a></small>
      <div class="d-none name"><code>gl.sim.WF.run.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function simulates populations made up of diploid organisms that
reproduce in non-overlapping generations. Each individual has a pair of
homologous chromosomes that contains interspersed selected and neutral loci.
For the initial generation, the genotype for each individual’s chromosomes is
 randomly drawn from distributions at linkage equilibrium and in
 Hardy-Weinberg equilibrium.</p>
<p>See documentation and tutorial for a complete description of the simulations.
These documents can be accessed at http://georges.biomatix.org/dartR</p>
<p>Take into account that the simulations will take a little bit longer the
 first time you use the function gl.sim.WF.run() because C++ functions must
  be compiled.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.sim.WF.run</span><span class="op">(</span></span>
<span>  <span class="va">file_var</span>,</span>
<span>  <span class="va">ref_table</span>,</span>
<span>  x <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  file_dispersal <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  number_iterations <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  every_gen <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  sample_percent <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  store_phase1 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  interactive_vars <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-file-var">file_var<a class="anchor" aria-label="anchor" href="#arg-file-var"></a></dt>
<dd><p>Path of the variables file 'sim_variables.csv' (see details)
[required if interactive_vars = FALSE].</p></dd>


<dt id="arg-ref-table">ref_table<a class="anchor" aria-label="anchor" href="#arg-ref-table"></a></dt>
<dd><p>Reference table created by the function
<code><a href="gl.sim.WF.table.html">gl.sim.WF.table</a></code> [required].</p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Name of the genlight object containing the SNP data to extract
values for some simulation variables (see details) [default NULL].</p></dd>


<dt id="arg-file-dispersal">file_dispersal<a class="anchor" aria-label="anchor" href="#arg-file-dispersal"></a></dt>
<dd><p>Path of the file with the dispersal table created with
the function <code><a href="gl.sim.create_dispersal.html">gl.sim.create_dispersal</a></code> [default NULL].</p></dd>


<dt id="arg-number-iterations">number_iterations<a class="anchor" aria-label="anchor" href="#arg-number-iterations"></a></dt>
<dd><p>Number of iterations of the simulations [default 1].</p></dd>


<dt id="arg-every-gen">every_gen<a class="anchor" aria-label="anchor" href="#arg-every-gen"></a></dt>
<dd><p>Generation interval at which simulations should be stored in
a genlight object [default 10].</p></dd>


<dt id="arg-sample-percent">sample_percent<a class="anchor" aria-label="anchor" href="#arg-sample-percent"></a></dt>
<dd><p>Percentage of individuals, from the total population,
to sample and save in the genlight object every generation [default 50].</p></dd>


<dt id="arg-store-phase-">store_phase1<a class="anchor" aria-label="anchor" href="#arg-store-phase-"></a></dt>
<dd><p>Whether to store simulations of phase 1 in genlight
objects [default FALSE].</p></dd>


<dt id="arg-interactive-vars">interactive_vars<a class="anchor" aria-label="anchor" href="#arg-interactive-vars"></a></dt>
<dd><p>Run a shiny app to input interactively the values of
simulations variables [default TRUE].</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Set the seed for the simulations [default NULL].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2, unless specified using gl.set.verbosity].</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Any variable and its value can be added separately within the
function, will be changed over the input value supplied by the csv file. See
tutorial.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns genlight objects with simulated data.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Values for simulation variables can be submitted into the function
interactively through a shiny app if interactive_vars = TRUE. Optionally, if
interactive_vars = FALSE, values for variables can be submitted by using the
csv file 'sim_variables.csv' which can be found by typing in the R console:
 system.file('extdata', 'sim_variables.csv', package ='dartR').</p>
<p>The values of the variables can be modified using the third column (“value”)
of this file.</p>
<p>The output of the simulations can be analysed seemingly with other dartR
functions.</p>
<p>If a genlight object is used as input for some of the simulation variables,
this function access the information stored in the slots x$position and
x$chromosome.</p>
<p>To show further information of the variables in interactive mode, it might be
 necessary to call first: 'library(shinyBS)' for the information to be
 displayed.</p>
<p>The main characteristics of the simulations are:</p><ul><li><p>Simulations can be parameterised with real-life genetic
characteristics such as the number, location, allele frequency and the
distribution of fitness effects (selection coefficients and dominance) of
loci under selection.</p></li>
<li><p>Simulations can recreate specific life histories and demographics, such
 as source populations, dispersal rate, number of generations, founder
 individuals, effective population size and census population size.</p></li>
<li><p>Each allele in each individual is an agent (i.e., each allele is
explicitly simulated).</p></li>
<li><p>Each locus can be customisable regarding its allele frequencies,
selection coefficients, and dominance.</p></li>
<li><p>The number of loci, individuals, and populations to be simulated is
only limited by computing resources.</p></li>
<li><p>Recombination is accurately modeled, and it is possible to use real
recombination maps as input.</p></li>
<li><p>The ratio between effective population size and census population size
can be easily controlled.</p></li>
<li><p>The output of the simulations are genlight objects for each generation
or a subset of generations.</p></li>
<li><p>Genlight objects can be used as input for some simulation variables.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="gl.sim.WF.table.html">gl.sim.WF.table</a></code></p>
<p>Other simulation functions:
<code><a href="gl.sim.WF.table.html">gl.sim.WF.table</a>()</code>,
<code><a href="gl.sim.create_dispersal.html">gl.sim.create_dispersal</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Luis Mijangos – Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">ref_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="gl.sim.WF.table.html">gl.sim.WF.table</a></span><span class="op">(</span>file_var<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, </span></span>
<span class="r-in"><span><span class="st">'ref_variables.csv'</span>, package <span class="op">=</span> <span class="st">'dartR'</span><span class="op">)</span>,interactive_vars <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res_sim</span> <span class="op">&lt;-</span> <span class="fu">gl.sim.WF.run</span><span class="op">(</span>file_var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, </span></span>
<span class="r-in"><span><span class="st">'sim_variables.csv'</span>, package <span class="op">=</span><span class="st">'dartR'</span><span class="op">)</span>,ref_table<span class="op">=</span><span class="va">ref_table</span>,</span></span>
<span class="r-in"><span>interactive_vars <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

