<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Converts a genlight object into eigenstrat format — gl2eigenstrat • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Converts a genlight object into eigenstrat format — gl2eigenstrat"><meta name="description" content="The output of this function are three files:
genotype file: contains genotype data for each individual at each SNP
with an extension 'eigenstratgeno.'
snp file: contains information about each SNP with an extension 'snp.'
indiv file: contains information about each individual with an
extension 'ind.'

"><meta property="og:description" content="The output of this function are three files:
genotype file: contains genotype data for each individual at each SNP
with an extension 'eigenstratgeno.'
snp file: contains information about each SNP with an extension 'snp.'
indiv file: contains information about each individual with an
extension 'ind.'

"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/green-striped-gecko/dartR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Converts a genlight object into eigenstrat format</h1>
      <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl2eigenstrat.r" class="external-link"><code>R/gl2eigenstrat.r</code></a></small>
      <div class="d-none name"><code>gl2eigenstrat.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The output of this function are three files:</p><ul><li><p>genotype file: contains genotype data for each individual at each SNP
with an extension 'eigenstratgeno.'</p></li>
<li><p>snp file: contains information about each SNP with an extension 'snp.'</p></li>
<li><p>indiv file: contains information about each individual with an
extension 'ind.'</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl2eigenstrat</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  outfile <span class="op">=</span> <span class="st">"gl_eigenstrat"</span>,</span>
<span>  outpath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  snp_pos <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  snp_chr <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  pos_cM <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  sex_code <span class="op">=</span> <span class="st">"unknown"</span>,</span>
<span>  phen_value <span class="op">=</span> <span class="st">"Case"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt id="arg-outfile">outfile<a class="anchor" aria-label="anchor" href="#arg-outfile"></a></dt>
<dd><p>File name of the output file [default 'gl_eigenstrat'].</p></dd>


<dt id="arg-outpath">outpath<a class="anchor" aria-label="anchor" href="#arg-outpath"></a></dt>
<dd><p>Path where to save the output file
[default tempdir(), mandated by CRAN]. Use outpath=getwd() or outpath='.'
 when calling this function to direct output files to your working directory.</p></dd>


<dt id="arg-snp-pos">snp_pos<a class="anchor" aria-label="anchor" href="#arg-snp-pos"></a></dt>
<dd><p>Field name from the slot loc.metrics where the SNP position is
stored [default 1].</p></dd>


<dt id="arg-snp-chr">snp_chr<a class="anchor" aria-label="anchor" href="#arg-snp-chr"></a></dt>
<dd><p>Field name from the slot loc.metrics where the chromosome of
each is stored [default 1].</p></dd>


<dt id="arg-pos-cm">pos_cM<a class="anchor" aria-label="anchor" href="#arg-pos-cm"></a></dt>
<dd><p>A vector, with as many elements as there are loci, containing
the SNP position in morgans or centimorgans [default 1].</p></dd>


<dt id="arg-sex-code">sex_code<a class="anchor" aria-label="anchor" href="#arg-sex-code"></a></dt>
<dd><p>A vector, with as many elements as there are individuals,
containing the sex code ('male', 'female', 'unknown') [default 'unknown'].</p></dd>


<dt id="arg-phen-value">phen_value<a class="anchor" aria-label="anchor" href="#arg-phen-value"></a></dt>
<dd><p>A vector, with as many elements as there are individuals,
containing the phenotype value ('Case', 'Control') [default 'Case'].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log ; 3, progress and results summary; 5, full report
[default 2 or as specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>returns no value (i.e. NULL)</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Eigenstrat only accepts chromosomes coded as numeric values, as follows:
X chromosome is encoded as 23, Y is encoded as 24, mtDNA is encoded as
90, and XY is encoded as 91. SNPs with illegal chromosome values, such
as 0, will be removed.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>

<ul><li><p>Patterson, N., Price, A. L., &amp; Reich, D. (2006). Population structure
and eigenanalysis. PLoS genetics, 2(12), e190.</p></li>
<li><p>Price, A. L., Patterson, N. J., Plenge, R. M., Weinblatt, M. E.,
Shadick, N. A., &amp; Reich, D. (2006). Principal components analysis corrects
for stratification in genome-wide association studies. Nature genetics,
38(8), 904-909.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Luis Mijangos (Post to <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/green-striped-gecko/dartR.data" class="external-link">"dartR.data"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">gl2eigenstrat</span><span class="op">(</span><span class="va">platypus.gl</span>,snp_pos<span class="op">=</span><span class="st">'ChromPos_Platypus_Chrom_NCBIv1'</span>,</span></span>
<span class="r-in"><span>snp_chr <span class="op">=</span> <span class="st">'Chrom_Platypus_Chrom_NCBIv1'</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl2eigenstrat </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl2eigenstrat </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

