<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Package SNPRelate relies on a bit-level representation of a SNP dataset that
competes with {adegenet} genlight objects and associated files. This
function converts a genlight object to a gds format file."><title>Converts a genlight object into gds format — gl2gds • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Converts a genlight object into gds format — gl2gds"><meta property="og:description" content="Package SNPRelate relies on a bit-level representation of a SNP dataset that
competes with {adegenet} genlight objects and associated files. This
function converts a genlight object to a gds format file."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Converts a genlight object into gds format</h1>
      
      <div class="d-none name"><code>gl2gds.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Package SNPRelate relies on a bit-level representation of a SNP dataset that
competes with {adegenet} genlight objects and associated files. This
function converts a genlight object to a gds format file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl2gds</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  outfile <span class="op">=</span> <span class="st">"gl_gds.gds"</span>,</span>
<span>  outpath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  snp_pos <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>  snp_chr <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>  chr_format <span class="op">=</span> <span class="st">"character"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt>outfile</dt>
<dd><p>File name of the output file (including extension)
[default 'gl_gds.gds'].</p></dd>


<dt>outpath</dt>
<dd><p>Path where to save the output file
[default tempdir(), mandated by CRAN]. Use outpath=getwd() or outpath='.'
 when calling this function to direct output files to your working directory.</p></dd>


<dt>snp_pos</dt>
<dd><p>Field name from the slot loc.metrics where the SNP position is
stored [default '0'].</p></dd>


<dt>snp_chr</dt>
<dd><p>Field name from the slot loc.metrics where the chromosome of
each is stored [default '0'].</p></dd>


<dt>chr_format</dt>
<dd><p>Whether chromosome information is stored as 'numeric' or as
'character', see details [default 'character'].</p></dd>


<dt>verbose</dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2 or as specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function orders the SNPS by chromosome and by position before converting
to SNPRelate format, as required by this package.</p>
<p>The chromosome of each SNP can be a character or numeric, as described in the
vignette of SNPRelate:
'snp.chromosome, an integer or character mapping for each chromosome.
Integer: numeric values 1-26, mapped in order from 1-22, 23=X, 24=XY
(the pseudoautosomal region), 25=Y, 26=M (the mitochondrial probes), and 0
for probes with unknown positions; it does not allow NA. Character: “X”,
 “XY”, “Y” and “M” can be used here, and a blank string indicating unknown
 position.'</p>
<p>When using some functions from package SNPRelate with datasets other than
humans it might be necessary to use the option autosome.only=FALSE to avoid
detecting chromosome coding. So, it is important to read the documentation of
the function before using it.</p>
<p>The chromosome information for unmapped SNPS is coded as 0, as required by
SNPRelate.</p>
<p>Remember to close the GDS file before working in a different GDS object with
the function <a href="https://rdrr.io/pkg/SNPRelate/man/snpgdsClose.html" class="external-link">snpgdsClose</a> (package SNPRelate).</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Luis Mijangos (Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="fu">gl2gds</span><span class="op">(</span><span class="va">platypus.gl</span>,snp_pos<span class="op">=</span><span class="st">'ChromPos_Platypus_Chrom_NCBIv1'</span>,</span></span>
<span class="r-in"><span>snp_chr <span class="op">=</span> <span class="st">'Chrom_Platypus_Chrom_NCBIv1'</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl2gds </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Converting SNP data to gds format</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Writing data to file /tmp/Rtmp7jvyOF/gl_gds.gds </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #0000BB;">Structure of gds file</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Some values of 'snp.position' are invalid (should be &gt; 0)!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The file name: /tmp/Rtmp7jvyOF/gl_gds.gds </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The total number of samples: 81 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The total number of SNPs: 1000 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SNP genotypes are stored in individual-major mode (SNP X Sample).</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The number of valid samples: 81 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The number of biallelic unique SNPs: 921 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> File: /tmp/Rtmp7jvyOF/gl_gds.gds <span style="font-weight: 100;">(31.2K)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> +    [  ] *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |--+ sample.id   <span style="font-weight: 100;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">81</span> <span style="font-weight: 100;">ZIP_ra(49.9%), 200B</span> <span style="font-weight: 100;">}</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |--+ snp.id   <span style="font-weight: 100;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">1000</span> <span style="font-weight: 100;">ZIP_ra(34.0%), 5.3K</span> <span style="font-weight: 100;">}</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |--+ snp.position   <span style="font-weight: 100;">{</span> <span style="color: #555555;">Int32</span> <span style="color: #555555;">1000</span> <span style="font-weight: 100;">ZIP_ra(86.6%), 3.4K</span> <span style="font-weight: 100;">}</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |--+ snp.chromosome   <span style="font-weight: 100;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">1000</span> <span style="font-weight: 100;">ZIP_ra(1.90%), 463B</span> <span style="font-weight: 100;">}</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> |--+ snp.allele   <span style="font-weight: 100;">{</span> <span style="color: #555555;">Str8</span> <span style="color: #555555;">1000</span> <span style="font-weight: 100;">ZIP_ra(18.9%), 765B</span> <span style="font-weight: 100;">}</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> \--+ genotype   <span style="font-weight: 100;">{</span> <span style="color: #555555;">Bit2</span> <span style="color: #555555;">1000x81</span><span style="font-weight: 100;">, 19.8K</span> <span style="font-weight: 100;">}</span> *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Closing file /tmp/Rtmp7jvyOF/gl_gds.gds </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl2gds </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

