<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function exports a genlight object into PLINK format and save it into a
file.
This function produces the following PLINK files: bed, bim, fam, ped and map."><title>Converts a genlight object into PLINK format — gl2plink • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Converts a genlight object into PLINK format — gl2plink"><meta property="og:description" content="This function exports a genlight object into PLINK format and save it into a
file.
This function produces the following PLINK files: bed, bim, fam, ped and map."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.6.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Converts a genlight object into PLINK format</h1>
      
      <div class="d-none name"><code>gl2plink.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function exports a genlight object into PLINK format and save it into a
file.
This function produces the following PLINK files: bed, bim, fam, ped and map.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl2plink</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  plink_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  bed_file <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  outfile <span class="op">=</span> <span class="st">"gl_plink"</span>,</span>
<span>  outpath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  chr_format <span class="op">=</span> <span class="st">"character"</span>,</span>
<span>  pos_cM <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>  ID_dad <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>  ID_mom <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>  sex_code <span class="op">=</span> <span class="st">"unknown"</span>,</span>
<span>  phen_value <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt>plink_path</dt>
<dd><p>Path of PLINK binary file [default getwd()].</p></dd>


<dt>bed_file</dt>
<dd><p>Whether create PLINK files .bed, .bim and .fam
[default FALSE].</p></dd>


<dt>outfile</dt>
<dd><p>File name of the output file [default 'gl_plink'].</p></dd>


<dt>outpath</dt>
<dd><p>Path where to save the output file
[default tempdir(), mandated by CRAN]. Use outpath=getwd() or outpath='.'
 when calling this function to direct output files to your working directory.</p></dd>


<dt>chr_format</dt>
<dd><p>Whether chromosome information is stored as 'numeric' or as
'character', see details [default 'character'].</p></dd>


<dt>pos_cM</dt>
<dd><p>A vector, with as many elements as there are loci, containing
the SNP position in morgans or centimorgans [default '0'].</p></dd>


<dt>ID_dad</dt>
<dd><p>A vector, with as many elements as there are individuals,
containing the ID of the father, '0' if father isn't in dataset [default '0'].</p></dd>


<dt>ID_mom</dt>
<dd><p>A vector, with as many elements as there are individuals,
containing the ID of the mother, '0' if mother isn't in dataset [default '0'].</p></dd>


<dt>sex_code</dt>
<dd><p>A vector, with as many elements as there are individuals,
containing the sex code ('male', 'female', 'unknown'). Sex information needs 
just to start with an "F" or "f" for females, with an "M" or "m" for males 
and with a "U", "u" or being empty if the sex is unknown [default 'unknown'].</p></dd>


<dt>phen_value</dt>
<dd><p>A vector, with as many elements as there are individuals,
containing the phenotype value. '1' = control, '2' = case, '0' = unknown
[default '0'].</p></dd>


<dt>verbose</dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2 or as specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>To create PLINK files .bed, .bim and .fam (bed_file = TRUE), it is necessary
to download the binary file of PLINK 1.9 and provide its path (plink_path).
The binary file can be downloaded from:
<a href="https://www.cog-genomics.org/plink/" class="external-link">https://www.cog-genomics.org/plink/</a></p>
<p>After downloading, unzip the file, access the unzipped folder and move the 
binary file ("plink") to your working directory.</p>
<p>If you are using a Mac, you might need to open the binary first to grant 
access to the binary.</p>
<p>The chromosome of each SNP can be a character or numeric. The chromosome
information for unmapped SNPS is coded as 0.
Family ID is taken from  x$pop.
Within-family ID (cannot be '0') is taken from indNames(x).
Variant identifier is taken from locNames(x).
SNP position is taken from the accessor x$position.
Chromosome name is taken from the accessor x$chromosome
Note that if names of populations or individuals contain spaces, they are 
replaced by an underscore "_".</p>
<p>If you like to use chromosome information when converting to plink format and
 your chromosome names are not from human, you need to change the chromosome 
 names as 'contig1', 'contig2', etc. as described in the section "Nonstandard
  chromosome IDs" in the following link:
  https://www.cog-genomics.org/plink/1.9/input</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Purcell, Shaun, et al. 'PLINK: a tool set for whole-genome association and
population-based linkage analyses.' The American journal of human genetics
81.3 (2007): 559-575.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Luis Mijangos (Post to
 <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/green-striped-gecko/dartR.data" class="external-link">"dartR.data"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">test</span> <span class="op">&lt;-</span> <span class="va">platypus.gl</span></span></span>
<span class="r-in"><span><span class="co"># assigning SNP position</span></span></span>
<span class="r-in"><span><span class="va">test</span><span class="op">$</span><span class="va">position</span> <span class="op">&lt;-</span> <span class="va">test</span><span class="op">$</span><span class="va">other</span><span class="op">$</span><span class="va">loc.metrics</span><span class="op">$</span><span class="va">ChromPos_Platypus_Chrom_NCBIv1</span></span></span>
<span class="r-in"><span><span class="co"># assigning a dummy name for chromosomes</span></span></span>
<span class="r-in"><span><span class="va">test</span><span class="op">$</span><span class="va">chromosome</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="st">"1"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">gl2plink</span><span class="op">(</span><span class="va">test</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl2plink </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl2plink </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

