<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Converts a genlight object to nexus format PAUP SVDquartets — gl2svdquartets • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Converts a genlight object to nexus format PAUP SVDquartets — gl2svdquartets"><meta name="description" content="The output nexus file contains the SNP data in one of two forms, depending
upon what you regard as most appropriate. One form, that used by Chifman and
 Kubatko, has two lines per individual, one providing the reference SNP the
  second providing the alternate SNP (method=1)."><meta property="og:description" content="The output nexus file contains the SNP data in one of two forms, depending
upon what you regard as most appropriate. One form, that used by Chifman and
 Kubatko, has two lines per individual, one providing the reference SNP the
  second providing the alternate SNP (method=1)."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/green-striped-gecko/dartR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Converts a genlight object to nexus format PAUP SVDquartets</h1>
      <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl2svdquartets.r" class="external-link"><code>R/gl2svdquartets.r</code></a></small>
      <div class="d-none name"><code>gl2svdquartets.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The output nexus file contains the SNP data in one of two forms, depending
upon what you regard as most appropriate. One form, that used by Chifman and
 Kubatko, has two lines per individual, one providing the reference SNP the
  second providing the alternate SNP (method=1).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl2svdquartets</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  outfile <span class="op">=</span> <span class="st">"svd.nex"</span>,</span>
<span>  outpath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  method <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Name of the genlight object containing the SNP data or tag P/A data
[required].</p></dd>


<dt id="arg-outfile">outfile<a class="anchor" aria-label="anchor" href="#arg-outfile"></a></dt>
<dd><p>File name of the output file (including extension)
[default 'svd.nex'].</p></dd>


<dt id="arg-outpath">outpath<a class="anchor" aria-label="anchor" href="#arg-outpath"></a></dt>
<dd><p>Path where to save the output file
[default tempdir(), mandated by CRAN]. Use outpath=getwd() when calling this
function or set.tempdir &lt;- getwd() elsewhere in your script to direct output
 files to your working directory.</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>Method = 1, nexus file with two lines per individual; method =
2, nexus file with one line per individual, ambiguity codes for SNP
genotypes, 0 or 1 for presence/absence data [default 2].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2 or as specified using gl.set.verbosity]</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>returns no value (i.e. NULL)</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>A second form, recommended by Dave Swofford, has a single line per
  individual, resolving heterozygous SNPs by replacing them with standard
ambiguity codes (method=2).</p>
<p>If the data are tag presence/absence, then method=2 is assumed.</p>
<p>Note that the genlight object must contain at least two populations for this
 function to work.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Chifman, J. and L. Kubatko. 2014. Quartet inference from SNP data
under the coalescent. Bioinformatics 30: 3317-3324</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Arthur Georges (Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="va">testset.gl</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">gg</span><span class="op">@</span><span class="va">other</span><span class="op">$</span><span class="va">loc.metrics</span> <span class="op">&lt;-</span> <span class="va">gg</span><span class="op">@</span><span class="va">other</span><span class="op">$</span><span class="va">loc.metrics</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>,<span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu">gl2svdquartets</span><span class="op">(</span><span class="va">gg</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl2svdquartets </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: genlight object may contain monomorphic loci</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    Extacting SNP bases and creating records for each individual</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    Sorting ....</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    Creating lookup table for ambiguity codes</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    Applying ambiguity codes (if method=2) and reconstructing sequences</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    Creating partition table</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">    Writing results to svd nexus file /tmp/RtmpZ70Nc1/svd.nex </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl2svdquartets </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

