<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Converts a genlight object into vcf format — gl2vcf • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Converts a genlight object into vcf format — gl2vcf"><meta name="description" content="This function exports a genlight object into VCF format and save it into a
file."><meta property="og:description" content="This function exports a genlight object into VCF format and save it into a
file."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.9.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/green-striped-gecko/dartR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Converts a genlight object into vcf format</h1>
      <small class="dont-index">Source: <a href="https://github.com/green-striped-gecko/dartR/blob/master/R/gl2vcf.r" class="external-link"><code>R/gl2vcf.r</code></a></small>
      <div class="d-none name"><code>gl2vcf.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function exports a genlight object into VCF format and save it into a
file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl2vcf</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  plink_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  outfile <span class="op">=</span> <span class="st">"gl_vcf"</span>,</span>
<span>  outpath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  snp_pos <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>  snp_chr <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>  chr_format <span class="op">=</span> <span class="st">"character"</span>,</span>
<span>  pos_cM <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>  ID_dad <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>  ID_mom <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>  sex_code <span class="op">=</span> <span class="st">"unknown"</span>,</span>
<span>  phen_value <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt id="arg-plink-path">plink_path<a class="anchor" aria-label="anchor" href="#arg-plink-path"></a></dt>
<dd><p>Path of PLINK binary file [default getwd())].</p></dd>


<dt id="arg-outfile">outfile<a class="anchor" aria-label="anchor" href="#arg-outfile"></a></dt>
<dd><p>File name of the output file [default 'gl_vcf'].</p></dd>


<dt id="arg-outpath">outpath<a class="anchor" aria-label="anchor" href="#arg-outpath"></a></dt>
<dd><p>Path where to save the output file
[default tempdir(), mandated by CRAN]. Use outpath=getwd() or outpath='.'
 when calling this function to direct output files to your working directory.</p></dd>


<dt id="arg-snp-pos">snp_pos<a class="anchor" aria-label="anchor" href="#arg-snp-pos"></a></dt>
<dd><p>Field name from the slot loc.metrics where the SNP position is
stored [default '0'].</p></dd>


<dt id="arg-snp-chr">snp_chr<a class="anchor" aria-label="anchor" href="#arg-snp-chr"></a></dt>
<dd><p>Field name from the slot loc.metrics where the chromosome of
each is stored [default '0'].</p></dd>


<dt id="arg-chr-format">chr_format<a class="anchor" aria-label="anchor" href="#arg-chr-format"></a></dt>
<dd><p>Whether chromosome information is stored as 'numeric' or as
'character', see details [default 'character'].</p></dd>


<dt id="arg-pos-cm">pos_cM<a class="anchor" aria-label="anchor" href="#arg-pos-cm"></a></dt>
<dd><p>A vector, with as many elements as there are loci, containing
the SNP position in morgans or centimorgans [default '0'].</p></dd>


<dt id="arg-id-dad">ID_dad<a class="anchor" aria-label="anchor" href="#arg-id-dad"></a></dt>
<dd><p>A vector, with as many elements as there are individuals,
containing the ID of the father, '0' if father isn't in dataset [default '0'].</p></dd>


<dt id="arg-id-mom">ID_mom<a class="anchor" aria-label="anchor" href="#arg-id-mom"></a></dt>
<dd><p>A vector, with as many elements as there are individuals,
containing the ID of the mother, '0' if mother isn't in dataset [default '0'].</p></dd>


<dt id="arg-sex-code">sex_code<a class="anchor" aria-label="anchor" href="#arg-sex-code"></a></dt>
<dd><p>A vector, with as many elements as there are individuals,
containing the sex code ('male', 'female', 'unknown') [default  'unknown'].</p></dd>


<dt id="arg-phen-value">phen_value<a class="anchor" aria-label="anchor" href="#arg-phen-value"></a></dt>
<dd><p>A vector, with as many elements as there are individuals,
containing the phenotype value. '1' = control, '2' = case, '0' = unknown
[default '0'].</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default 2 or as specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>returns no value (i.e. NULL)</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function requires to download the binary file of PLINK 1.9 and provide
its path (plink_path).
The binary file can be downloaded from:
<a href="https://www.cog-genomics.org/plink/" class="external-link">https://www.cog-genomics.org/plink/</a></p>
<p>The chromosome information for unmapped SNPS is coded as 0.
Family ID is taken from  x$pop
Within-family ID (cannot be '0') is taken from indNames(x)
Variant identifier is taken from locNames(x)</p>
<p>#' Note that if names of populations or individuals contain spaces, they are
replaced by an underscore "_".</p>
<p>If you like to use chromosome information when converting to plink format and
 your chromosome names are not from human, you need to change the chromosome
 names as 'contig1', 'contig2', etc. as described in the section "Nonstandard
  chromosome IDs" in the following link:
  https://www.cog-genomics.org/plink/1.9/input</p>
<p>Note that the function might not work if there are spaces in the path to the
  plink executable.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Danecek, P., Auton, A., Abecasis, G., Albers, C. A., Banks, E., DePristo, M.
 A., ... &amp; 1000 Genomes Project Analysis Group. (2011). The variant call
 format and VCFtools. Bioinformatics, 27(15), 2156-2158.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Luis Mijangos (Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/green-striped-gecko/dartR.data" class="external-link">"dartR.data"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">gl2vcf</span><span class="op">(</span><span class="va">platypus.gl</span>,snp_pos<span class="op">=</span><span class="st">'ChromPos_Platypus_Chrom_NCBIv1'</span>,</span></span>
<span class="r-in"><span> snp_chr <span class="op">=</span> <span class="st">'Chrom_Platypus_Chrom_NCBIv1'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni, Diana Robledo-Ruiz.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

