<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ dartR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dartR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/dartRTutorials.html">dartR Tutorials</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="bandicoot.gl.html">bandicoot.gl</a></code> </p>
        </td>
        <td><p>A genlight object created via the read.dart functions</p></td>
      </tr><tr><td>
          <p><code><a href="cbind.dartR.html">cbind(<i>&lt;dartR&gt;</i>)</a></code> </p>
        </td>
        <td><p>adjust cbind for dartR</p></td>
      </tr><tr><td>
          <p><code><a href="gi2gl.html">gi2gl()</a></code> </p>
        </td>
        <td><p>Converts a genind object into a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.He.html">gl.He()</a></code> </p>
        </td>
        <td><p>Estimates expected Heterozygosity</p></td>
      </tr><tr><td>
          <p><code><a href="gl.Ho.html">gl.Ho()</a></code> </p>
        </td>
        <td><p>Estimates observed Heterozygosity</p></td>
      </tr><tr><td>
          <p><code><a href="gl.LDNe.html">gl.LDNe()</a></code> </p>
        </td>
        <td><p>Estimates effective population size using the Linkage Disequilibrium
method based on NeEstimator (V2)</p></td>
      </tr><tr><td>
          <p><code><a href="gl.alf.html">gl.alf()</a></code> </p>
        </td>
        <td><p>Calculates allele frequency of the first and second allele for each loci
A very simple function to report allele frequencies</p></td>
      </tr><tr><td>
          <p><code><a href="gl.amova.html">gl.amova()</a></code> </p>
        </td>
        <td><p>Performs AMOVA using genlight data</p></td>
      </tr><tr><td>
          <p><code><a href="gl.assign.grm.html">gl.assign.grm()</a></code> </p>
        </td>
        <td><p>Population assignment using grm</p></td>
      </tr><tr><td>
          <p><code><a href="gl.assign.mahalanobis.html">gl.assign.mahalanobis()</a></code> </p>
        </td>
        <td><p>Assign an individual of unknown provenance to population based on 
Mahalanobis Distance</p></td>
      </tr><tr><td>
          <p><code><a href="gl.assign.pa.html">gl.assign.pa()</a></code> </p>
        </td>
        <td><p>Eliminates populations as possible source populations for an 
individual of unknown provenance, using private alleles</p></td>
      </tr><tr><td>
          <p><code><a href="gl.assign.pca.html">gl.assign.pca()</a></code> </p>
        </td>
        <td><p>Assign an individual of unknown provenance to population based on PCA</p></td>
      </tr><tr><td>
          <p><code><a href="gl.basic.stats.html">gl.basic.stats()</a></code> </p>
        </td>
        <td><p>Calculates basic statistics for each loci (Hs, Ho, Fis etc.)</p></td>
      </tr><tr><td>
          <p><code><a href="gl.blast.html">gl.blast()</a></code> </p>
        </td>
        <td><p>Aligns nucleotides sequences against those present in a target database 
using blastn</p></td>
      </tr><tr><td>
          <p><code><a href="gl.check.verbosity.html">gl.check.verbosity()</a></code> </p>
        </td>
        <td><p>Checks the current global verbosity</p></td>
      </tr><tr><td>
          <p><code><a href="gl.collapse.html">gl.collapse()</a></code> </p>
        </td>
        <td><p>Collapses a distance matrix by amalgamating populations with pairwise
 fixed difference count less that a threshold</p></td>
      </tr><tr><td>
          <p><code><a href="gl.compliance.check.html">gl.compliance.check()</a></code> </p>
        </td>
        <td><p>Checks a genlight object to see if it complies with dartR
 expectations and amends it to comply if necessary</p></td>
      </tr><tr><td>
          <p><code><a href="gl.costdistances.html">gl.costdistances()</a></code> </p>
        </td>
        <td><p>Calculates cost distances for a given landscape (resistance matrix)</p></td>
      </tr><tr><td>
          <p><code><a href="gl.define.pop.html">gl.define.pop()</a></code> </p>
        </td>
        <td><p>Defines a new population in a genlight object for specified 
individuals</p></td>
      </tr><tr><td>
          <p><code><a href="gl.diagnostics.hwe.html">gl.diagnostics.hwe()</a></code> </p>
        </td>
        <td><p>Provides descriptive stats and plots to diagnose potential problems
  with Hardy-Weinberg proportions</p></td>
      </tr><tr><td>
          <p><code><a href="gl.diagnostics.sim.html">gl.diagnostics.sim()</a></code> </p>
        </td>
        <td><p>Comparing simulations against theoretical expectations</p></td>
      </tr><tr><td>
          <p><code><a href="gl.dist.ind.html">gl.dist.ind()</a></code> </p>
        </td>
        <td><p>Calculates a distance matrix for individuals defined in a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.dist.pop.html">gl.dist.pop()</a></code> </p>
        </td>
        <td><p>Calculates a distance matrix for populations with SNP genotypes in a
 genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.drop.ind.html">gl.drop.ind()</a></code> </p>
        </td>
        <td><p>Removes specified individuals from a genlight {adegenet} object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.drop.loc.html">gl.drop.loc()</a></code> </p>
        </td>
        <td><p>Removes specified loci from a genlight {adegenet} object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.drop.pop.html">gl.drop.pop()</a></code> </p>
        </td>
        <td><p>Removes specified populations from a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.edit.recode.ind.html">gl.edit.recode.ind()</a></code> </p>
        </td>
        <td><p>Creates or edits individual (=specimen) names, creates a recode_ind
 file and applies the changes to a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.edit.recode.pop.html">gl.edit.recode.pop()</a></code> </p>
        </td>
        <td><p>Creates or edits a population re-assignment table</p></td>
      </tr><tr><td>
          <p><code><a href="gl.evanno.html">gl.evanno()</a></code> </p>
        </td>
        <td><p>Creates an Evanno plot from a STRUCTURE run object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.fdsim.html">gl.fdsim()</a></code> </p>
        </td>
        <td><p>Estimates the rate of false positives in a fixed difference analysis</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.allna.html">gl.filter.allna()</a></code> </p>
        </td>
        <td><p>Filters loci that are all NA across individuals and/or populations 
with all NA across loci</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.callrate.html">gl.filter.callrate()</a></code> </p>
        </td>
        <td><p>Filters loci or specimens in a genlight {adegenet} object based on
call rate</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.hamming.html">gl.filter.hamming()</a></code> </p>
        </td>
        <td><p>Filters loci based on pairwise Hamming distance between sequence tags</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.heterozygosity.html">gl.filter.heterozygosity()</a></code> </p>
        </td>
        <td><p>Filters individuals with average heterozygosity greater than a
 specified upper threshold or less than a specified lower threshold</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.hwe.html">gl.filter.hwe()</a></code> </p>
        </td>
        <td><p>Filters loci that show significant departure from Hardy-Weinberg
 Equilibrium</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.ld.html">gl.filter.ld()</a></code> </p>
        </td>
        <td><p>Filters loci based on linkage disequilibrium (LD)</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.locmetric.html">gl.filter.locmetric()</a></code> </p>
        </td>
        <td><p>Filters loci on the basis of numeric information stored in
 other$loc.metrics in a genlight {adegenet} object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.maf.html">gl.filter.maf()</a></code> </p>
        </td>
        <td><p>Filters loci on the basis of minor allele frequency (MAF) in a 
genlight
 adegenet object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.monomorphs.html">gl.filter.monomorphs()</a></code> </p>
        </td>
        <td><p>Filters monomorphic loci, including those with all NAs</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.overshoot.html">gl.filter.overshoot()</a></code> </p>
        </td>
        <td><p>Filters loci for which the SNP has been trimmed from the sequence tag
along with the adaptor</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.pa.html">gl.filter.pa()</a></code> </p>
        </td>
        <td><p>Filters loci that contain private (and fixed alleles) between two populations</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.parent.offspring.html">gl.filter.parent.offspring()</a></code> </p>
        </td>
        <td><p>Filters putative parent offspring within a population</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.rdepth.html">gl.filter.rdepth()</a></code> </p>
        </td>
        <td><p>Filters loci based on counts of sequence tags scored at a locus (read
 depth)</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.reproducibility.html">gl.filter.reproducibility()</a></code> </p>
        </td>
        <td><p>Filters loci in a genlight {adegenet} object based on average
 repeatability of alleles at a locus</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.secondaries.html">gl.filter.secondaries()</a></code> </p>
        </td>
        <td><p>Filters loci that represent secondary SNPs in a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.sexlinked.html">gl.filter.sexlinked()</a></code> </p>
        </td>
        <td><p>Filters loci that are sex linked</p></td>
      </tr><tr><td>
          <p><code><a href="gl.filter.taglength.html">gl.filter.taglength()</a></code> </p>
        </td>
        <td><p>Filters loci in a genlight {adegenet} object based on sequence tag
 length</p></td>
      </tr><tr><td>
          <p><code><a href="gl.fixed.diff.html">gl.fixed.diff()</a></code> </p>
        </td>
        <td><p>Generates a matrix of fixed differences and associated statistics for
 populations taken pairwise</p></td>
      </tr><tr><td>
          <p><code><a href="gl.fst.pop.html">gl.fst.pop()</a></code> </p>
        </td>
        <td><p>Calculates a pairwise Fst values for populations in a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.genleastcost.html">gl.genleastcost()</a></code> </p>
        </td>
        <td><p>Performs least-cost path analysis based on a friction matrix</p></td>
      </tr><tr><td>
          <p><code><a href="gl.grm.html">gl.grm()</a></code> </p>
        </td>
        <td><p>Calculates an identity by descent matrix</p></td>
      </tr><tr><td>
          <p><code><a href="gl.grm.network.html">gl.grm.network()</a></code> </p>
        </td>
        <td><p>Represents a genomic relationship matrix (GRM) as a network</p></td>
      </tr><tr><td>
          <p><code><a href="gl.hwe.pop.html">gl.hwe.pop()</a></code> </p>
        </td>
        <td><p>Performs Hardy-Weinberg tests over loci and populations</p></td>
      </tr><tr><td>
          <p><code><a href="gl.ibd.html">gl.ibd()</a></code> </p>
        </td>
        <td><p>Performs isolation by distance analysis</p></td>
      </tr><tr><td>
          <p><code><a href="gl.impute.html">gl.impute()</a></code> </p>
        </td>
        <td><p>Imputates missing data</p></td>
      </tr><tr><td>
          <p><code><a href="gl.install.vanilla.dartR.html">gl.install.vanilla.dartR()</a></code> </p>
        </td>
        <td><p>Installs all required packages for using all functions
available in dartR</p></td>
      </tr><tr><td>
          <p><code><a href="gl.join.html">gl.join()</a></code> </p>
        </td>
        <td><p>Combines two genlight objects</p></td>
      </tr><tr><td>
          <p><code><a href="gl.keep.ind.html">gl.keep.ind()</a></code> </p>
        </td>
        <td><p>Removes all but the specified individuals from a genlight {adegenet}
object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.keep.loc.html">gl.keep.loc()</a></code> </p>
        </td>
        <td><p>Removes all but the specified loci from a genlight {adegenet} object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.keep.pop.html">gl.keep.pop()</a></code> </p>
        </td>
        <td><p>Removes all but the specified populations from a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.ld.distance.html">gl.ld.distance()</a></code> </p>
        </td>
        <td><p>Plots linkage disequilibrium against distance by population 
disequilibrium patterns</p></td>
      </tr><tr><td>
          <p><code><a href="gl.ld.haplotype.html">gl.ld.haplotype()</a></code> </p>
        </td>
        <td><p>Visualize patterns of linkage disequilibrium and identification of 
haplotypes</p></td>
      </tr><tr><td>
          <p><code><a href="gl.list.reports.html">gl.list.reports()</a></code> </p>
        </td>
        <td><p>Prints dartR reports saved in tempdir</p></td>
      </tr><tr><td>
          <p><code><a href="gl.load.html">gl.load()</a></code> </p>
        </td>
        <td><p>Loads an object from compressed binary format produced by gl.save()</p></td>
      </tr><tr><td>
          <p><code><a href="gl.make.recode.ind.html">gl.make.recode.ind()</a></code> </p>
        </td>
        <td><p>Creates a proforma recode_ind file for reassigning individual
 (=specimen) names</p></td>
      </tr><tr><td>
          <p><code><a href="gl.make.recode.pop.html">gl.make.recode.pop()</a></code> </p>
        </td>
        <td><p>Creates a proforma recode_pop_table file for reassigning population
 names</p></td>
      </tr><tr><td>
          <p><code><a href="gl.map.interactive.html">gl.map.interactive()</a></code> </p>
        </td>
        <td><p>Creates an interactive map (based on latlon) from a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.map.structure.html">gl.map.structure()</a></code> </p>
        </td>
        <td><p>Maps a STRUCTURE plot using a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.merge.pop.html">gl.merge.pop()</a></code> </p>
        </td>
        <td><p>Merges two or more populations in a genlight object into one population</p></td>
      </tr><tr><td>
          <p><code><a href="gl.nhybrids.html">gl.nhybrids()</a></code> </p>
        </td>
        <td><p>Creates an input file for the program NewHybrids and runs it if 
NewHybrids is installed</p></td>
      </tr><tr><td>
          <p><code><a href="gl.outflank.html">gl.outflank()</a></code> </p>
        </td>
        <td><p>Identifies loci under selection per population using the outflank
 method of Whitlock and Lotterhos (2015)</p></td>
      </tr><tr><td>
          <p><code><a href="gl.pcoa.html">gl.pcoa()</a></code> </p>
        </td>
        <td><p>Ordination applied to genotypes in a genlight object (PCA), in an fd
object, or to a distance matrix (PCoA)</p></td>
      </tr><tr><td>
          <p><code><a href="gl.pcoa.plot.html">gl.pcoa.plot()</a></code> </p>
        </td>
        <td><p>Bivariate or trivariate plot of the results of an ordination generated
 using gl.pcoa()</p></td>
      </tr><tr><td>
          <p><code><a href="gl.percent.freq.html">gl.percent.freq()</a></code> </p>
        </td>
        <td><p>Generates percentage allele frequencies by locus and population</p></td>
      </tr><tr><td>
          <p><code><a href="gl.play.history.html">gl.play.history()</a></code> </p>
        </td>
        <td><p>Replays the history and applies it to a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.plot.heatmap.html">gl.plot.heatmap()</a></code> </p>
        </td>
        <td><p>Represents a distance matrix as a heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="gl.plot.network.html">gl.plot.network()</a></code> </p>
        </td>
        <td><p>Represents a distance or dissimilarity matrix as a network</p></td>
      </tr><tr><td>
          <p><code><a href="gl.plot.structure.html">gl.plot.structure()</a></code> </p>
        </td>
        <td><p>Plots STRUCTURE analysis results (Q-matrix)</p></td>
      </tr><tr><td>
          <p><code><a href="gl.print.history.html">gl.print.history()</a></code> </p>
        </td>
        <td><p>Prints history of a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.print.reports.html">gl.print.reports()</a></code> </p>
        </td>
        <td><p>Prints dartR reports saved in tempdir</p></td>
      </tr><tr><td>
          <p><code><a href="gl.propShared.html">gl.propShared()</a></code> </p>
        </td>
        <td><p>Calculates a similarity (distance) matrix for individuals on the proportion of
 shared alleles</p></td>
      </tr><tr><td>
          <p><code><a href="gl.random.snp.html">gl.random.snp()</a></code> </p>
        </td>
        <td><p>Randomly changes the allocation of 0's and 2's in a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.read.csv.html">gl.read.csv()</a></code> </p>
        </td>
        <td><p>Reads SNP data from a csv file into a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.read.dart.html">gl.read.dart()</a></code> </p>
        </td>
        <td><p>Imports DArT data into dartR and converts it into a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.read.fasta.html">gl.read.fasta()</a></code> </p>
        </td>
        <td><p>Reads FASTA files and converts them to genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.read.silicodart.html">gl.read.silicodart()</a></code> </p>
        </td>
        <td><p>Imports presence/absence data from SilicoDArT to genlight {agegenet}
format (ploidy=1)</p></td>
      </tr><tr><td>
          <p><code><a href="gl.read.vcf.html">gl.read.vcf()</a></code> </p>
        </td>
        <td><p>Converts a vcf file into a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.reassign.pop.html">gl.reassign.pop()</a></code> </p>
        </td>
        <td><p>Assigns an individual metric as pop in a genlight {adegenet} object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.recalc.metrics.html">gl.recalc.metrics()</a></code> </p>
        </td>
        <td><p>Recalculates locus metrics when individuals or populations are deleted from a
 genlight {adegenet} object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.recode.ind.html">gl.recode.ind()</a></code> </p>
        </td>
        <td><p>Recodes individual (=specimen = sample) labels in a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.recode.pop.html">gl.recode.pop()</a></code> </p>
        </td>
        <td><p>Recodes population assignments in a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.rename.pop.html">gl.rename.pop()</a></code> </p>
        </td>
        <td><p>Renames a population in a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.bases.html">gl.report.bases()</a></code> </p>
        </td>
        <td><p>Reports summary of base pair frequencies</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.callrate.html">gl.report.callrate()</a></code> </p>
        </td>
        <td><p>Reports summary of Call Rate for loci or individuals</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.diversity.html">gl.report.diversity()</a></code> </p>
        </td>
        <td><p>Calculates diversity indexes for SNPs</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.hamming.html">gl.report.hamming()</a></code> </p>
        </td>
        <td><p>Calculates the pairwise Hamming distance between DArT trimmed DNA
sequences</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.heterozygosity.html">gl.report.heterozygosity()</a></code> </p>
        </td>
        <td><p>Reports observed, expected and unbiased heterozygosities and FIS
(inbreeding coefficient) by population or by individual from SNP data</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.hwe.html">gl.report.hwe()</a></code> </p>
        </td>
        <td><p>Reports departure from Hardy-Weinberg proportions</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.ld.html">gl.report.ld()</a></code> </p>
        </td>
        <td><p>Calculates pairwise population based Linkage Disequilibrium across all loci
 using the specified number of cores</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.ld.map.html">gl.report.ld.map()</a></code> </p>
        </td>
        <td><p>Calculates pairwise linkage disequilibrium by population</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.locmetric.html">gl.report.locmetric()</a></code> </p>
        </td>
        <td><p>Reports summary of the slot $other$loc.metrics</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.maf.html">gl.report.maf()</a></code> </p>
        </td>
        <td><p>Reports minor allele frequency (MAF) for each locus in a SNP dataset</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.monomorphs.html">gl.report.monomorphs()</a></code> </p>
        </td>
        <td><p>Reports monomorphic loci</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.overshoot.html">gl.report.overshoot()</a></code> </p>
        </td>
        <td><p>Reports loci for which the SNP has been trimmed from the sequence tag
 along with the adaptor</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.pa.html">gl.report.pa()</a></code> </p>
        </td>
        <td><p>Reports private alleles (and fixed alleles) per pair of populations</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.parent.offspring.html">gl.report.parent.offspring()</a></code> </p>
        </td>
        <td><p>Identifies putative parent offspring within a population</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.rdepth.html">gl.report.rdepth()</a></code> </p>
        </td>
        <td><p>Reports summary of Read Depth for each locus</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.reproducibility.html">gl.report.reproducibility()</a></code> </p>
        </td>
        <td><p>Reports summary of RepAvg (repeatability averaged over both alleles 
for
each locus) or reproducibility (repeatability of the scores for fragment
presence/absence)</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.secondaries.html">gl.report.secondaries()</a></code> </p>
        </td>
        <td><p>Reports loci containing secondary SNPs in sequence tags and calculates
  number of invariant sites</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.sexlinked.html">gl.report.sexlinked()</a></code> </p>
        </td>
        <td><p>Identifies loci that are sex linked</p></td>
      </tr><tr><td>
          <p><code><a href="gl.report.taglength.html">gl.report.taglength()</a></code> </p>
        </td>
        <td><p>Reports summary of sequence tag length across loci</p></td>
      </tr><tr><td>
          <p><code><a href="gl.run.structure.html">gl.run.structure()</a></code> </p>
        </td>
        <td><p>Runs a STRUCTURE analysis using a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.save.html">gl.save()</a></code> </p>
        </td>
        <td><p>Saves an object in compressed binary format for later rapid retrieval</p></td>
      </tr><tr><td>
          <p><code><a href="gl.select.colors.html">gl.select.colors()</a></code> </p>
        </td>
        <td><p>Selects colors from one of several palettes and output as a vector</p></td>
      </tr><tr><td>
          <p><code><a href="gl.select.shapes.html">gl.select.shapes()</a></code> </p>
        </td>
        <td><p>Selects shapes from the base R shape palette and outputs as a vector</p></td>
      </tr><tr><td>
          <p><code><a href="gl.set.verbosity.html">gl.set.verbosity()</a></code> </p>
        </td>
        <td><p>Sets the default verbosity level</p></td>
      </tr><tr><td>
          <p><code><a href="gl.sfs.html">gl.sfs()</a></code> </p>
        </td>
        <td><p>Creates a site frequency spectrum based on a dartR or genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.sim.WF.run.html">gl.sim.WF.run()</a></code> </p>
        </td>
        <td><p>Runs Wright-Fisher simulations</p></td>
      </tr><tr><td>
          <p><code><a href="gl.sim.WF.table.html">gl.sim.WF.table()</a></code> </p>
        </td>
        <td><p>Creates the reference table for running gl.sim.WF.run</p></td>
      </tr><tr><td>
          <p><code><a href="gl.sim.create_dispersal.html">gl.sim.create_dispersal()</a></code> </p>
        </td>
        <td><p>Creates a dispersal file as input for the function gl.sim.WF.run</p></td>
      </tr><tr><td>
          <p><code><a href="gl.sim.emigration.html">gl.sim.emigration()</a></code> </p>
        </td>
        <td><p>Simulates emigration between populations</p></td>
      </tr><tr><td>
          <p><code><a href="gl.sim.ind.html">gl.sim.ind()</a></code> </p>
        </td>
        <td><p>Simulates individuals based on the allele frequencies provided via a genlight
object.</p></td>
      </tr><tr><td>
          <p><code><a href="gl.sim.mutate.html">gl.sim.mutate()</a></code> </p>
        </td>
        <td><p>Simulates mutations within a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.sim.offspring.html">gl.sim.offspring()</a></code> </p>
        </td>
        <td><p>Simulates a specified number of offspring based on alleles provided by
 potential father(s) and mother(s)</p></td>
      </tr><tr><td>
          <p><code><a href="gl.smearplot.html">gl.smearplot()</a></code> </p>
        </td>
        <td><p>Smear plot of SNP or presence/absence (SilicoDArT) data</p></td>
      </tr><tr><td>
          <p><code><a href="gl.spatial.autoCorr.html">gl.spatial.autoCorr()</a></code> </p>
        </td>
        <td><p>Spatial autocorrelation following Smouse and Peakall 1999</p></td>
      </tr><tr><td>
          <p><code><a href="gl.subsample.loci.html">gl.subsample.loci()</a></code> </p>
        </td>
        <td><p>Subsamples n loci from a genlight object and return it as a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="gl.test.heterozygosity.html">gl.test.heterozygosity()</a></code> </p>
        </td>
        <td><p>Tests the difference in heterozygosity between populations taken
 pairwise</p></td>
      </tr><tr><td>
          <p><code><a href="gl.tree.nj.html">gl.tree.nj()</a></code> </p>
        </td>
        <td><p>Outputs an nj tree to summarize genetic similarity among populations</p></td>
      </tr><tr><td>
          <p><code><a href="gl.write.csv.html">gl.write.csv()</a></code> </p>
        </td>
        <td><p>Writes out data from a genlight object to csv file</p></td>
      </tr><tr><td>
          <p><code><a href="gl2bayescan.html">gl2bayescan()</a></code> </p>
        </td>
        <td><p>Converts a genlight object into a format suitable for input to Bayescan</p></td>
      </tr><tr><td>
          <p><code><a href="gl2demerelate.html">gl2demerelate()</a></code> </p>
        </td>
        <td><p>Creates a dataframe suitable for input to package {Demerelate} from a
genlight {adegenet} object</p></td>
      </tr><tr><td>
          <p><code><a href="gl2eigenstrat.html">gl2eigenstrat()</a></code> </p>
        </td>
        <td><p>Converts a genlight object into eigenstrat format</p></td>
      </tr><tr><td>
          <p><code><a href="gl2fasta.html">gl2fasta()</a></code> </p>
        </td>
        <td><p>Concatenates DArT trimmed sequences and outputs a FASTA file</p></td>
      </tr><tr><td>
          <p><code><a href="gl2faststructure.html">gl2faststructure()</a></code> </p>
        </td>
        <td><p>Converts a genlight object into faststructure format (to run faststructure
 elsewhere)</p></td>
      </tr><tr><td>
          <p><code><a href="gl2gds.html">gl2gds()</a></code> </p>
        </td>
        <td><p>Converts a genlight object into gds format</p></td>
      </tr><tr><td>
          <p><code><a href="gl2genalex.html">gl2genalex()</a></code> </p>
        </td>
        <td><p>Converts a genlight object into a format suitable for input to genalex</p></td>
      </tr><tr><td>
          <p><code><a href="gl2genepop.html">gl2genepop()</a></code> </p>
        </td>
        <td><p>Converts a genlight object into genepop format (and file)</p></td>
      </tr><tr><td>
          <p><code><a href="gl2geno.html">gl2geno()</a></code> </p>
        </td>
        <td><p>Converts a genlight object to geno format from package LEA</p></td>
      </tr><tr><td>
          <p><code><a href="gl2gi.html">gl2gi()</a></code> </p>
        </td>
        <td><p>Converts a genlight object to genind object</p></td>
      </tr><tr><td>
          <p><code><a href="gl2hiphop.html">gl2hiphop()</a></code> </p>
        </td>
        <td><p>Converts a genlight objects into hiphop format</p></td>
      </tr><tr><td>
          <p><code><a href="gl2phylip.html">gl2phylip()</a></code> </p>
        </td>
        <td><p>Creates a Phylip input distance matrix from a genlight (SNP) {adegenet}
 object</p></td>
      </tr><tr><td>
          <p><code><a href="gl2plink.html">gl2plink()</a></code> </p>
        </td>
        <td><p>Converts a genlight object into PLINK format</p></td>
      </tr><tr><td>
          <p><code><a href="gl2related.html">gl2related()</a></code> </p>
        </td>
        <td><p>Converts a genlight object to format suitable to be run with Coancestry</p></td>
      </tr><tr><td>
          <p><code><a href="gl2sa.html">gl2sa()</a></code> </p>
        </td>
        <td><p>Converts genlight objects to the format used in the SNPassoc package</p></td>
      </tr><tr><td>
          <p><code><a href="gl2sfs.html">gl2sfs()</a></code> </p>
        </td>
        <td><p>Converts a genlight object into a sfs input file</p></td>
      </tr><tr><td>
          <p><code><a href="gl2shp.html">gl2shp()</a></code> </p>
        </td>
        <td><p>Converts a genlight object to ESRI shapefiles or kml files</p></td>
      </tr><tr><td>
          <p><code><a href="gl2snapp.html">gl2snapp()</a></code> </p>
        </td>
        <td><p>Converts a genlight object to nexus format suitable for phylogenetic analysis
 by SNAPP (via BEAUti)</p></td>
      </tr><tr><td>
          <p><code><a href="gl2structure.html">gl2structure()</a></code> </p>
        </td>
        <td><p>Converts a genlight object to STRUCTURE formated files</p></td>
      </tr><tr><td>
          <p><code><a href="gl2svdquartets.html">gl2svdquartets()</a></code> </p>
        </td>
        <td><p>Converts a genlight object to nexus format PAUP SVDquartets</p></td>
      </tr><tr><td>
          <p><code><a href="gl2treemix.html">gl2treemix()</a></code> </p>
        </td>
        <td><p>Converts a genlight object to a treemix input file</p></td>
      </tr><tr><td>
          <p><code><a href="gl2vcf.html">gl2vcf()</a></code> </p>
        </td>
        <td><p>Converts a genlight object into vcf format</p></td>
      </tr><tr><td>
          <p><code><a href="interactive_reference.html">interactive_reference()</a></code> </p>
        </td>
        <td><p>Shiny app for the input of the reference table for the simulations</p></td>
      </tr><tr><td>
          <p><code><a href="interactive_sim_run.html">interactive_sim_run()</a></code> </p>
        </td>
        <td><p>Shiny app for the input of the simulations variables</p></td>
      </tr><tr><td>
          <p><code><a href="is.fixed.html">is.fixed()</a></code> </p>
        </td>
        <td><p>Tests if two populations are fixed at a given locus</p></td>
      </tr><tr><td>
          <p><code><a href="platy.html">platy</a></code> </p>
        </td>
        <td><p>Example data set as text file to be imported into a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="possums.gl.html">possums.gl</a></code> </p>
        </td>
        <td><p>A simulated genlight object created to run a landscape genetic example</p></td>
      </tr><tr><td>
          <p><code><a href="rbind.dartR.html">rbind(<i>&lt;dartR&gt;</i>)</a></code> </p>
        </td>
        <td><p>adjust rbind for dartR</p></td>
      </tr><tr><td>
          <p><code><a href="sub-dartR-ANY-ANY-ANY-method.html">`[`(<i>&lt;dartR&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>)</a></code> </p>
        </td>
        <td><p>indexing dartR objects correctly...</p></td>
      </tr><tr><td>
          <p><code><a href="testset.gl.html">testset.gl</a></code> </p>
        </td>
        <td><p>A genlight object created via the gl.read.dart function</p></td>
      </tr><tr><td>
          <p><code><a href="testset.gs.html">testset.gs</a></code> </p>
        </td>
        <td><p>A genlight object created via the gl.read.silicodart function</p></td>
      </tr><tr><td>
          <p><code><a href="testset_SNPs_2Row.html">testset_SNPs_2Row</a></code> </p>
        </td>
        <td><p>Testfile in DArT format (as provided by DArT)</p></td>
      </tr><tr><td>
          <p><code><a href="testset_metadata.html">testset_metadata</a></code> </p>
        </td>
        <td><p>Metadata file. Can be integrated via the dart2genlight function.</p></td>
      </tr><tr><td>
          <p><code><a href="testset_pop_recode.html">testset_pop_recode</a></code> </p>
        </td>
        <td><p>Recode file to be used with the function.</p></td>
      </tr><tr><td>
          <p><code><a href="theme_dartR.html">theme_dartR()</a></code> </p>
        </td>
        <td><p>dartR theme</p></td>
      </tr><tr><td>
          <p><code><a href="utils.assignment.html">utils.assignment()</a></code> </p>
        </td>
        <td><p>Population assignment probabilities</p></td>
      </tr><tr><td>
          <p><code><a href="utils.assignment_2.html">utils.assignment_2()</a></code> </p>
        </td>
        <td><p>Population assignment probabilities</p></td>
      </tr><tr><td>
          <p><code><a href="utils.assignment_3.html">utils.assignment_3()</a></code> </p>
        </td>
        <td><p>Population assignment probabilities</p></td>
      </tr><tr><td>
          <p><code><a href="utils.assignment_4.html">utils.assignment_4()</a></code> </p>
        </td>
        <td><p>Population assignment probabilities</p></td>
      </tr><tr><td>
          <p><code><a href="utils.basic.stats.html">utils.basic.stats()</a></code> </p>
        </td>
        <td><p>Calculates mean observed heterozygosity, mean expected heterozygosity and Fis
 per locus, per population and Fst across all populations</p></td>
      </tr><tr><td>
          <p><code><a href="utils.check.datatype.html">utils.check.datatype()</a></code> </p>
        </td>
        <td><p>Utility function to check the class of an object passed to a function</p></td>
      </tr><tr><td>
          <p><code><a href="utils.dart2genlight.html">utils.dart2genlight()</a></code> </p>
        </td>
        <td><p>Converts DarT to genlight</p></td>
      </tr><tr><td>
          <p><code><a href="utils.dist.binary.html">utils.dist.binary()</a></code> </p>
        </td>
        <td><p>Calculates a distance matrix for individuals defined in a dartR
genlight object using binary P/A data (SilicoDArT)</p></td>
      </tr><tr><td>
          <p><code><a href="utils.dist.ind.snp.html">utils.dist.ind.snp()</a></code> </p>
        </td>
        <td><p>Calculates a distance matrix for individuals defined in a dartR
genlight object using SNP data (DArTseq)</p></td>
      </tr><tr><td>
          <p><code><a href="utils.flag.start.html">utils.flag.start()</a></code> </p>
        </td>
        <td><p>A utility script to flag the start of a script</p></td>
      </tr><tr><td>
          <p><code><a href="utils.hamming.html">utils.hamming()</a></code> </p>
        </td>
        <td><p>Calculates the Hamming distance between two DArT trimmed DNA sequences</p></td>
      </tr><tr><td>
          <p><code><a href="utils.het.pop.html">utils.het.pop()</a></code> </p>
        </td>
        <td><p>Calculates expected mean expected heterozygosity per population</p></td>
      </tr><tr><td>
          <p><code><a href="utils.jackknife.html">utils.jackknife()</a></code> </p>
        </td>
        <td><p>Conducts jackknife resampling using a genlight object</p></td>
      </tr><tr><td>
          <p><code><a href="utils.n.var.invariant.html">utils.n.var.invariant()</a></code> </p>
        </td>
        <td><p>A utility script to calculate the number of variant and invariant sites by
locus</p></td>
      </tr><tr><td>
          <p><code><a href="utils.outflank.MakeDiploidFSTMat.html">utils.outflank.MakeDiploidFSTMat()</a></code> </p>
        </td>
        <td><p>Creates OutFLANK input file from individual genotype info.</p></td>
      </tr><tr><td>
          <p><code><a href="utils.outflank.html">utils.outflank()</a></code> </p>
        </td>
        <td><p>OutFLANK:  An Fst outlier approach by Mike Whitlock and Katie Lotterhos,
University of British Columbia.</p></td>
      </tr><tr><td>
          <p><code><a href="utils.outflank.plotter.html">utils.outflank.plotter()</a></code> </p>
        </td>
        <td><p>Plotting functions for Fst distributions after OutFLANK</p></td>
      </tr><tr><td>
          <p><code><a href="utils.read.dart.html">utils.read.dart()</a></code> </p>
        </td>
        <td><p>Imports DarT data to R</p></td>
      </tr><tr><td>
          <p><code><a href="utils.recalc.avgpic.html">utils.recalc.avgpic()</a></code> </p>
        </td>
        <td><p>A utility script to recalculate the OneRatioRef, OneRatioSnp, PICRef, PICSnp,
 and AvgPIC by locus after some individuals or populations have been deleted.</p></td>
      </tr><tr><td>
          <p><code><a href="utils.recalc.callrate.html">utils.recalc.callrate()</a></code> </p>
        </td>
        <td><p>A utility script to recalculate the callrate by locus after some populations
have been deleted</p></td>
      </tr><tr><td>
          <p><code><a href="utils.recalc.freqhets.html">utils.recalc.freqhets()</a></code> </p>
        </td>
        <td><p>A utility script to recalculate the frequency of the heterozygous SNPs by
locus after some populations have been deleted</p></td>
      </tr><tr><td>
          <p><code><a href="utils.recalc.freqhomref.html">utils.recalc.freqhomref()</a></code> </p>
        </td>
        <td><p>A utility script to recalculate the frequency of the homozygous reference
SNP by locus after some populations have been deleted</p></td>
      </tr><tr><td>
          <p><code><a href="utils.recalc.freqhomsnp.html">utils.recalc.freqhomsnp()</a></code> </p>
        </td>
        <td><p>A utility script to recalculate the frequency of the homozygous alternate
SNP by locus after some populations have been deleted</p></td>
      </tr><tr><td>
          <p><code><a href="utils.recalc.maf.html">utils.recalc.maf()</a></code> </p>
        </td>
        <td><p>A utility script to recalculate the minor allele frequency by locus,
typically after some populations have been deleted</p></td>
      </tr><tr><td>
          <p><code><a href="utils.reset.flags.html">utils.reset.flags()</a></code> </p>
        </td>
        <td><p>A utility script to reset to FALSE (or TRUE) the locus metric flags after
some individuals or populations have been deleted.</p></td>
      </tr><tr><td>
          <p><code><a href="utils.spautocorr.html">utils.spautocor()</a></code> </p>
        </td>
        <td><p>Spatial autocorrelation coefficient calculations</p></td>
      </tr><tr><td>
          <p><code><a href="utils.structure.evanno.html">utils.structure.evanno()</a></code> </p>
        </td>
        <td><p>Util function for evanno plots</p></td>
      </tr><tr><td>
          <p><code><a href="utils.structure.genind2gtypes.html">utils.structure.genind2gtypes()</a></code> </p>
        </td>
        <td><p>structure util functions</p></td>
      </tr><tr><td>
          <p><code><a href="utils.structure.run.html">utils.structure.run()</a></code> </p>
        </td>
        <td><p>Utility function to run Structure</p></td>
      </tr><tr><td>
          <p><code><a href="zzz.html">zzz</a></code> </p>
        </td>
        <td><p>Setting up the package</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

