<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function reports private alleles in one population compared with a
second population, for all populations taken pairwise. It also reports a
count of fixed allelic differences and the mean absolute allele frequency
differences (AFD) between pairs of populations."><title>Reports private alleles (and fixed alleles) per pair of populations — gl.report.pa • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reports private alleles (and fixed alleles) per pair of populations — gl.report.pa"><meta property="og:description" content="This function reports private alleles in one population compared with a
second population, for all populations taken pairwise. It also reports a
count of fixed allelic differences and the mean absolute allele frequency
differences (AFD) between pairs of populations."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Reports private alleles (and fixed alleles) per pair of populations</h1>
      
      <div class="d-none name"><code>gl.report.pa.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function reports private alleles in one population compared with a
second population, for all populations taken pairwise. It also reports a
count of fixed allelic differences and the mean absolute allele frequency
differences (AFD) between pairs of populations.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gl.report.pa</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  x2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"pairwise"</span>,</span>
<span>  plot.out <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  font_plot <span class="op">=</span> <span class="fl">14</span>,</span>
<span>  map.interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  palette_discrete <span class="op">=</span> <span class="va">discrete_palette</span>,</span>
<span>  save2tmp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>Name of the genlight object containing the SNP data [required].</p></dd>


<dt>x2</dt>
<dd><p>If two separate genlight objects are to be compared this can be
provided here, but they must have the same number of SNPs [default NULL].</p></dd>


<dt>method</dt>
<dd><p>Method to calculate private alleles: 'pairwise' comparison or
compare each population against the rest 'one2rest' [default 'pairwise'].</p></dd>


<dt>plot.out</dt>
<dd><p>Specify if Sankey plot is to be produced [default TRUE].</p></dd>


<dt>font_plot</dt>
<dd><p>Numeric font size in pixels for the node text labels
[default 14].</p></dd>


<dt>map.interactive</dt>
<dd><p>Specify whether an interactive map showing private
alleles between populations is to be produced [default FALSE].</p></dd>


<dt>palette_discrete</dt>
<dd><p>A discrete palette for the color of populations or a
list with as many colors as there are populations in the dataset
 [default discrete_palette].</p></dd>


<dt>save2tmp</dt>
<dd><p>If TRUE, saves any ggplots and listings to the session
temporary directory (tempdir) [default FALSE].</p></dd>


<dt>verbose</dt>
<dd><p>Verbosity: 0, silent, fatal errors only; 1, flag function
begin and end; 2, progress log; 3, progress and results summary; 5, full
report [default 2 or as specified using gl.set.verbosity].</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A data.frame. Each row shows, for each pair of populations the number
 of individuals in each population, the number of loci with fixed differences
 (same for both populations) in pop1 (compared to pop2) and vice versa. Same
 for private alleles and finally the absolute mean allele frequency
 difference between loci (AFD).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Note that the number of paired alleles between two populations is not a
symmetric dissimilarity measure.</p>
<p>If no x2 is provided, the function uses the pop(gl) hierarchy to determine
pairs of populations, otherwise it runs a single comparison between x and
x2.</p>
<p><strong>Hint:</strong> in case you want to run comparisons between individuals
(assuming individual names are unique), you can simply redefine your
population names with your individual names, as below:</p>
<p><code>pop(gl) &lt;- indNames(gl)</code></p>
<p><strong> Definition of fixed and private alleles </strong></p>
<p>The table below shows the possible cases of allele frequencies between
two populations (0 = homozygote for Allele 1, x = both Alleles are present,
1 = homozygote for Allele 2).</p>
<ul><li><p>p: cases where there is a private allele in pop1 compared to pop2 (but
not vice versa)</p></li>
<li><p>f: cases where there is a fixed allele in pop1 (and pop2, as those cases
are symmetric)</p></li>
</ul><table class="table table"><tr><td></td><td></td><td></td><td><em>pop1</em></td><td></td></tr><tr><td></td><td></td><td><strong>0</strong></td><td><strong>x</strong></td><td><strong>1</strong></td></tr><tr><td></td><td><strong>0</strong></td><td>-</td><td>p</td><td>p,f</td></tr><tr><td><em>pop2</em></td><td><strong>x</strong></td><td>-</td><td>-</td><td>-</td></tr><tr><td></td><td><strong>1</strong></td><td>p,f</td><td>p</td><td>-</td></tr></table><p>The absolute allele frequency difference (AFD) in this function is a simple
differentiation metric displaying intuitive properties which provides a
valuable alternative to FST. For details about its properties and how it is
calculated see Berner (2019).</p>
<p>In this function a Sankey Diagram is used to visualize patterns of private
alleles between populations. This diagram allows to display flows (private
alleles) between nodes (populations). Their links are represented with arcs
that have a width proportional to the importance of the flow (number of
private alleles).</p>
<p>if save2temp=TRUE, resultant plot(s) and the tabulation(s) are saved to the
session's temporary directory.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    
<ul><li><p>Berner, D. (2019). Allele frequency difference AFD – an intuitive
alternative to FST for quantifying genetic population differentiation. Genes,
 10(4), 308.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="gl.list.reports.html">gl.list.reports</a></code>,
 <code><a href="gl.print.reports.html">gl.print.reports</a></code></p>
<p>Other report functions: 
<code><a href="gl.report.bases.html">gl.report.bases</a>()</code>,
<code><a href="gl.report.callrate.html">gl.report.callrate</a>()</code>,
<code><a href="gl.report.diversity.html">gl.report.diversity</a>()</code>,
<code><a href="gl.report.hamming.html">gl.report.hamming</a>()</code>,
<code><a href="gl.report.heterozygosity.html">gl.report.heterozygosity</a>()</code>,
<code><a href="gl.report.hwe.html">gl.report.hwe</a>()</code>,
<code><a href="gl.report.ld.map.html">gl.report.ld.map</a>()</code>,
<code><a href="gl.report.locmetric.html">gl.report.locmetric</a>()</code>,
<code><a href="gl.report.maf.html">gl.report.maf</a>()</code>,
<code><a href="gl.report.monomorphs.html">gl.report.monomorphs</a>()</code>,
<code><a href="gl.report.overshoot.html">gl.report.overshoot</a>()</code>,
<code><a href="gl.report.parent.offspring.html">gl.report.parent.offspring</a>()</code>,
<code><a href="gl.report.rdepth.html">gl.report.rdepth</a>()</code>,
<code><a href="gl.report.reproducibility.html">gl.report.reproducibility</a>()</code>,
<code><a href="gl.report.secondaries.html">gl.report.secondaries</a>()</code>,
<code><a href="gl.report.sexlinked.html">gl.report.sexlinked</a>()</code>,
<code><a href="gl.report.taglength.html">gl.report.taglength</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Custodian: Bernd Gruber -- Post to
<a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">gl.report.pa</span><span class="op">(</span><span class="va">testset.gl</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,<span class="op">]</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.pa </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: data include loci that are scored NA across all individuals.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Consider filtering using gl &lt;- gl.filter.allna(gl)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">  Warning: no loci listed to keep! Genlight object returned unchanged</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    p1 p2          pop1          pop2 N1 N2 fixed priv1 priv2 Chao1 Chao2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   1  2 EmmacBurdMist EmmacBurnBara 10  1     0     7     1     7    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   1  3 EmmacBurdMist EmmacCoopEulb 10  5     4    12     6     7     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   1  4 EmmacBurdMist EmmacMaclGeor 10  1     0     8     2     7     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   1  5 EmmacBurdMist  EmmacMDBForb 10  1     2    10     4     7     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5   1  6 EmmacBurdMist  EmmacMDBMaci 10  1     1     8     2     7     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6   1  7 EmmacBurdMist  EmmacMDBSanf 10  1     3    11     3     7     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7   2  3 EmmacBurnBara EmmacCoopEulb  1  5     3     5     5     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8   2  4 EmmacBurnBara EmmacMaclGeor  1  1     1     3     3     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9   2  5 EmmacBurnBara  EmmacMDBForb  1  1     2     4     3     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10  2  6 EmmacBurnBara  EmmacMDBMaci  1  1     1     3     2     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11  2  7 EmmacBurnBara  EmmacMDBSanf  1  1     3     5     3     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12  3  4 EmmacCoopEulb EmmacMaclGeor  5  1     3     4     4     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13  3  5 EmmacCoopEulb  EmmacMDBForb  5  1     4     6     5     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14  3  6 EmmacCoopEulb  EmmacMDBMaci  5  1     4     5     4     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15  3  7 EmmacCoopEulb  EmmacMDBSanf  5  1     5     7     5     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16  4  5 EmmacMaclGeor  EmmacMDBForb  1  1     1     3     2     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17  4  6 EmmacMaclGeor  EmmacMDBMaci  1  1     1     1     1    NA    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18  4  7 EmmacMaclGeor  EmmacMDBSanf  1  1     2     4     2     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19  5  6  EmmacMDBForb  EmmacMDBMaci  1  1     0     2     1     0    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20  5  7  EmmacMDBForb  EmmacMDBSanf  1  1     1     3     1     0    NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21  6  7  EmmacMDBMaci  EmmacMDBSanf  1  1     0     1     0    NA  6817</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    totalpriv   AFD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          8 0.009</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2         18 0.026</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3         10 0.008</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         14 0.018</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         10 0.011</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6         14 0.016</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7         10 0.024</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8          6 0.014</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9          7 0.017</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10         5 0.013</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11         8 0.018</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12         8 0.018</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13        11 0.027</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14         9 0.024</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15        12 0.027</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16         5 0.012</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17         2 0.005</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18         6 0.013</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19         3 0.008</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20         4 0.009</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21         1 0.002</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Table of private alleles and fixed differences returned</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.pa </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

