<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This script uses any field with numeric values stored in $other$loc.metrics
to produce summary statistics (mean, minimum, average, quantiles), histograms
and boxplots to assist the decision of choosing thresholds for the filter
function gl.filter.locmetric."><title>Reports summary of the slot $other$loc.metrics — gl.report.locmetric • dartR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reports summary of the slot $other$loc.metrics — gl.report.locmetric"><meta property="og:description" content="This script uses any field with numeric values stored in $other$loc.metrics
to produce summary statistics (mean, minimum, average, quantiles), histograms
and boxplots to assist the decision of choosing thresholds for the filter
function gl.filter.locmetric."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dartR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/dartRTutorials.html">dartR Tutorials</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Reports summary of the slot $other$loc.metrics</h1>
      
      <div class="d-none name"><code>gl.report.locmetric.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This script uses any field with numeric values stored in $other$loc.metrics
to produce summary statistics (mean, minimum, average, quantiles), histograms
and boxplots to assist the decision of choosing thresholds for the filter
function <code><a href="gl.filter.locmetrics.html">gl.filter.locmetric</a></code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">gl.report.locmetric</span><span class="op">(</span>
  <span class="va">x</span>,
  <span class="va">metric</span>,
  plot.out <span class="op">=</span> <span class="cn">TRUE</span>,
  plot_theme <span class="op">=</span> <span class="fu"><a href="theme_dartR.html">theme_dartR</a></span><span class="op">(</span><span class="op">)</span>,
  plot_colors <span class="op">=</span> <span class="va">two_colors</span>,
  save2tmp <span class="op">=</span> <span class="cn">FALSE</span>,
  verbose <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>Name of the genlight object containing the SNP or presence/absence
(SilicoDArT) data [required].</p></dd>
<dt>metric</dt>
<dd><p>Name of the metric to be used for filtering [required].</p></dd>
<dt>plot.out</dt>
<dd><p>Specify if plot is to be produced [default TRUE].</p></dd>
<dt>plot_theme</dt>
<dd><p>Theme for the plot. See Details for options
[default theme_dartR()].</p></dd>
<dt>plot_colors</dt>
<dd><p>List of two color names for the borders and fill of the
plots [default two_colors].</p></dd>
<dt>save2tmp</dt>
<dd><p>If TRUE, saves any ggplots and listings to the session
temporary directory (tempdir) [default FALSE].</p></dd>
<dt>verbose</dt>
<dd><p>Verbosity: 0, silent or fatal errors; 1, begin and end; 2,
progress log; 3, progress and results summary; 5, full report
[default NULL, unless specified using gl.set.verbosity].</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An unaltered genlight object.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function <code><a href="gl.filter.locmetrics.html">gl.filter.locmetric</a></code> will filter out the
 loci with a locmetric value below a specified threshold.</p>
<p>The fields that are included in dartR, and a short description, are found
below. Optionally, the user can also set his/her own field by adding a vector
into $other$loc.metrics as shown in the example. You can check the names of
all available loc.metrics via: names(gl$other$loc.metrics).</p>
<ul><li><p>SnpPosition - position (zero is position 1) in the sequence tag of the
defined SNP variant base.</p></li>
<li><p>CallRate - proportion of samples for which the genotype call is
non-missing (that is, not '-' ).</p></li>
<li><p>OneRatioRef - proportion of samples for which the genotype score is 0.</p></li>
<li><p>OneRatioSnp - proportion of samples for which the genotype score is 2.</p></li>
<li><p>FreqHomRef - proportion of samples homozygous for the Reference allele.</p></li>
<li><p>FreqHomSnp - proportion of samples homozygous for the Alternate (SNP)
allele.</p></li>
<li><p>FreqHets - proportion of samples which score as heterozygous, that is,
scored as 1.</p></li>
<li><p>PICRef - polymorphism information content (PIC) for the Reference allele.</p></li>
<li><p>PICSnp - polymorphism information content (PIC) for the SNP.</p></li>
<li><p>AvgPIC - average of the polymorphism information content (PIC) of the
reference and SNP alleles.</p></li>
<li><p>AvgCountRef - sum of the tag read counts for all samples, divided by the
number of samples with non-zero tag read counts, for the Reference allele row.</p></li>
<li><p>AvgCountSnp - sum of the tag read counts for all samples, divided by the
number of samples with non-zero tag read counts, for the Alternate (SNP) allele
row.</p></li>
<li><p>RepAvg - proportion of technical replicate assay pairs for which the
marker score is consistent.</p></li>
<li><p>rdepth - read depth.</p></li>
</ul><p><strong> Function's output </strong></p>
<p>The minimum, maximum, mean and a tabulation of quantiles of the locmetric
values against thresholds rate are provided. Output also includes a boxplot
and a histogram.</p>
<p>Quantiles are partitions of a finite set of values into q subsets of (nearly)
equal sizes. In this function q = 20. Quantiles are useful measures because
they are less susceptible to long-tailed distributions and outliers.</p>
<p>Plots and table were saved to the temporal directory (tempdir) and can be
 accessed with the function <code><a href="gl.print.reports.html">gl.print.reports</a></code> and listed with
 the function <code><a href="gl.list.reports.html">gl.list.reports</a></code>. Note that they can be accessed
 only in the current R session because tempdir is cleared each time that the
 R session is closed.</p>
<p>Examples of other themes that can be used can be consulted in:</p><ul><li><p><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">https://ggplot2.tidyverse.org/reference/ggtheme.html</a> and</p></li>
<li><p><a href="https://yutannihilation.github.io/allYourFigureAreBelongToUs/ggthemes/" class="external-link">https://yutannihilation.github.io/allYourFigureAreBelongToUs/ggthemes/</a></p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="gl.filter.locmetrics.html">gl.filter.locmetric</a></code>, <code><a href="gl.list.reports.html">gl.list.reports</a></code>,
 <code><a href="gl.print.reports.html">gl.print.reports</a></code></p>
<p>Other filters and filter reports: 
<code><a href="gl.filter.allna.html">gl.filter.allna</a>()</code>,
<code><a href="gl.filter.monomorphs.html">gl.filter.monomorphs</a>()</code>,
<code><a href="gl.filter.rdepth.html">gl.filter.rdepth</a>()</code>,
<code><a href="gl.filter.reproducibility.html">gl.filter.reproducibility</a>()</code>,
<code><a href="gl.filter.secondaries.html">gl.filter.secondaries</a>()</code>,
<code><a href="gl.report.callrate.html">gl.report.callrate</a>()</code>,
<code><a href="gl.report.hamming.html">gl.report.hamming</a>()</code>,
<code><a href="gl.report.monomorphs.html">gl.report.monomorphs</a>()</code>,
<code><a href="gl.report.rdepth.html">gl.report.rdepth</a>()</code>,
<code><a href="gl.report.reproducibility.html">gl.report.reproducibility</a>()</code>,
<code><a href="gl.report.taglength.html">gl.report.taglength</a>()</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Luis Mijangos (Post to <a href="https://groups.google.com/d/forum/dartr" class="external-link">https://groups.google.com/d/forum/dartr</a>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># adding dummy data</span></span>
<span class="r-in"><span class="va">test</span> <span class="op">&lt;-</span> <span class="va">testset.gl</span></span>
<span class="r-in"><span class="va">test</span><span class="op">$</span><span class="va">other</span><span class="op">$</span><span class="va">loc.metrics</span><span class="op">$</span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nLoc</a></span><span class="op">(</span><span class="va">test</span><span class="op">)</span></span>
<span class="r-in"><span class="co"># SNP data</span></span>
<span class="r-in"><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">gl.report.locmetric</span><span class="op">(</span><span class="va">test</span>,metric<span class="op">=</span><span class="st">'test'</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.locmetric </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with SNP data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Reporting test by Locus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   No. of loci = 255 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   No. of individuals = 250 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Minimum      :  Min.   :  1.0   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     1st quantile :  1st Qu.: 64.5   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Median       :  Median :128.0   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Mean         :  Mean   :128.0   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     3r quantile  :  3rd Qu.:191.5   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Maximum      :  Max.   :255.0   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.locmetric-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Quantile Threshold Retained Percent Filtered Percent</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      100%       255        1     0.4      254    99.6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       95%       243       13     5.1      242    94.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       90%       230       26    10.2      229    89.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       85%       217       39    15.3      216    84.7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       80%       204       52    20.4      203    79.6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       75%       192       64    25.1      191    74.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7       70%       179       77    30.2      178    69.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8       65%       166       90    35.3      165    64.7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9       60%       154      102    40.0      153    60.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10      55%       141      115    45.1      140    54.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11      50%       128      128    50.2      127    49.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12      45%       115      141    55.3      114    44.7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13      40%       102      154    60.4      101    39.6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14      35%        90      166    65.1       89    34.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15      30%        77      179    70.2       76    29.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16      25%        64      192    75.3       63    24.7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17      20%        51      205    80.4       50    19.6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18      15%        39      217    85.1       38    14.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19      10%        26      230    90.2       25     9.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20       5%        13      243    95.3       12     4.7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21       0%         1      255   100.0        0     0.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.locmetric </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># adding dummy data</span></span>
<span class="r-in"><span class="va">test.gs</span> <span class="op">&lt;-</span> <span class="va">testset.gs</span></span>
<span class="r-in"><span class="va">test.gs</span><span class="op">$</span><span class="va">other</span><span class="op">$</span><span class="va">loc.metrics</span><span class="op">$</span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/adegenet/man/accessors.html" class="external-link">nLoc</a></span><span class="op">(</span><span class="va">test.gs</span><span class="op">)</span></span>
<span class="r-in"><span class="co"># Tag P/A data</span></span>
<span class="r-in"><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">gl.report.locmetric</span><span class="op">(</span><span class="va">test.gs</span>,metric<span class="op">=</span><span class="st">'test'</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Starting gl.report.locmetric </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">  Processing genlight object with Presence/Absence (SilicoDArT) data</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Reporting test by Locus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   No. of loci = 255 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   No. of individuals = 218 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Minimum      :  Min.   :  1.0   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     1st quantile :  1st Qu.: 64.5   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Median       :  Median :128.0   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Mean         :  Mean   :128.0   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     3r quantile  :  3rd Qu.:191.5   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Maximum      :  Max.   :255.0   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-plt img"><img src="gl.report.locmetric-2.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Quantile Threshold Retained Percent Filtered Percent</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      100%       255        1     0.4      254    99.6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2       95%       243       13     5.1      242    94.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       90%       230       26    10.2      229    89.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       85%       217       39    15.3      216    84.7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5       80%       204       52    20.4      203    79.6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6       75%       192       64    25.1      191    74.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7       70%       179       77    30.2      178    69.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8       65%       166       90    35.3      165    64.7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9       60%       154      102    40.0      153    60.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10      55%       141      115    45.1      140    54.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11      50%       128      128    50.2      127    49.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12      45%       115      141    55.3      114    44.7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13      40%       102      154    60.4      101    39.6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14      35%        90      166    65.1       89    34.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15      30%        77      179    70.2       76    29.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16      25%        64      192    75.3       63    24.7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17      20%        51      205    80.4       50    19.6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18      15%        39      217    85.1       38    14.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19      10%        26      230    90.2       25     9.8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20       5%        13      243    95.3       12     4.7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21       0%         1      255   100.0        0     0.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">Completed: gl.report.locmetric </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Bernd Gruber, Arthur Georges, Jose L. Mijangos, Carlo Pacioni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer></div>

  

  

  </body></html>

